// dts-backoffice-util/esm2015/lib/services/breadcrumb-control.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
Versão: 1.003
Data Criação: 06/08/2018
*/
import { Injectable } from '@angular/core';
export class BreadcrumbControlService {
    constructor() { }
    /**
     * @return {?}
     */
    newBreadcrumb() {
        this.breadcrumb = { items: [] };
    }
    /**
     * @param {?} literal
     * @param {?} activatedRoute
     * @return {?}
     */
    addBreadcrumb(literal, activatedRoute) {
        this.addBreadcrumbURL(literal, decodeURIComponent(activatedRoute.snapshot['_routerState'].url));
    }
    /**
     * @param {?} literal
     * @param {?} url
     * @return {?}
     */
    addBreadcrumbURL(literal, url) {
        if (!literal || literal === '') {
            return;
        }
        if (!this.breadcrumb) {
            this.newBreadcrumb();
        }
        /** @type {?} */
        const breadcrumbAux = { items: [] };
        /** @type {?} */
        let idx = 0;
        for (idx = 0; idx < this.breadcrumb.items.length; idx++) {
            if (this.breadcrumb.items[idx].label !== literal) {
                breadcrumbAux.items.push(this.breadcrumb.items[idx]);
            }
            else {
                break;
            }
        }
        breadcrumbAux.items.push({ label: literal, link: url });
        this.breadcrumb = breadcrumbAux;
    }
    /**
     * @param {?} literal
     * @param {?} valueOld
     * @param {?} valueNew
     * @return {?}
     */
    updBreadcrumbURL(literal, valueOld, valueNew) {
        /** @type {?} */
        let currentRouterURL = this.getCurrentRouter();
        if (!currentRouterURL) {
            return;
        }
        if (!valueOld) {
            return;
        }
        if (!valueNew) {
            return;
        }
        currentRouterURL = currentRouterURL.replace(valueOld, valueNew);
        this.addBreadcrumbURL(literal, currentRouterURL);
    }
    /**
     * @return {?}
     */
    getBreadcrumb() {
        if (!this.breadcrumb) {
            this.newBreadcrumb();
        }
        return this.breadcrumb;
    }
    /**
     * @return {?}
     */
    getCurrentRouter() {
        if (!this.breadcrumb) {
            this.newBreadcrumb();
        }
        if (this.breadcrumb.items.length <= 0) {
            return '';
        }
        return this.breadcrumb.items[this.breadcrumb.items.length - 1].link;
    }
    /**
     * @return {?}
     */
    getPrevRouter() {
        if (!this.breadcrumb) {
            this.newBreadcrumb();
        }
        if (this.breadcrumb.items.length <= 1) {
            return '';
        }
        return this.breadcrumb.items[this.breadcrumb.items.length - 2].link;
    }
    /**
     * @return {?}
     */
    hasPreviousRouter() {
        return this.breadcrumb.items.length > 1;
    }
}
BreadcrumbControlService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
BreadcrumbControlService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    BreadcrumbControlService.prototype.breadcrumb;
}
