// dts-backoffice-util/esm2015/lib/services/cache-params.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class CacheParamsService {
    constructor() {
        this.values = {};
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    setValue(key, value) {
        this.values[key] = value;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getValue(key) {
        return this.values[key];
    }
    /**
     * @param {?} key
     * @return {?}
     */
    removeValue(key) {
        delete this.values[key];
    }
}
CacheParamsService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
CacheParamsService.ctorParameters = () => [];
/** @nocollapse */ CacheParamsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CacheParamsService_Factory() { return new CacheParamsService(); }, token: CacheParamsService, providedIn: "root" });
if (false) {
    /** @type {?} */
    CacheParamsService.prototype.values;
}
