// dts-backoffice-util/esm2015/lib/services/menu-datasul.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class MenuDatasulService {
    constructor() { }
    /**
     * @param {?} program
     * @return {?}
     */
    callProgress(program) {
        /**
         * var program = {};
         * program['prg'] = "cd0210";
         * program['params'] = '01|123456'
         */
        parent.postMessage({ program }, '*');
    }
    /**
     * @param {?} notification
     * @return {?}
     */
    sendNotification(notification) {
        /**
         * var notification = {};
         * notification['type'] = 'success';
         * notification['title'] = 'Notification Title'
         * notification['detail'] = 'Notification Detail'
         */
        parent.postMessage({ notification }, '*');
    }
    /**
     * @param {?} programName
     * @param {?} params
     * @param {?=} parent
     * @return {?}
     */
    openPath(programName, params, parent = true) {
        /** @type {?} */
        const datasulPath = window.location.href.indexOf('totvs-menu') > 0 ? 'totvs-menu' : 'menu-html';
        /** @type {?} */
        const baseUrl = `/${datasulPath}/#/${datasulPath}/program-html/${programName}/#`;
        if (parent) {
            window.open(`${baseUrl}/${params}`, '_parent');
        }
        else {
            window.open(`${baseUrl}/${params}`);
        }
    }
}
MenuDatasulService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
MenuDatasulService.ctorParameters = () => [];
