// dts-backoffice-util/esm2015/lib/services/report.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators/map';
/**
 * @record
 */
export function IProperty() { }
if (false) {
    /** @type {?} */
    IProperty.prototype.name;
    /** @type {?} */
    IProperty.prototype.value;
}
/**
 * @record
 */
export function IReportServiceParams() { }
if (false) {
    /** @type {?} */
    IReportServiceParams.prototype.reportName;
    /** @type {?} */
    IReportServiceParams.prototype.programName;
    /** @type {?} */
    IReportServiceParams.prototype.properties;
    /** @type {?} */
    IReportServiceParams.prototype.dialect;
    /** @type {?} */
    IReportServiceParams.prototype.downloadName;
    /** @type {?} */
    IReportServiceParams.prototype.download;
    /** @type {?} */
    IReportServiceParams.prototype.format;
}
/** @enum {string} */
const ReportFormats = {
    XLSX: 'xlsx',
    PDF: 'pdf',
    DOCX: 'docx',
    HTML: 'html',
};
export { ReportFormats };
export class ReportService {
    /**
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
        this.URL = window.location.href.indexOf('totvs-menu') > 0 ? '/totvs-menu/rest/report/run/'
            : '/dts/datasul-report/resources/run/';
    }
    /**
     * @param {?} params
     * @param {?=} showLoading
     * @return {?}
     */
    generate(params, showLoading = true) {
        /** @type {?} */
        const headers = { 'X-PO-Screen-Lock': showLoading ? 'true' : 'false' };
        /** @type {?} */
        let reportURL = `${this.URL}${params.reportName}`;
        if (params.properties.length > 0) {
            reportURL += '?';
            params.properties.forEach((/**
             * @param {?} property
             * @param {?} index
             * @return {?}
             */
            (property, index) => {
                if (index > 0) {
                    reportURL += '&';
                }
                reportURL += `c_properties=${property.name}&c_values=${property.value}`;
            }));
        }
        reportURL += `&dialect=${params.dialect}&format=${params.format}`;
        reportURL += `&program=${params.programName}&resultFileName=${params.downloadName}`;
        return this.httpClient.post(reportURL, {}, { headers, responseType: 'blob' })
            .pipe(map((/**
         * @param {?} report
         * @return {?}
         */
        report => {
            if (params.download) {
                this.download(report, `${params.downloadName}.${params.format}`);
            }
            return report;
        })));
    }
    /**
     * @param {?} file
     * @param {?} fileName
     * @return {?}
     */
    download(file, fileName) {
        /** @type {?} */
        const binaryData = [file];
        /** @type {?} */
        const downloadLink = document.createElement('a');
        /** @type {?} */
        const urlDownload = window.URL.createObjectURL(new Blob(binaryData, { type: file.type }));
        downloadLink.href = urlDownload;
        downloadLink.setAttribute('download', fileName);
        document.body.appendChild(downloadLink);
        downloadLink.click();
        window.URL.revokeObjectURL(urlDownload);
        downloadLink.remove();
    }
}
ReportService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ReportService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /** @type {?} */
    ReportService.prototype.URL;
    /** @type {?} */
    ReportService.prototype.httpClient;
}
