// dts-backoffice-util/esm2015/lib/services/session-info.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { of } from 'rxjs';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class UserLoginService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.apiUrl = '/totvs-menu/rest/getSessionInfo';
        this.userLogin = undefined;
    }
    /**
     * @return {?}
     */
    getUserLogin() {
        if (localStorage.getItem('username')) {
            return of(localStorage.getItem('username'));
        }
        if (this.userLogin) {
            return of(this.userLogin);
        }
        else {
            return this.http.get(`${this.apiUrl}`)
                .pipe(map((/**
             * @param {?} response
             * @return {?}
             */
            (response) => {
                this.userLogin = response.userName;
                return response.userName;
            })));
        }
    }
}
UserLoginService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
UserLoginService.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ UserLoginService.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserLoginService_Factory() { return new UserLoginService(i0.ɵɵinject(i1.HttpClient)); }, token: UserLoginService, providedIn: "root" });
if (false) {
    /** @type {?} */
    UserLoginService.prototype.apiUrl;
    /** @type {?} */
    UserLoginService.prototype.userLogin;
    /** @type {?} */
    UserLoginService.prototype.http;
}
