// @po-ui/ng-templates/esm5/lib/components/po-page-dynamic-detail/po-page-dynamic-detail-actions.service.js
import { __decorate, __metadata } from "tslib";
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
var PoPageDynamicDetailActionsService = /** @class */ (function () {
    function PoPageDynamicDetailActionsService(http) {
        this.http = http;
        this.headers = new HttpHeaders({
            'X-PO-SCREEN-LOCK': 'true'
        });
    }
    PoPageDynamicDetailActionsService.prototype.beforeBack = function (action) {
        return this.executeAction({ action: action });
    };
    PoPageDynamicDetailActionsService.prototype.beforeEdit = function (action, id, body) {
        var resource = body !== null && body !== void 0 ? body : {};
        return this.executeAction({ action: action, resource: resource, id: id });
    };
    PoPageDynamicDetailActionsService.prototype.beforeRemove = function (action, id, body) {
        var resource = body !== null && body !== void 0 ? body : {};
        return this.executeAction({ action: action, resource: resource, id: id });
    };
    PoPageDynamicDetailActionsService.prototype.executeAction = function (_a) {
        var action = _a.action, _b = _a.resource, resource = _b === void 0 ? {} : _b, id = _a.id;
        if (!action) {
            return of({});
        }
        if (typeof action === 'string') {
            var url = id ? action + "/" + id : action;
            return this.http.post(url, resource, { headers: this.headers });
        }
        return of(action(id, resource));
    };
    PoPageDynamicDetailActionsService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    PoPageDynamicDetailActionsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PoPageDynamicDetailActionsService_Factory() { return new PoPageDynamicDetailActionsService(i0.ɵɵinject(i1.HttpClient)); }, token: PoPageDynamicDetailActionsService, providedIn: "root" });
    PoPageDynamicDetailActionsService = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __metadata("design:paramtypes", [HttpClient])
    ], PoPageDynamicDetailActionsService);
    return PoPageDynamicDetailActionsService;
}());
export { PoPageDynamicDetailActionsService };
