// @po-ui/ng-templates/esm5/lib/components/po-page-dynamic-edit/po-page-dynamic-edit.component.js
import { __assign, __decorate, __metadata, __read, __spread } from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { concat, of, EMPTY, throwError } from 'rxjs';
import { tap, catchError, map, switchMap } from 'rxjs/operators';
import { PoBreadcrumb, PoDialogService, PoDynamicFormComponent, PoGridComponent, PoGridRowActions, PoNotificationService, PoPageAction } from '@po-ui/ng-components';
import * as util from './../../utils/util';
import { PoPageDynamicService } from '../../services/po-page-dynamic/po-page-dynamic.service';
import { PoPageCustomizationService } from '../../services/po-page-customization/po-page-customization.service';
import { PoPageDynamicEditActionsService } from './po-page-dynamic-edit-actions.service';
export var poPageDynamicEditLiteralsDefault = {
    en: {
        cancelConfirmMessage: 'Are you sure you want to cancel this operation?',
        detailActionNew: 'New',
        pageActionCancel: 'Cancel',
        pageActionSave: 'Save',
        pageActionSaveNew: 'Save and new',
        registerNotFound: 'Register not found.',
        saveNotificationSuccessSave: 'Resource successfully saved.',
        saveNotificationSuccessUpdate: 'Resource successfully updated.',
        saveNotificationWarning: 'Form must be filled out correctly.'
    },
    es: {
        cancelConfirmMessage: 'Está seguro de que desea cancelar esta operación?',
        detailActionNew: 'Nuevo',
        pageActionCancel: 'Cancelar',
        pageActionSave: 'Guardar',
        pageActionSaveNew: 'Guardar y nuevo',
        registerNotFound: 'Registro no encontrado.',
        saveNotificationSuccessSave: 'Recurso salvo con éxito.',
        saveNotificationSuccessUpdate: 'Recurso actualizado con éxito.',
        saveNotificationWarning: 'El formulario debe llenarse correctamente.'
    },
    pt: {
        cancelConfirmMessage: 'Tem certeza que deseja cancelar esta operação?',
        detailActionNew: 'Novo',
        pageActionCancel: 'Cancelar',
        pageActionSave: 'Salvar',
        pageActionSaveNew: 'Salvar e novo',
        registerNotFound: 'Registro não encontrado.',
        saveNotificationSuccessSave: 'Recurso salvo com sucesso.',
        saveNotificationSuccessUpdate: 'Recurso atualizado com sucesso.',
        saveNotificationWarning: 'Formulário precisa ser preenchido corretamente.'
    }
};
/**
 * @description
 *
 * O `po-page-dynamic-edit` é uma página que pode servir para editar ou criar novos registros,
 * o mesmo também suporta metadados conforme especificado na documentação.
 *
 * ### Utilização via rota
 *
 * Ao utilizar as rotas para carregar o template, o `page-dynamic-edit` disponibiliza propriedades para
 * poder especificar o endpoint dos dados e dos metadados. Exemplo de utilização:
 *
 * O componente primeiro irá carregar o metadado da rota definida na propriedade serviceMetadataApi
 * e depois irá buscar da rota definida na propriedade serviceLoadApi
 *
 * > Caso o servidor retornar um erro ao recuperar o metadados, será repassado o metadados salvo em cache,
 * se o cache não existe será disparado uma notificação.
 *
 * ```
 * {
 *   path: 'people',
 *   component: PoPageDynamicEditComponent,
 *   data: {
 *     serviceApi: 'http://localhost:3000/v1/people', // endpoint dos dados
 *     serviceMetadataApi: 'http://localhost:3000/v1/metadata', // endpoint dos metadados utilizando o método HTTP Get
 *     serviceLoadApi: 'http://localhost:3000/load-metadata' // endpoint de customizações dos metadados utilizando o método HTTP Post
 *   }
 * }
 *
 * ```
 *
 * Para carregar com um recurso já existente, deve-se ser incluído um parâmetro na rota chamado `id`:
 *
 * ```
 * {
 *   path: 'people/:id',
 *   component: PoPageDynamicEditComponent,
 *   data: {
 *     serviceApi: 'http://localhost:3000/v1/people', // endpoint dos dados
 *     serviceMetadataApi: 'http://localhost:3000/v1/metadata', // endpoint dos metadados
 *     serviceLoadApi: 'http://localhost:3000/load-metadata' // endpoint de customizações dos metadados
 *   }
 * }
 * ```
 *
 * A requisição dos metadados é feita na inicialização do template para buscar os metadados da página passando o
 * tipo do metadado esperado e a versão cacheada pelo browser.
 *
 * O formato esperado na resposta da requisição está especificado na interface
 * [PoPageDynamicEditMetadata](/documentation/po-page-dynamic-edit#po-page-dynamic-edit-metadata). Por exemplo:
 *
 * ```
 *  {
 *   version: 1,
 *   title: 'Person edit',
 *   fields: [
 *     { property: 'id', key: true, disabled: true },
 *     { property: 'status' },
 *     { property: 'name' },
 *     { property: 'nickname' },
 *     { property: 'birthdate', label: 'Birth date' },
 *     { property: 'genre' },
 *     { property: 'city' },
 *     { property: 'country' }
 *   ]
 * }
 * ```
 *
 * > Caso o endpoint dos metadados não seja especificado, será feito uma requisição utilizando o `serviceApi` da seguinte forma:
 * ```
 * GET {end-point}/metadata?type=edit&version={version}
 * ```
 *
 * @example
 *
 * <example name="po-page-dynamic-edit-basic" title="PO Page Dynamic Edit Basic">
 *  <file name="sample-po-page-dynamic-edit-basic/sample-po-page-dynamic-edit-basic.component.html"> </file>
 *  <file name="sample-po-page-dynamic-edit-basic/sample-po-page-dynamic-edit-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-dynamic-edit-user" title="PO Page Dynamic Edit - User">
 *  <file name="sample-po-page-dynamic-edit-user/sample-po-page-dynamic-edit-user.component.html"> </file>
 *  <file name="sample-po-page-dynamic-edit-user/sample-po-page-dynamic-edit-user.component.ts"> </file>
 * </example>
 */
var PoPageDynamicEditComponent = /** @class */ (function () {
    function PoPageDynamicEditComponent(router, activatedRoute, poNotification, poDialogService, poPageDynamicService, poPageCustomizationService, poPageDynamicEditActionsService) {
        this.router = router;
        this.activatedRoute = activatedRoute;
        this.poNotification = poNotification;
        this.poDialogService = poDialogService;
        this.poPageDynamicService = poPageDynamicService;
        this.poPageCustomizationService = poPageCustomizationService;
        this.poPageDynamicEditActionsService = poPageDynamicEditActionsService;
        this.subscriptions = [];
        this._actions = {};
        this._autoRouter = false;
        this._controlFields = [];
        this._detailFields = [];
        this._duplicates = [];
        this._fields = [];
        this._keys = [];
        this._pageActions = [];
        this.literals = __assign(__assign({}, poPageDynamicEditLiteralsDefault[util.poLocaleDefault]), poPageDynamicEditLiteralsDefault[util.browserLanguage()]);
        this.model = {};
        // beforeSave: return boolean
        // afterSave
        // beforeRemove: return boolean
        // afterRemove
        // beforeInsert: : return boolean
        this.detailActions = {};
        /** Objeto com propriedades do breadcrumb. */
        this.breadcrumb = { items: [] };
    }
    Object.defineProperty(PoPageDynamicEditComponent.prototype, "actions", {
        get: function () {
            return __assign({}, this._actions);
        },
        /**
         * @optional
         *
         * @description
         *
         * Ações da página.
         */
        set: function (value) {
            this._actions = this.isObject(value) ? value : {};
            this._pageActions = this.getPageActions(this._actions);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicEditComponent.prototype, "autoRouter", {
        get: function () {
            return this._autoRouter;
        },
        /**
         * @todo Validar rotas na mão pois se existir uma rota '**' o catch do navigation não funciona.
         *
         * @optional
         *
         * @description
         *
         * Cria automaticamente as rotas de edição (novo/duplicate) e detalhes caso as ações
         * estejam definidas nas ações.
         *
         * > Para o correto funcionamento não pode haver nenhum rota coringa (`**`) especificada.
         *
         * @default false
         */
        set: function (value) {
            this._autoRouter = util.convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicEditComponent.prototype, "fields", {
        get: function () {
            return this._fields;
        },
        /** Lista dos campos usados na tabela e busca avançada. */
        set: function (value) {
            this._fields = Array.isArray(value) ? __spread(value) : [];
            this._keys = this.getKeysByFields(this._fields);
            this._duplicates = this.getDuplicatesByFields(this._fields);
            this._controlFields = this.getControlFields(this._fields);
            this._detailFields = this.getDetailFields(this._fields);
        },
        enumerable: true,
        configurable: true
    });
    PoPageDynamicEditComponent.prototype.ngOnInit = function () {
        this.loadDataFromAPI();
    };
    PoPageDynamicEditComponent.prototype.ngOnDestroy = function () {
        if (this.subscriptions) {
            this.subscriptions.forEach(function (subscription) {
                subscription.unsubscribe();
            });
        }
    };
    PoPageDynamicEditComponent.prototype.loadDataFromAPI = function () {
        var _a = this.activatedRoute.snapshot.data, serviceApiFromRoute = _a.serviceApi, serviceMetadataApi = _a.serviceMetadataApi, serviceLoadApi = _a.serviceLoadApi;
        var id = this.activatedRoute.snapshot.params.id;
        var duplicate = this.activatedRoute.snapshot.queryParams.duplicate;
        var onLoad = serviceLoadApi || this.onLoad;
        this.serviceApi = serviceApiFromRoute || this.serviceApi;
        this.poPageDynamicService.configServiceApi({ endpoint: this.serviceApi, metadata: serviceMetadataApi });
        var metadata$ = this.getMetadata(serviceApiFromRoute, id, onLoad);
        var data$ = this.loadData(id, duplicate);
        this.subscriptions.push(concat(metadata$, data$).subscribe());
    };
    Object.defineProperty(PoPageDynamicEditComponent.prototype, "duplicates", {
        get: function () {
            return __spread(this._duplicates);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicEditComponent.prototype, "keys", {
        get: function () {
            return __spread(this._keys);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicEditComponent.prototype, "pageActions", {
        get: function () {
            return __spread(this._pageActions);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicEditComponent.prototype, "controlFields", {
        get: function () {
            return this._controlFields;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicEditComponent.prototype, "detailFields", {
        get: function () {
            return this._detailFields;
        },
        enumerable: true,
        configurable: true
    });
    PoPageDynamicEditComponent.prototype.detailActionNew = function () {
        this.gridDetail.insertRow();
    };
    PoPageDynamicEditComponent.prototype.cancel = function (actionCancel, actionBeforeCancel) {
        if (this.dynamicForm && this.dynamicForm.form.dirty) {
            this.poDialogService.confirm({
                message: this.literals.cancelConfirmMessage,
                title: this.literals.pageActionCancel,
                confirm: this.goBack.bind(this, actionCancel, actionBeforeCancel)
            });
        }
        else {
            this.goBack(actionCancel, actionBeforeCancel);
        }
    };
    PoPageDynamicEditComponent.prototype.formatUniqueKey = function (item) {
        var keys = util.mapObjectByProperties(item, this.keys);
        return util.valuesFromObject(keys).join('|');
    };
    PoPageDynamicEditComponent.prototype.goBack = function (actionCancel, actionBeforeCancel) {
        var _this = this;
        this.subscriptions.push(this.poPageDynamicEditActionsService
            .beforeCancel(actionBeforeCancel)
            .subscribe(function (beforeCancelResult) {
            _this.executeBackAction(actionCancel, beforeCancelResult === null || beforeCancelResult === void 0 ? void 0 : beforeCancelResult.allowAction, beforeCancelResult === null || beforeCancelResult === void 0 ? void 0 : beforeCancelResult.newUrl);
        }));
    };
    PoPageDynamicEditComponent.prototype.executeBackAction = function (actionCancel, allowAction, newUrl) {
        var isAllowedAction = typeof allowAction === 'boolean' ? allowAction : true;
        if (isAllowedAction) {
            if (actionCancel === undefined || typeof actionCancel === 'boolean') {
                return window.history.back();
            }
            if (typeof actionCancel === 'string' || newUrl) {
                return this.router.navigate([newUrl || actionCancel]);
            }
            return actionCancel();
        }
    };
    PoPageDynamicEditComponent.prototype.loadData = function (id, duplicate) {
        var _this = this;
        if (!id) {
            try {
                this.model = duplicate ? JSON.parse(duplicate) : {};
            }
            catch (_a) {
                this.model = {};
            }
            return EMPTY;
        }
        return this.poPageDynamicService.getResource(id).pipe(tap(function (response) { return (_this.model = response); }), catchError(function (error) {
            _this.model = undefined;
            _this.actions = undefined;
            _this._pageActions = [];
            return throwError(error);
        }));
    };
    PoPageDynamicEditComponent.prototype.loadOptionsOnInitialize = function (onLoad) {
        var _this = this;
        if (onLoad) {
            return this.getPoDynamicPageOptions(onLoad).pipe(tap(function (responsePoOption) {
                return _this.poPageCustomizationService.changeOriginalOptionsToNewOptions(_this, responsePoOption);
            }));
        }
        return EMPTY;
    };
    PoPageDynamicEditComponent.prototype.getPoDynamicPageOptions = function (onLoad) {
        var originalOption = {
            fields: this.fields,
            actions: this.actions,
            breadcrumb: this.breadcrumb,
            title: this.title
        };
        var pageOptionSchema = {
            schema: [
                {
                    nameProp: 'fields',
                    merge: true,
                    keyForMerge: 'property'
                },
                {
                    nameProp: 'actions',
                    merge: true
                },
                {
                    nameProp: 'breadcrumb'
                },
                {
                    nameProp: 'title'
                }
            ]
        };
        return this.poPageCustomizationService.getCustomOptions(onLoad, originalOption, pageOptionSchema);
    };
    PoPageDynamicEditComponent.prototype.getMetadata = function (serviceApiFromRoute, paramId, onLoad) {
        var _this = this;
        var typeMetadata = paramId ? 'edit' : 'create';
        if (serviceApiFromRoute) {
            return this.poPageDynamicService.getMetadata(typeMetadata).pipe(tap(function (response) {
                _this.autoRouter = response.autoRouter || _this.autoRouter;
                _this.actions = response.actions || _this.actions;
                _this.breadcrumb = response.breadcrumb || _this.breadcrumb;
                _this.fields = response.fields || _this.fields;
                _this.title = response.title || _this.title;
            }), switchMap(function () { return _this.loadOptionsOnInitialize(onLoad); }));
        }
        return this.loadOptionsOnInitialize(onLoad);
    };
    PoPageDynamicEditComponent.prototype.navigateTo = function (path) {
        if (path) {
            var url = this.resolveUrl(this.model, path);
            this.router.navigate([url]);
        }
        else {
            window.history.back();
        }
    };
    PoPageDynamicEditComponent.prototype.resolveUniqueKey = function (item) {
        return this.activatedRoute.snapshot.params['id'] ? this.formatUniqueKey(item) : undefined;
    };
    PoPageDynamicEditComponent.prototype.resolveUrl = function (item, path) {
        var uniqueKey = this.formatUniqueKey(item);
        return path.replace(/:id/g, uniqueKey);
    };
    PoPageDynamicEditComponent.prototype.executeSave = function (saveRedirectPath) {
        var _this = this;
        var saveOperation$ = this.saveOperation();
        return saveOperation$.pipe(tap(function (message) {
            _this.poNotification.success(message);
            _this.navigateTo(saveRedirectPath);
        }));
    };
    PoPageDynamicEditComponent.prototype.updateModel = function (newResource) {
        if (newResource === void 0) { newResource = {}; }
        var dynamicNgForm = this.dynamicForm.form;
        util.removeKeysProperties(this.keys, newResource);
        this.model = __assign(__assign({}, this.model), newResource);
        dynamicNgForm.form.patchValue(this.model);
    };
    PoPageDynamicEditComponent.prototype.saveOperation = function () {
        if (this.dynamicForm.form.invalid) {
            this.poNotification.warning(this.literals.saveNotificationWarning);
            return EMPTY;
        }
        var paramId = this.activatedRoute.snapshot.params['id'];
        var successMsg = paramId
            ? this.literals.saveNotificationSuccessUpdate
            : this.literals.saveNotificationSuccessSave;
        var saveOperation$ = paramId
            ? this.poPageDynamicService.updateResource(paramId, this.model)
            : this.poPageDynamicService.createResource(this.model);
        return saveOperation$.pipe(map(function () { return successMsg; }));
    };
    PoPageDynamicEditComponent.prototype.save = function (action, before) {
        var _this = this;
        if (before === void 0) { before = 'beforeSave'; }
        var executeOperation = {
            beforeSave: this.executeSave.bind(this),
            beforeSaveNew: this.executeSaveNew.bind(this)
        };
        var uniqueKey = this.resolveUniqueKey(this.model);
        this.subscriptions.push(this.poPageDynamicEditActionsService[before](this.actions[before], uniqueKey, __assign({}, this.model))
            .pipe(switchMap(function (returnBefore) {
            var _a, _b;
            var newAction = (_a = returnBefore === null || returnBefore === void 0 ? void 0 : returnBefore.newUrl) !== null && _a !== void 0 ? _a : action;
            var allowAction = (_b = returnBefore === null || returnBefore === void 0 ? void 0 : returnBefore.allowAction) !== null && _b !== void 0 ? _b : true;
            _this.updateModel(returnBefore === null || returnBefore === void 0 ? void 0 : returnBefore.resource);
            if (!allowAction) {
                return of({});
            }
            if (typeof newAction === 'string') {
                return executeOperation[before](newAction);
            }
            else {
                newAction(__assign({}, _this.model), uniqueKey);
                return EMPTY;
            }
        }))
            .subscribe());
    };
    PoPageDynamicEditComponent.prototype.executeSaveNew = function (path) {
        var _this = this;
        var paramId = this.activatedRoute.snapshot.params['id'];
        var saveOperation$ = this.saveOperation();
        return saveOperation$.pipe(tap(function (message) {
            if (paramId) {
                _this.poNotification.success(message);
                _this.navigateTo(path);
            }
            else {
                _this.poNotification.success(message);
                _this.model = {};
                _this.dynamicForm.form.reset();
            }
        }));
    };
    PoPageDynamicEditComponent.prototype.getKeysByFields = function (fields) {
        if (fields === void 0) { fields = []; }
        return fields.filter(function (field) { return field.key === true; }).map(function (field) { return field.property; });
    };
    PoPageDynamicEditComponent.prototype.getControlFields = function (fields) {
        if (fields === void 0) { fields = []; }
        return fields.filter(function (field) { return field.type !== 'detail'; });
    };
    PoPageDynamicEditComponent.prototype.getDetailFields = function (fields) {
        if (fields === void 0) { fields = []; }
        return fields.filter(function (field) { return field.type === 'detail'; });
    };
    PoPageDynamicEditComponent.prototype.getDuplicatesByFields = function (fields) {
        if (fields === void 0) { fields = []; }
        return fields.filter(function (field) { return field.duplicate === true; }).map(function (field) { return field.property; });
    };
    PoPageDynamicEditComponent.prototype.getPageActions = function (actions) {
        if (actions === void 0) { actions = {}; }
        var pageActions = [{ label: this.literals.pageActionSave, action: this.save.bind(this, actions.save) }];
        if (actions.saveNew) {
            pageActions.push({
                label: this.literals.pageActionSaveNew,
                action: this.save.bind(this, actions.saveNew, 'beforeSaveNew')
            });
        }
        if (actions.cancel === undefined || actions.cancel) {
            pageActions.push({
                label: this.literals.pageActionCancel,
                action: this.cancel.bind(this, actions.cancel, this.actions.beforeCancel)
            });
        }
        return pageActions;
    };
    PoPageDynamicEditComponent.prototype.isObject = function (value) {
        return !!value && typeof value === 'object' && !Array.isArray(value);
    };
    PoPageDynamicEditComponent.ctorParameters = function () { return [
        { type: Router },
        { type: ActivatedRoute },
        { type: PoNotificationService },
        { type: PoDialogService },
        { type: PoPageDynamicService },
        { type: PoPageCustomizationService },
        { type: PoPageDynamicEditActionsService }
    ]; };
    __decorate([
        Input('p-actions'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PoPageDynamicEditComponent.prototype, "actions", null);
    __decorate([
        Input('p-auto-router'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], PoPageDynamicEditComponent.prototype, "autoRouter", null);
    __decorate([
        Input('p-breadcrumb'),
        __metadata("design:type", Object)
    ], PoPageDynamicEditComponent.prototype, "breadcrumb", void 0);
    __decorate([
        Input('p-fields'),
        __metadata("design:type", Array),
        __metadata("design:paramtypes", [Array])
    ], PoPageDynamicEditComponent.prototype, "fields", null);
    __decorate([
        Input('p-service-api'),
        __metadata("design:type", String)
    ], PoPageDynamicEditComponent.prototype, "serviceApi", void 0);
    __decorate([
        Input('p-title'),
        __metadata("design:type", String)
    ], PoPageDynamicEditComponent.prototype, "title", void 0);
    __decorate([
        Input('p-load'),
        __metadata("design:type", Object)
    ], PoPageDynamicEditComponent.prototype, "onLoad", void 0);
    __decorate([
        ViewChild('dynamicForm'),
        __metadata("design:type", PoDynamicFormComponent)
    ], PoPageDynamicEditComponent.prototype, "dynamicForm", void 0);
    __decorate([
        ViewChild('gridDetail'),
        __metadata("design:type", PoGridComponent)
    ], PoPageDynamicEditComponent.prototype, "gridDetail", void 0);
    PoPageDynamicEditComponent = __decorate([
        Component({
            selector: 'po-page-dynamic-edit',
            template: "<po-page-default [p-actions]=\"pageActions\" [p-breadcrumb]=\"breadcrumb\" [p-title]=\"title\">\n  <ng-container *ngIf=\"model; then formFieldsTemplate; else registerNotFoundTemplate\"> </ng-container>\n</po-page-default>\n\n<ng-template #registerNotFoundTemplate>\n  <po-widget>\n    <span class=\"po-icon po-icon-info\"></span>\n    <span class=\"po-font-text-large\">\n      {{ literals.registerNotFound }}\n    </span>\n  </po-widget>\n</ng-template>\n\n<ng-template #formFieldsTemplate>\n  <po-dynamic-form #dynamicForm [p-fields]=\"controlFields\" [p-value]=\"model\"> </po-dynamic-form>\n\n  <div *ngIf=\"detailFields.length > 0\" class=\"po-sm-12\">\n    <po-divider [p-label]=\"detailFields[0].divider\"></po-divider>\n\n    <div class=\"po-row po-mb-2\">\n      <po-button [p-label]=\"literals.detailActionNew\" (p-click)=\"detailActionNew()\" p-small></po-button>\n    </div>\n\n    <po-grid\n      #gridDetail\n      [p-row-actions]=\"detailActions\"\n      [p-columns]=\"detailFields[0].detail.columns\"\n      [p-data]=\"model[detailFields[0].property]\"\n    >\n    </po-grid>\n  </div>\n</ng-template>\n",
            providers: [PoPageDynamicService]
        }),
        __metadata("design:paramtypes", [Router,
            ActivatedRoute,
            PoNotificationService,
            PoDialogService,
            PoPageDynamicService,
            PoPageCustomizationService,
            PoPageDynamicEditActionsService])
    ], PoPageDynamicEditComponent);
    return PoPageDynamicEditComponent;
}());
export { PoPageDynamicEditComponent };
