// @po-ui/ng-templates/esm5/lib/components/po-page-dynamic-edit/po-page-dynamic-edit-actions.service.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { of } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
var PoPageDynamicEditActionsService = /** @class */ (function () {
    function PoPageDynamicEditActionsService(http) {
        this.http = http;
        this.headers = new HttpHeaders({
            'X-PO-SCREEN-LOCK': 'true'
        });
    }
    PoPageDynamicEditActionsService.prototype.beforeCancel = function (action) {
        return this.executeAction({ action: action });
    };
    PoPageDynamicEditActionsService.prototype.beforeSave = function (action, id, body) {
        var resource = body !== null && body !== void 0 ? body : {};
        return this.executeAction({ action: action, resource: resource, id: id });
    };
    PoPageDynamicEditActionsService.prototype.beforeSaveNew = function (action, id, body) {
        var resource = body !== null && body !== void 0 ? body : {};
        return this.executeAction({ action: action, resource: resource, id: id });
    };
    PoPageDynamicEditActionsService.prototype.executeAction = function (_a) {
        var action = _a.action, _b = _a.resource, resource = _b === void 0 ? {} : _b, id = _a.id;
        if (!action) {
            return of({});
        }
        if (typeof action === 'string') {
            var url = id ? action + "/" + id : action;
            return this.http.post(url, resource, { headers: this.headers });
        }
        return of(action(resource, id));
    };
    PoPageDynamicEditActionsService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    PoPageDynamicEditActionsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PoPageDynamicEditActionsService_Factory() { return new PoPageDynamicEditActionsService(i0.ɵɵinject(i1.HttpClient)); }, token: PoPageDynamicEditActionsService, providedIn: "root" });
    PoPageDynamicEditActionsService = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __metadata("design:paramtypes", [HttpClient])
    ], PoPageDynamicEditActionsService);
    return PoPageDynamicEditActionsService;
}());
export { PoPageDynamicEditActionsService };
