// dts-backoffice-util/esm5/lib/dts-backoffice-util.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { TotvsScheduleExecutionComponent } from './components/totvs-schedule-execution/totvs-schedule-execution.component';
import { FormsModule } from '@angular/forms';
import { PoModule } from '@po-ui/ng-components';
import { HttpClientModule } from '@angular/common/http';
// import { BrowserModule } from '@angular/platform-browser';
import { RpwService } from './components/totvs-schedule-execution/totvs-schedule-execution.service';
import { CommonModule } from '@angular/common';
import { DtsDateFormatPipe } from './pipes/dts-date-format.pipe';
var DtsBackofficeUtilsModule = /** @class */ (function () {
    function DtsBackofficeUtilsModule() {
    }
    /**
     * @return {?}
     */
    DtsBackofficeUtilsModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: DtsBackofficeUtilsModule,
            providers: [RpwService]
        };
    };
    DtsBackofficeUtilsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        TotvsScheduleExecutionComponent,
                        DtsDateFormatPipe
                    ],
                    imports: [
                        CommonModule,
                        PoModule,
                        FormsModule,
                        HttpClientModule
                    ],
                    exports: [
                        TotvsScheduleExecutionComponent,
                        DtsDateFormatPipe
                    ],
                    providers: [
                        RpwService
                    ]
                },] }
    ];
    return DtsBackofficeUtilsModule;
}());
export { DtsBackofficeUtilsModule };
