// @po-ui/ng-templates/esm5/lib/components/po-page-job-scheduler/po-page-job-scheduler-execution/po-page-job-scheduler-execution.component.js
import { __decorate, __metadata } from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { NgForm } from '@angular/forms';
import { isTypeof } from '../../../utils/util';
import { PoPageJobSchedulerLookupService } from '../po-page-job-scheduler-lookup.service';
import { PoPageJobSchedulerService } from '../po-page-job-scheduler.service';
var PoPageJobSchedulerExecutionComponent = /** @class */ (function () {
    function PoPageJobSchedulerExecutionComponent(poPageJobSchedulerService, poPageJobSchedulerLookup) {
        this.poPageJobSchedulerService = poPageJobSchedulerService;
        this.poPageJobSchedulerLookup = poPageJobSchedulerLookup;
        this._value = {};
        this.dayPattern = '^(3[0-1]|[0-2][0-9]|[1-9]|0[1-9])$';
        this.existProcessAPI = true;
        this.minDateFirstExecution = new Date();
        this.periodicityOptions = [];
        this.timePattern = '^(2[0-3]|[01][0-9]):?([0-5][0-9])$';
        this.weekDays = [];
        this.isEdit = false;
        this.literals = {};
        this.changeProcess = new EventEmitter();
    }
    Object.defineProperty(PoPageJobSchedulerExecutionComponent.prototype, "value", {
        get: function () {
            return this._value;
        },
        set: function (value) {
            this._value = value && isTypeof(value, 'object') ? value : {};
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageJobSchedulerExecutionComponent.prototype, "startDateFirstExecution", {
        get: function () {
            return this.isEdit ? undefined : this.minDateFirstExecution;
        },
        enumerable: true,
        configurable: true
    });
    PoPageJobSchedulerExecutionComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        setTimeout(function () {
            _this.subscribeProcessIdValueChanges();
        });
    };
    PoPageJobSchedulerExecutionComponent.prototype.ngOnInit = function () {
        this.periodicityTemplates = {
            daily: this.dailyTemplate,
            monthly: this.monthlyTemplate,
            weekly: this.weeklyTempalte
        };
        this.checkExistsProcessesAPI();
        this.periodicityOptions = this.getPeriodicityOptions();
        this.weekDays = this.getWeekDays();
    };
    PoPageJobSchedulerExecutionComponent.prototype.checkExistsProcessesAPI = function () {
        var _this = this;
        this.poPageJobSchedulerService.getHeadProcesses().subscribe(undefined, function (error) {
            _this.existProcessAPI = false;
        });
    };
    PoPageJobSchedulerExecutionComponent.prototype.getPeriodicityOptions = function () {
        return [
            { label: this.literals.single, value: 'single' },
            { label: this.literals.daily, value: 'daily' },
            { label: this.literals.weekly, value: 'weekly' },
            { label: this.literals.monthly, value: 'monthly' }
        ];
    };
    PoPageJobSchedulerExecutionComponent.prototype.getWeekDays = function () {
        return [
            { label: this.literals.sunday, value: 'Sunday' },
            { label: this.literals.monday, value: 'Monday' },
            { label: this.literals.tuesday, value: 'Tuesday' },
            { label: this.literals.wednesday, value: 'Wednesday' },
            { label: this.literals.thursday, value: 'Thursday' },
            { label: this.literals.friday, value: 'Friday' },
            { label: this.literals.saturday, value: 'Saturday' }
        ];
    };
    PoPageJobSchedulerExecutionComponent.prototype.subscribeProcessIdValueChanges = function () {
        var _this = this;
        this.form.controls['processID'].valueChanges.subscribe(function (processId) {
            _this.changeProcess.emit({ processId: processId, existAPI: _this.existProcessAPI });
        });
    };
    PoPageJobSchedulerExecutionComponent.ctorParameters = function () { return [
        { type: PoPageJobSchedulerService },
        { type: PoPageJobSchedulerLookupService }
    ]; };
    __decorate([
        ViewChild('formExecution', { static: true }),
        __metadata("design:type", NgForm)
    ], PoPageJobSchedulerExecutionComponent.prototype, "form", void 0);
    __decorate([
        ViewChild('dailyTemplate', { static: true }),
        __metadata("design:type", Object)
    ], PoPageJobSchedulerExecutionComponent.prototype, "dailyTemplate", void 0);
    __decorate([
        ViewChild('monthlyTemplate', { static: true }),
        __metadata("design:type", Object)
    ], PoPageJobSchedulerExecutionComponent.prototype, "monthlyTemplate", void 0);
    __decorate([
        ViewChild('weeklyTemplate', { static: true }),
        __metadata("design:type", Object)
    ], PoPageJobSchedulerExecutionComponent.prototype, "weeklyTempalte", void 0);
    __decorate([
        Input('p-is-edit'),
        __metadata("design:type", Boolean)
    ], PoPageJobSchedulerExecutionComponent.prototype, "isEdit", void 0);
    __decorate([
        Input('p-literals'),
        __metadata("design:type", Object)
    ], PoPageJobSchedulerExecutionComponent.prototype, "literals", void 0);
    __decorate([
        Input('p-value'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PoPageJobSchedulerExecutionComponent.prototype, "value", null);
    __decorate([
        Output('p-change-process'),
        __metadata("design:type", EventEmitter)
    ], PoPageJobSchedulerExecutionComponent.prototype, "changeProcess", void 0);
    PoPageJobSchedulerExecutionComponent = __decorate([
        Component({
            selector: 'po-page-job-scheduler-execution',
            template: "<form #formExecution=\"ngForm\">\n  <div class=\"po-row\">\n    <ng-container *ngIf=\"existProcessAPI; then lookupProcessesID; else inputProcessesID\"></ng-container>\n  </div>\n\n  <div class=\"po-row\">\n    <po-divider class=\"po-md-12\" [p-label]=\"literals.firstExecution\"></po-divider>\n\n    <po-datepicker\n      class=\"po-md-4\"\n      name=\"firstExecution\"\n      [(ngModel)]=\"value.firstExecution\"\n      p-placeholder=\"dd/mm/aaaa\"\n      p-required\n      [p-disabled]=\"isEdit\"\n      [p-label]=\"literals.day\"\n      [p-min-date]=\"startDateFirstExecution\"\n    >\n    </po-datepicker>\n\n    <po-input\n      class=\"po-md-2\"\n      name=\"firstExecutionHour\"\n      [(ngModel)]=\"value.firstExecutionHour\"\n      p-mask=\"99:99\"\n      p-mask-format-model\n      p-placeholder=\"HH:mm\"\n      p-required\n      [p-disabled]=\"isEdit\"\n      [p-label]=\"literals.hour\"\n      [p-pattern]=\"timePattern\"\n    >\n    </po-input>\n  </div>\n\n  <hr />\n\n  <div class=\"po-row\">\n    <po-radio-group\n      class=\"po-sm-12\"\n      name=\"periodicity\"\n      [(ngModel)]=\"value.periodicity\"\n      p-columns=\"4\"\n      p-required\n      [p-label]=\"literals.periodicity\"\n      [p-options]=\"periodicityOptions\"\n    >\n    </po-radio-group>\n\n    <ng-container *ngIf=\"value.periodicity !== 'single'\">\n      <po-divider class=\"po-md-12\" [p-label]=\"literals.periodicityData\"> </po-divider>\n\n      <div class=\"po-row\">\n        <ng-container *ngTemplateOutlet=\"periodicityTemplates[value.periodicity]\"> </ng-container>\n      </div>\n\n      <div class=\"po-row\">\n        <po-switch\n          class=\"po-md-3\"\n          name=\"recurrent\"\n          [(ngModel)]=\"value.recurrent\"\n          [p-label]=\"literals.recurrent\"\n          [p-label-off]=\"literals.no\"\n          [p-label-on]=\"literals.yes\"\n        >\n        </po-switch>\n      </div>\n    </ng-container>\n  </div>\n\n  <ng-template #lookupProcessesID>\n    <po-lookup\n      class=\"po-md-12\"\n      name=\"processID\"\n      [(ngModel)]=\"value.processID\"\n      p-field-label=\"description\"\n      p-field-value=\"processID\"\n      p-required\n      [p-disabled]=\"isEdit\"\n      [p-filter-service]=\"poPageJobSchedulerLookup\"\n      [p-label]=\"literals.process\"\n      [p-placeholder]=\"literals.enterProcess\"\n    >\n    </po-lookup>\n  </ng-template>\n\n  <ng-template #inputProcessesID>\n    <po-input\n      class=\"po-md-12\"\n      name=\"processID\"\n      [(ngModel)]=\"value.processID\"\n      p-required\n      [p-disabled]=\"isEdit\"\n      [p-label]=\"literals.process\"\n      [p-placeholder]=\"literals.enterProcess\"\n    >\n    </po-input>\n  </ng-template>\n\n  <ng-template #dailyTemplate>\n    <ng-container *ngTemplateOutlet=\"inputHourTemplate\"> </ng-container>\n  </ng-template>\n\n  <ng-template #weeklyTemplate>\n    <div class=\"po-row\">\n      <ng-container *ngTemplateOutlet=\"inputHourTemplate\"> </ng-container>\n    </div>\n\n    <po-checkbox-group\n      class=\"po-md-12\"\n      name=\"daysOfWeek\"\n      [(ngModel)]=\"value.daysOfWeek\"\n      p-columns=\"4\"\n      p-required\n      [p-label]=\"literals.weekDays\"\n      [p-options]=\"weekDays\"\n    >\n    </po-checkbox-group>\n  </ng-template>\n\n  <ng-template #monthlyTemplate>\n    <po-number\n      class=\"po-md-3\"\n      name=\"dayOfMonth\"\n      [(ngModel)]=\"value.dayOfMonth\"\n      p-required\n      [p-error-pattern]=\"'Dia inv\u00E1lido'\"\n      [p-label]=\"literals.day\"\n      [p-max]=\"31\"\n      [p-pattern]=\"dayPattern\"\n    >\n    </po-number>\n\n    <ng-container *ngTemplateOutlet=\"inputHourTemplate\"> </ng-container>\n  </ng-template>\n\n  <ng-template #inputHourTemplate>\n    <po-input\n      class=\"po-md-2\"\n      name=\"hour\"\n      [(ngModel)]=\"value.hour\"\n      p-mask=\"99:99\"\n      p-mask-format-model\n      p-placeholder=\"HH:mm\"\n      p-required\n      [p-label]=\"literals.hour\"\n      [p-pattern]=\"timePattern\"\n    >\n    </po-input>\n  </ng-template>\n</form>\n"
        }),
        __metadata("design:paramtypes", [PoPageJobSchedulerService,
            PoPageJobSchedulerLookupService])
    ], PoPageJobSchedulerExecutionComponent);
    return PoPageJobSchedulerExecutionComponent;
}());
export { PoPageJobSchedulerExecutionComponent };
