// @po-ui/ng-templates/esm5/lib/components/po-page-dynamic-search/po-advanced-filter/po-advanced-filter.component.js
import { __decorate, __extends, __metadata } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { PoDynamicFormComponent, PoLanguageService } from '@po-ui/ng-components';
import { PoAdvancedFilterBaseComponent } from './po-advanced-filter-base.component';
/**
 * @docsPrivate
 *
 * @docsExtends PoAdvancedFilterBaseComponent
 *
 * @examplePrivate
 *
 * <example-private name="po-advanced-filter" title="PO Busca Avançada">
 *   <file name="sample-po-advanced-filter.component.html"> </file>
 *   <file name="sample-po-advanced-filter.component.ts"> </file>
 * </example-private>
 */
var PoAdvancedFilterComponent = /** @class */ (function (_super) {
    __extends(PoAdvancedFilterComponent, _super);
    function PoAdvancedFilterComponent(languageService) {
        return _super.call(this, languageService) || this;
    }
    PoAdvancedFilterComponent.prototype.open = function () {
        this.filter = this.keepFilters ? this.getInitialValuesFromFilter(this.filters) : {};
        this.poModal.open();
    };
    PoAdvancedFilterComponent.prototype.getInitialValuesFromFilter = function (filters) {
        return filters.reduce(function (result, item) {
            var _a;
            return Object.assign(result, (_a = {}, _a[item.property] = item.initValue, _a));
        }, {});
    };
    PoAdvancedFilterComponent.ctorParameters = function () { return [
        { type: PoLanguageService }
    ]; };
    __decorate([
        ViewChild(PoDynamicFormComponent, { static: true }),
        __metadata("design:type", PoDynamicFormComponent)
    ], PoAdvancedFilterComponent.prototype, "poDynamicForm", void 0);
    PoAdvancedFilterComponent = __decorate([
        Component({
            selector: 'po-advanced-filter',
            template: "<po-modal\n  p-hide-close\n  [p-primary-action]=\"primaryAction\"\n  [p-secondary-action]=\"secondaryAction\"\n  [p-title]=\"literals.title\"\n>\n  <po-dynamic-form [p-fields]=\"filters\" [p-value]=\"filter\"> </po-dynamic-form>\n</po-modal>\n"
        }),
        __metadata("design:paramtypes", [PoLanguageService])
    ], PoAdvancedFilterComponent);
    return PoAdvancedFilterComponent;
}(PoAdvancedFilterBaseComponent));
export { PoAdvancedFilterComponent };
