// @po-ui/ng-templates/esm5/lib/components/po-page-change-password/po-page-change-password.component.js
import { __assign, __decorate, __extends, __metadata } from "tslib";
import { AfterViewInit, Component, ComponentRef, OnDestroy, OnInit, ViewChild, ViewContainerRef } from '@angular/core';
import { NgForm } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { PoComponentInjectorService, PoModalAction, PoModalComponent } from '@po-ui/ng-components';
import { browserLanguage, getParentRef, isExternalLink, isTypeof, poLocaleDefault } from '../../utils/util';
import { PoModalPasswordRecoveryComponent } from '../po-modal-password-recovery/po-modal-password-recovery.component';
import { PoModalPasswordRecoveryType } from '../po-modal-password-recovery/enums/po-modal-password-recovery-type.enum';
import { PoPageChangePasswordBaseComponent } from './po-page-change-password-base.component';
import { poPageChangePasswordLiterals } from './literals/i18n/po-page-change-password-literals';
import { PoPageChangePasswordService } from './po-page-change-password.service';
/**
 * @docsExtends PoPageChangePasswordBaseComponent
 *
 * @example
 *
 * <example name="po-page-change-password-basic" title="PO Page Change Password Basic">
 *  <file name="sample-po-page-change-password-basic/sample-po-page-change-password-basic.component.html"> </file>
 *  <file name="sample-po-page-change-password-basic/sample-po-page-change-password-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-labs" title="PO Page Change Password Labs">
 *  <file name="sample-po-page-change-password-labs/sample-po-page-change-password-labs.component.html"> </file>
 *  <file name="sample-po-page-change-password-labs/sample-po-page-change-password-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-modify" title="PO Page Change Password Modify">
 *  <file name="sample-po-page-change-password-modify/sample-po-page-change-password-modify.component.html"> </file>
 *  <file name="sample-po-page-change-password-modify/sample-po-page-change-password-modify.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-create" title="PO Page Change Password Create">
 *  <file name="sample-po-page-change-password-create/sample-po-page-change-password-create.component.html"> </file>
 *  <file name="sample-po-page-change-password-create/sample-po-page-change-password-create.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-change-password-request" title="PO Page Change Password Request">
 *  <file name="sample-po-page-change-password-request/sample-po-page-change-password-request.component.html"> </file>
 *  <file name="sample-po-page-change-password-request/sample-po-page-change-password-request.component.ts"> </file>
 * </example>
 */
var PoPageChangePasswordComponent = /** @class */ (function (_super) {
    __extends(PoPageChangePasswordComponent, _super);
    function PoPageChangePasswordComponent(activatedRoute, route, router, service, poComponentInjector, viewRef) {
        var _this = _super.call(this) || this;
        _this.activatedRoute = activatedRoute;
        _this.route = route;
        _this.router = router;
        _this.service = service;
        _this.poComponentInjector = poComponentInjector;
        _this.literals = __assign(__assign({}, poPageChangePasswordLiterals[poLocaleDefault]), poPageChangePasswordLiterals[browserLanguage()]);
        _this.componentRef = null;
        _this.modalAction = {
            action: _this.navigateTo.bind(_this, _this.urlHome),
            label: _this.literals.enterSystemButton
        };
        _this.parentRef = getParentRef(viewRef);
        return _this;
    }
    PoPageChangePasswordComponent.prototype.ngAfterViewInit = function () {
        if (this.urlNewPassword) {
            this.subscribeToTokenParameter();
        }
    };
    PoPageChangePasswordComponent.prototype.ngOnDestroy = function () {
        if (this.newPasswordSubscription) {
            this.newPasswordSubscription.unsubscribe();
        }
    };
    PoPageChangePasswordComponent.prototype.ngOnInit = function () {
        this.checkingForRouteMetadata(this.activatedRoute.snapshot.data);
    };
    PoPageChangePasswordComponent.prototype.navigateTo = function (url) {
        isExternalLink(url) ? window.open(url) : this.router.navigate([url || '/']);
    };
    PoPageChangePasswordComponent.prototype.onForgotPasswordClick = function (recovery) {
        if (isTypeof(recovery, 'function')) {
            recovery();
        }
        else {
            this.createModalPasswordRecoveryComponent(recovery);
        }
    };
    PoPageChangePasswordComponent.prototype.onLoginSubmit = function () {
        var form = this.getLoginForm();
        if (this.urlNewPassword) {
            this.postUrlNewPassword(form);
        }
        else if (this.submit && this.submit.observers.length) {
            this.emitSubmit(form);
        }
    };
    /**
     * Abre uma modal de confirmação com texto, imagem e botão que redireciona para o link definido na propriedade `p-url-home`
     */
    PoPageChangePasswordComponent.prototype.openConfirmation = function () {
        this.modal.open();
    };
    PoPageChangePasswordComponent.prototype.validatePassword = function () {
        var _this = this;
        var controls = this.passwordForm.form.controls;
        var controlConfirmPassword = controls['confirmPassword'];
        var controlNewPassword = controls['newPassword'];
        if (!this.newPassword) {
            this.setFormErrors({ 'required': true }, [controlNewPassword]);
        }
        else if (!this.confirmPassword) {
            this.setFormErrors({ 'required': true }, [controlConfirmPassword]);
        }
        else if (this.newPassword && this.confirmPassword && this.newPassword !== this.confirmPassword) {
            this.setFormErrors({ 'equalPassword': true }, [controlNewPassword, controlConfirmPassword]);
        }
        else {
            this.setFormErrors(null, [controlConfirmPassword, controlNewPassword]);
        }
        if (this.requirements.length &&
            this.requirements.find(function (requirement) { return _this.validateRequirement(requirement) === false; })) {
            this.setFormErrors({ 'requirement': true }, [controlNewPassword]);
        }
    };
    PoPageChangePasswordComponent.prototype.validateRequirement = function (requirement) {
        return typeof requirement.status === 'function'
            ? requirement.status.call(this.parentRef, this.newPassword)
            : requirement.status;
    };
    PoPageChangePasswordComponent.prototype.checkingForMetadataProperty = function (object, property) {
        if (Object.prototype.hasOwnProperty.call(object, property)) {
            return object[property];
        }
    };
    PoPageChangePasswordComponent.prototype.checkingForRouteMetadata = function (data) {
        if (Object.keys(data).length !== 0) {
            this.urlNewPassword = this.checkingForMetadataProperty(data, 'serviceApi') || this.urlNewPassword;
            this.recovery = this.checkingForMetadataProperty(data, 'recovery') || this.recovery;
            this.hideCurrentPassword =
                this.checkingForMetadataProperty(data, 'hideCurrentPassword') || this.hideCurrentPassword;
        }
    };
    PoPageChangePasswordComponent.prototype.createModalPasswordRecoveryComponent = function (recovery) {
        var _this = this;
        if (this.componentRef) {
            this.poComponentInjector.destroyComponentInApplication(this.componentRef);
        }
        this.componentRef = this.poComponentInjector.createComponentInApplication(PoModalPasswordRecoveryComponent);
        this.componentRef.instance.recovery = recovery.url;
        this.componentRef.instance.contactEmail = recovery.contactMail;
        this.componentRef.instance.phoneMask = recovery.phoneMask;
        this.componentRef.instance.type = recovery.type || PoModalPasswordRecoveryType.Email;
        this.componentRef.changeDetectorRef.detectChanges();
        setTimeout(function () {
            _this.componentRef.instance.open();
        });
    };
    PoPageChangePasswordComponent.prototype.emitSubmit = function (form) {
        this.submit.emit(form);
    };
    PoPageChangePasswordComponent.prototype.getLoginForm = function () {
        return {
            currentPassword: this.currentPassword,
            newPassword: this.newPassword
        };
    };
    PoPageChangePasswordComponent.prototype.postUrlNewPassword = function (form) {
        var _this = this;
        form['token'] = this.token;
        this.service.post(this.urlNewPassword, form).subscribe(function (response) {
            if (response.status === 204) {
                _this.openConfirmation();
            }
        });
    };
    PoPageChangePasswordComponent.prototype.setFormErrors = function (error, controls) {
        controls.forEach(function (control) {
            control.setErrors(error);
        });
    };
    PoPageChangePasswordComponent.prototype.subscribeToTokenParameter = function () {
        var _this = this;
        this.route.queryParams.subscribe(function (params) {
            var token = params['token'];
            if (token) {
                _this.token = token;
            }
        });
    };
    PoPageChangePasswordComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: ActivatedRoute },
        { type: Router },
        { type: PoPageChangePasswordService },
        { type: PoComponentInjectorService },
        { type: ViewContainerRef }
    ]; };
    __decorate([
        ViewChild(PoModalComponent, { static: true }),
        __metadata("design:type", PoModalComponent)
    ], PoPageChangePasswordComponent.prototype, "modal", void 0);
    __decorate([
        ViewChild('pageChangePassword', { read: ViewContainerRef, static: true }),
        __metadata("design:type", ViewContainerRef)
    ], PoPageChangePasswordComponent.prototype, "pageChangePassword", void 0);
    __decorate([
        ViewChild('passwordForm', { read: NgForm, static: true }),
        __metadata("design:type", NgForm)
    ], PoPageChangePasswordComponent.prototype, "passwordForm", void 0);
    PoPageChangePasswordComponent = __decorate([
        Component({
            selector: 'po-page-change-password',
            template: "<div class=\"po-page-change-password-wrapper\" #pageChangePassword>\n  <div class=\"po-page-change-password-container\">\n    <po-page-background class=\"po-page-change-password\" [p-logo]=\"logo\" [p-secondary-logo]=\"secondaryLogo\">\n      <div class=\"po-page-blocked-user-header po-row\">\n        <div class=\"po-md-12\">{{ literals.createNewPassword }}</div>\n        <div class=\"po-md-12 po-text-center po-font-text\">\n          {{ literals.createNewPasswordPhrase }}\n        </div>\n      </div>\n\n      <form #passwordForm=\"ngForm\">\n        <div class=\"po-row po-mt-3 po-mt-sm-1\">\n          <div class=\"po-md-6\" [ngClass]=\"{ 'po-offset-md-3 po-offset-lg-3 po-offset-xl-3': !showRequirements }\">\n            <div *ngIf=\"!hideCurrentPassword\">\n              <po-password\n                name=\"currentPassword\"\n                [(ngModel)]=\"currentPassword\"\n                p-required\n                [p-label]=\"literals.currentPassword\"\n              >\n              </po-password>\n            </div>\n            <div *ngIf=\"!hideCurrentPassword && recovery\" class=\"po-text-center po-mb-sm-1 po-mb-2\">\n              <a\n                *ngIf=\"recoveryUrlType === 'internalLink'\"\n                class=\"po-font-text-bold po-page-change-password-forgot-password\"\n                [routerLink]=\"recovery\"\n              >\n                {{ literals.forgotPassword }}\n              </a>\n              <a\n                *ngIf=\"recoveryUrlType === 'externalLink'\"\n                class=\"po-font-text-bold po-page-change-password-forgot-password\"\n                [href]=\"recovery\"\n              >\n                {{ literals.forgotPassword }}\n              </a>\n              <a\n                *ngIf=\"!recoveryUrlType\"\n                class=\"po-font-text-bold po-page-change-password-forgot-password po-clickable\"\n                (click)=\"onForgotPasswordClick(recovery)\"\n              >\n                {{ literals.forgotPassword }}\n              </a>\n            </div>\n            <div>\n              <po-password\n                name=\"newPassword\"\n                [(ngModel)]=\"newPassword\"\n                p-required\n                [p-label]=\"literals.newPassword\"\n                (p-change-model)=\"validatePassword()\"\n              >\n              </po-password>\n            </div>\n            <div>\n              <po-password\n                name=\"confirmPassword\"\n                [(ngModel)]=\"confirmPassword\"\n                p-required\n                [p-label]=\"literals.confirmPassword\"\n                (p-change-model)=\"validatePassword()\"\n              >\n              </po-password>\n            </div>\n          </div>\n          <div *ngIf=\"showRequirements\" class=\"po-md-6 po-page-change-password-required-container\">\n            <po-container>\n              <ul class=\"po-page-change-password-required-list\">\n                <li class=\"po-page-change-password-required-item\">\n                  <p class=\"po-mb-1 po-font-text-bold po-page-change-password-required-title\">\n                    {{ literals?.requirements }}\n                  </p>\n                </li>\n                <li *ngFor=\"let requirement of requirements\" class=\"po-page-change-password-required-item\">\n                  <span\n                    class=\"po-icon po-page-change-password-required-icon\"\n                    [ngClass]=\"\n                      validateRequirement(requirement)\n                        ? 'po-icon-ok po-page-change-password-required-ok'\n                        : 'po-icon-minus po-page-change-password-required-minus'\n                    \"\n                  >\n                  </span>\n                  <p class=\"po-page-change-password-required-text po-font-text\">\n                    {{ requirement.requirement }}\n                  </p>\n                </li>\n              </ul>\n            </po-container>\n          </div>\n        </div>\n      </form>\n\n      <po-divider class=\"po-sm-12\"></po-divider>\n\n      <div class=\"po-page-change-password-tips po-row\">\n        <div class=\"po-md-2 po-mr-3\">\n          <img\n            class=\"po-page-change-password-lock-image po-mb-3\"\n            alt=\"Blocked user image\"\n            src=\"./assets/images/big-lock.svg\"\n          />\n        </div>\n        <div class=\"po-md-9 po-page-change-password-tips-text\">\n          <div class=\"po-font-text-bold po-pb-1\">{{ literals.safetyTips }}</div>\n          <div class=\"po-font-text-small\">\n            <div class=\"po-row po-pb-1\">{{ literals.safetyTipsPhrase }}</div>\n            <div class=\"po-row po-pb-1\">\u2022 {{ literals.safetyTipsFirst }}</div>\n            <div class=\"po-row po-pb-1\">\u2022 {{ literals.safetyTipsSecond }}</div>\n            <div class=\"po-row po-pb-1\">\u2022 {{ literals.safetyTipsThird }}</div>\n          </div>\n        </div>\n      </div>\n\n      <div class=\"po-page-change-password-buttons po-row\">\n        <po-divider class=\"po-sm-12 po-mobile-only\"></po-divider>\n\n        <po-button\n          *ngIf=\"!hideCurrentPassword\"\n          class=\"po-sm-6 po-mb-sm-1\"\n          [p-label]=\"literals.backButton\"\n          (p-click)=\"navigateTo(urlBack)\"\n        >\n        </po-button>\n\n        <po-button\n          [ngClass]=\"hideCurrentPassword ? 'po-md-12' : 'po-sm-6'\"\n          p-type=\"primary\"\n          [p-disabled]=\"passwordForm.invalid\"\n          [p-label]=\"literals.saveButton\"\n          (p-click)=\"onLoginSubmit()\"\n        >\n        </po-button>\n      </div>\n    </po-page-background>\n  </div>\n</div>\n\n<po-modal\n  [p-primary-action]=\"modalAction\"\n  p-hide-close\n  p-size=\"auto\"\n  [p-title]=\"hideCurrentPassword ? literals.passwordSuccessfullyCreated : literals.passwordSuccessfullyUpdated\"\n>\n  <div class=\"po-text-center\">\n    <img\n      alt=\"Blocked user image\"\n      class=\"po-page-change-password-lock-image po-mb-3\"\n      src=\"./assets/images/success.svg\"\n    />\n  </div>\n</po-modal>\n"
        }),
        __metadata("design:paramtypes", [ActivatedRoute,
            ActivatedRoute,
            Router,
            PoPageChangePasswordService,
            PoComponentInjectorService,
            ViewContainerRef])
    ], PoPageChangePasswordComponent);
    return PoPageChangePasswordComponent;
}(PoPageChangePasswordBaseComponent));
export { PoPageChangePasswordComponent };
