// dts-backoffice-util/esm2015/lib/utils/date.util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
Versão: 1.000
Data Criação: 18/04/2019
*/
export class DateUtil {
    constructor() { }
    /* Transforma data para o padrão YYYY-MM-DD */
    /**
     * @param {?} date
     * @return {?}
     */
    static dateToQueryParam(date) {
        /** @type {?} */
        const iDay = date.getDate();
        /** @type {?} */
        const iMonth = date.getMonth() + 1;
        /** @type {?} */
        const iYear = date.getFullYear();
        return `${iYear}-${this.pad(iMonth)}-${this.pad(iDay)}`;
    }
    /* Transforma data no padrão YYYY-MM-DD para DATE */
    /**
     * @param {?} param
     * @return {?}
     */
    static queryParamToDate(param) {
        /** @type {?} */
        const iDay = +param.split('-')[2];
        /** @type {?} */
        const iMonth = +param.split('-')[1];
        /** @type {?} */
        const iYear = +param.split('-')[0];
        return new Date(iYear, iMonth - 1, iDay);
    }
    /* Valida se foi informado uma data válida */
    /**
     * @param {?} date
     * @return {?}
     */
    static isValidDate(date) {
        return date instanceof Date && !isNaN(date.getTime());
    }
    /* Ajusta a data retornando o padrão DATE */
    /**
     * @param {?} param
     * @return {?}
     */
    static ajustDate(param) {
        if (param instanceof Date) {
            return param;
        }
        if (!param) {
            return param;
        }
        return this.queryParamToDate(param.split('T')[0]);
    }
    /* Ajusta a data para o padrão DATE - utilizada nos construtores dos modelos */
    /**
     * @param {?=} values
     * @param {?=} fieldName
     * @return {?}
     */
    static ajustDateToModel(values = {}, fieldName) {
        if (values.hasOwnProperty(fieldName) && (values[fieldName])) {
            if (values[fieldName].indexOf('T') >= 0) {
                return new Date(values[fieldName]);
            }
            /** @type {?} */
            const iDay = +values[fieldName].split('-')[2];
            /** @type {?} */
            const iMonth = +values[fieldName].split('-')[1];
            /** @type {?} */
            const iYear = +values[fieldName].split('-')[0];
            return new Date(iYear, iMonth - 1, iDay);
        }
        return null;
    }
    /* Adicona zero a esquerda do número */
    /**
     * @param {?} number
     * @return {?}
     */
    static pad(number) {
        if (number < 10) {
            return '0' + number;
        }
        return number;
    }
}
