// dts-backoffice-util/esm2015/lib/utils/filter-range.util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
Versão: 1.000
Data Criação: 06/08/2018
*/
export class FilterRangeUtil {
    constructor() {
    }
    /**
     * @param {?} iniInitial
     * @param {?} iniFinal
     * @return {?}
     */
    static makeFilterRangeCharacter(iniInitial, iniFinal) {
        return {
            iniInitial,
            iniFinal,
            valInitial: iniInitial,
            valFinal: iniFinal
        };
    }
    /**
     * @param {?} iniInitial
     * @param {?} iniFinal
     * @return {?}
     */
    static makeFilterRangeNumber(iniInitial, iniFinal) {
        return {
            iniInitial,
            iniFinal,
            valInitial: iniInitial,
            valFinal: iniFinal
        };
    }
    /**
     * @param {?} iniInitial
     * @param {?} iniFinal
     * @return {?}
     */
    static makeFilterRangeDate(iniInitial, iniFinal) {
        return {
            iniInitial,
            iniFinal,
            valInitial: iniInitial,
            valFinal: iniFinal
        };
    }
}
