// @po-ui/ng-templates/esm5/lib/components/po-page-job-scheduler/po-page-job-scheduler-parameters/po-page-job-scheduler-parameters.component.js
import { __decorate, __metadata } from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { NgForm } from '@angular/forms';
var PoPageJobSchedulerParametersComponent = /** @class */ (function () {
    function PoPageJobSchedulerParametersComponent() {
        this.literals = {};
        this.parameters = [];
        this.valueChange = new EventEmitter();
    }
    PoPageJobSchedulerParametersComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (this.form) {
            setTimeout(function () {
                _this.form.valueChanges.subscribe(function (value) {
                    _this.valueChange.emit(value);
                });
            });
        }
    };
    __decorate([
        ViewChild('parametersForm'),
        __metadata("design:type", NgForm)
    ], PoPageJobSchedulerParametersComponent.prototype, "form", void 0);
    __decorate([
        Input('p-literals'),
        __metadata("design:type", Object)
    ], PoPageJobSchedulerParametersComponent.prototype, "literals", void 0);
    __decorate([
        Input('p-parameters'),
        __metadata("design:type", Array)
    ], PoPageJobSchedulerParametersComponent.prototype, "parameters", void 0);
    __decorate([
        Input('p-value'),
        __metadata("design:type", Object)
    ], PoPageJobSchedulerParametersComponent.prototype, "value", void 0);
    __decorate([
        Output('p-valueChange'),
        __metadata("design:type", EventEmitter)
    ], PoPageJobSchedulerParametersComponent.prototype, "valueChange", void 0);
    PoPageJobSchedulerParametersComponent = __decorate([
        Component({
            selector: 'po-page-job-scheduler-parameters',
            template: "<ng-container *ngIf=\"parameters && parameters.length; then formFieldsTemplate; else parametersNotFoundTemplate\">\n</ng-container>\n\n<ng-template #parametersNotFoundTemplate>\n  <div class=\"po-text-center\">\n    <span class=\"po-icon po-icon-info\"></span>\n    <span class=\"po-font-text-large\">\n      {{ literals.parametersNotFound }}\n    </span>\n  </div>\n</ng-template>\n\n<ng-template #formFieldsTemplate>\n  <form #parametersForm=\"ngForm\">\n    <po-dynamic-form p-group-form [p-fields]=\"parameters\" [p-value]=\"value\"> </po-dynamic-form>\n  </form>\n</ng-template>\n"
        })
    ], PoPageJobSchedulerParametersComponent);
    return PoPageJobSchedulerParametersComponent;
}());
export { PoPageJobSchedulerParametersComponent };
