// @po-ui/ng-templates/esm5/lib/components/po-page-job-scheduler/po-page-job-scheduler-summary/po-page-job-scheduler-summary.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input } from '@angular/core';
import { DatePipe } from '@angular/common';
import { PoInfoOrientation } from '@po-ui/ng-components';
var PoPageJobSchedulerSummaryComponent = /** @class */ (function () {
    function PoPageJobSchedulerSummaryComponent(datePipe) {
        this.datePipe = datePipe;
        this.executionValue = '';
        this.firstExecutionValue = '';
        this.infoOrientation = PoInfoOrientation.Horizontal;
        this.periodicityValue = '';
        this.recurrentValue = '';
        this.literals = {};
        this.parameters = [];
        this.value = {};
    }
    PoPageJobSchedulerSummaryComponent.prototype.ngOnInit = function () {
        var _a = this.value, periodicity = _a.periodicity, hour = _a.hour, dayOfMonth = _a.dayOfMonth, daysOfWeek = _a.daysOfWeek, recurrent = _a.recurrent, firstExecution = _a.firstExecution, firstExecutionHour = _a.firstExecutionHour;
        this.periodicityValue = this.getPeriodicityLabel(periodicity);
        this.executionValue = this.getExecutionValue(periodicity, hour, daysOfWeek, dayOfMonth);
        this.firstExecutionValue = this.getFirstExecutionLabel(firstExecution, firstExecutionHour);
        this.recurrentValue = this.getRecurrentValue(recurrent);
    };
    PoPageJobSchedulerSummaryComponent.prototype.getExecutionValue = function (periodicity, hour, daysOfWeek, dayOfMonth) {
        switch (periodicity) {
            case 'daily':
                return this.getHourLabel(hour);
            case 'monthly':
                return this.getMonthlyLabelExecution(dayOfMonth, hour);
            case 'weekly':
                return this.getWeeklyLabelExecution(daysOfWeek, hour);
            default:
                return this.literals.notReported;
        }
    };
    PoPageJobSchedulerSummaryComponent.prototype.getFirstExecutionLabel = function (firstExecution, firstExecutionHour) {
        if (firstExecution) {
            var date = this.datePipe.transform(firstExecution, 'dd/MM/yyyy', '-0200');
            return date + " " + this.getHourLabel(firstExecutionHour);
        }
        else {
            return this.literals.notReported;
        }
    };
    PoPageJobSchedulerSummaryComponent.prototype.getHourLabel = function (hour) {
        return this.literals.at + " " + (hour || '00:00') + "h";
    };
    PoPageJobSchedulerSummaryComponent.prototype.getMonthlyLabelExecution = function (dayOfMonth, hour) {
        var hourLabel = this.getHourLabel(hour);
        return this.literals.day + " " + dayOfMonth + " " + hourLabel;
    };
    PoPageJobSchedulerSummaryComponent.prototype.getPeriodicityLabel = function (periodicity) {
        switch (periodicity) {
            case 'daily':
                return this.literals.daily;
            case 'monthly':
                return this.literals.monthly;
            case 'weekly':
                return this.literals.weekly;
            default:
                return this.literals.single;
        }
    };
    PoPageJobSchedulerSummaryComponent.prototype.getRecurrentValue = function (recurrent) {
        return recurrent ? this.literals.yes : this.literals.no;
    };
    PoPageJobSchedulerSummaryComponent.prototype.getSorterWeekDays = function () {
        return {
            'sunday': 0,
            'monday': 1,
            'tuesday': 2,
            'wednesday': 3,
            'thursday': 4,
            'friday': 5,
            'saturday': 6
        };
    };
    PoPageJobSchedulerSummaryComponent.prototype.getTranslateWeekDay = function (day) {
        var days = {
            Sunday: this.literals.sunday,
            Monday: this.literals.monday,
            Tuesday: this.literals.tuesday,
            Wednesday: this.literals.wednesday,
            Thursday: this.literals.thursday,
            Friday: this.literals.friday,
            Saturday: this.literals.saturday
        };
        return days[day] || '';
    };
    PoPageJobSchedulerSummaryComponent.prototype.getWeekDaysLabel = function (days) {
        var _this = this;
        if (days === void 0) { days = []; }
        var weekDaysSorted = this.sortWeekDays(days);
        return weekDaysSorted.map(function (day) { return _this.getTranslateWeekDay(day); }).join(', ');
    };
    PoPageJobSchedulerSummaryComponent.prototype.getWeeklyLabelExecution = function (daysOfWeek, hour) {
        if (daysOfWeek && Array.isArray(daysOfWeek)) {
            return this.getWeekDaysLabel(daysOfWeek) + " " + this.getHourLabel(hour);
        }
        else {
            return this.literals.notReported;
        }
    };
    PoPageJobSchedulerSummaryComponent.prototype.sortWeekDays = function (days) {
        if (days === void 0) { days = []; }
        var sorterWeekDays = this.getSorterWeekDays();
        return days.sort(function (a, b) {
            var currDay = a.toLowerCase();
            var nextDay = b.toLowerCase();
            return sorterWeekDays[currDay] > sorterWeekDays[nextDay] ? 1 : -1;
        });
    };
    PoPageJobSchedulerSummaryComponent.ctorParameters = function () { return [
        { type: DatePipe }
    ]; };
    __decorate([
        Input('p-literals'),
        __metadata("design:type", Object)
    ], PoPageJobSchedulerSummaryComponent.prototype, "literals", void 0);
    __decorate([
        Input('p-parameters'),
        __metadata("design:type", Array)
    ], PoPageJobSchedulerSummaryComponent.prototype, "parameters", void 0);
    __decorate([
        Input('p-value'),
        __metadata("design:type", Object)
    ], PoPageJobSchedulerSummaryComponent.prototype, "value", void 0);
    PoPageJobSchedulerSummaryComponent = __decorate([
        Component({
            selector: 'po-page-job-scheduler-summary',
            template: "<div class=\"po-md-12\">\n  <po-info [p-label]=\"literals.process\" [p-orientation]=\"infoOrientation\" [p-value]=\"value.processID\"> </po-info>\n\n  <po-info [p-label]=\"literals.periodicity\" [p-orientation]=\"infoOrientation\" [p-value]=\"periodicityValue\"> </po-info>\n\n  <po-info\n    *ngIf=\"value.periodicity !== 'single'\"\n    [p-label]=\"literals.execution\"\n    [p-orientation]=\"infoOrientation\"\n    [p-value]=\"executionValue\"\n  >\n  </po-info>\n\n  <po-info [p-label]=\"literals.recurrent\" [p-orientation]=\"infoOrientation\" [p-value]=\"recurrentValue\"> </po-info>\n\n  <po-info [p-label]=\"literals.firstExecution\" [p-orientation]=\"infoOrientation\" [p-value]=\"firstExecutionValue\">\n  </po-info>\n</div>\n\n<po-widget *ngIf=\"parameters && parameters.length\" class=\"po-pt-1 po-pb-1 po-md-12\" [p-title]=\"literals.parameters\">\n  <po-dynamic-view [p-fields]=\"parameters\" [p-value]=\"value.executionParameter\"> </po-dynamic-view>\n</po-widget>\n"
        }),
        __metadata("design:paramtypes", [DatePipe])
    ], PoPageJobSchedulerSummaryComponent);
    return PoPageJobSchedulerSummaryComponent;
}());
export { PoPageJobSchedulerSummaryComponent };
