// dts-backoffice-util/esm5/lib/services/breadcrumb-control.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
Versão: 1.003
Data Criação: 06/08/2018
*/
import { Injectable } from '@angular/core';
var BreadcrumbControlService = /** @class */ (function () {
    function BreadcrumbControlService() {
    }
    /**
     * @return {?}
     */
    BreadcrumbControlService.prototype.newBreadcrumb = /**
     * @return {?}
     */
    function () {
        this.breadcrumb = { items: [] };
    };
    /**
     * @param {?} literal
     * @param {?} activatedRoute
     * @return {?}
     */
    BreadcrumbControlService.prototype.addBreadcrumb = /**
     * @param {?} literal
     * @param {?} activatedRoute
     * @return {?}
     */
    function (literal, activatedRoute) {
        this.addBreadcrumbURL(literal, decodeURIComponent(activatedRoute.snapshot['_routerState'].url));
    };
    /**
     * @param {?} literal
     * @param {?} url
     * @return {?}
     */
    BreadcrumbControlService.prototype.addBreadcrumbURL = /**
     * @param {?} literal
     * @param {?} url
     * @return {?}
     */
    function (literal, url) {
        if (!literal || literal === '') {
            return;
        }
        if (!this.breadcrumb) {
            this.newBreadcrumb();
        }
        /** @type {?} */
        var breadcrumbAux = { items: [] };
        /** @type {?} */
        var idx = 0;
        for (idx = 0; idx < this.breadcrumb.items.length; idx++) {
            if (this.breadcrumb.items[idx].label !== literal) {
                breadcrumbAux.items.push(this.breadcrumb.items[idx]);
            }
            else {
                break;
            }
        }
        breadcrumbAux.items.push({ label: literal, link: url });
        this.breadcrumb = breadcrumbAux;
    };
    /**
     * @param {?} literal
     * @param {?} valueOld
     * @param {?} valueNew
     * @return {?}
     */
    BreadcrumbControlService.prototype.updBreadcrumbURL = /**
     * @param {?} literal
     * @param {?} valueOld
     * @param {?} valueNew
     * @return {?}
     */
    function (literal, valueOld, valueNew) {
        /** @type {?} */
        var currentRouterURL = this.getCurrentRouter();
        if (!currentRouterURL) {
            return;
        }
        if (!valueOld) {
            return;
        }
        if (!valueNew) {
            return;
        }
        currentRouterURL = currentRouterURL.replace(valueOld, valueNew);
        this.addBreadcrumbURL(literal, currentRouterURL);
    };
    /**
     * @return {?}
     */
    BreadcrumbControlService.prototype.getBreadcrumb = /**
     * @return {?}
     */
    function () {
        if (!this.breadcrumb) {
            this.newBreadcrumb();
        }
        return this.breadcrumb;
    };
    /**
     * @return {?}
     */
    BreadcrumbControlService.prototype.getCurrentRouter = /**
     * @return {?}
     */
    function () {
        if (!this.breadcrumb) {
            this.newBreadcrumb();
        }
        if (this.breadcrumb.items.length <= 0) {
            return '';
        }
        return this.breadcrumb.items[this.breadcrumb.items.length - 1].link;
    };
    /**
     * @return {?}
     */
    BreadcrumbControlService.prototype.getPrevRouter = /**
     * @return {?}
     */
    function () {
        if (!this.breadcrumb) {
            this.newBreadcrumb();
        }
        if (this.breadcrumb.items.length <= 1) {
            return '';
        }
        return this.breadcrumb.items[this.breadcrumb.items.length - 2].link;
    };
    /**
     * @return {?}
     */
    BreadcrumbControlService.prototype.hasPreviousRouter = /**
     * @return {?}
     */
    function () {
        return this.breadcrumb.items.length > 1;
    };
    BreadcrumbControlService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    BreadcrumbControlService.ctorParameters = function () { return []; };
    return BreadcrumbControlService;
}());
export { BreadcrumbControlService };
if (false) {
    /** @type {?} */
    BreadcrumbControlService.prototype.breadcrumb;
}
