// dts-backoffice-util/esm5/lib/services/cache-params.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var CacheParamsService = /** @class */ (function () {
    function CacheParamsService() {
        this.values = {};
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    CacheParamsService.prototype.setValue = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        this.values[key] = value;
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CacheParamsService.prototype.getValue = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this.values[key];
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CacheParamsService.prototype.removeValue = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        delete this.values[key];
    };
    CacheParamsService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    CacheParamsService.ctorParameters = function () { return []; };
    /** @nocollapse */ CacheParamsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CacheParamsService_Factory() { return new CacheParamsService(); }, token: CacheParamsService, providedIn: "root" });
    return CacheParamsService;
}());
export { CacheParamsService };
if (false) {
    /** @type {?} */
    CacheParamsService.prototype.values;
}
