// dts-backoffice-util/esm5/lib/services/menu-datasul.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var MenuDatasulService = /** @class */ (function () {
    function MenuDatasulService() {
    }
    /**
     * @param {?} program
     * @return {?}
     */
    MenuDatasulService.prototype.callProgress = /**
     * @param {?} program
     * @return {?}
     */
    function (program) {
        /**
         * var program = {};
         * program['prg'] = "cd0210";
         * program['params'] = '01|123456'
         */
        parent.postMessage({ program: program }, '*');
    };
    /**
     * @param {?} notification
     * @return {?}
     */
    MenuDatasulService.prototype.sendNotification = /**
     * @param {?} notification
     * @return {?}
     */
    function (notification) {
        /**
         * var notification = {};
         * notification['type'] = 'success';
         * notification['title'] = 'Notification Title'
         * notification['detail'] = 'Notification Detail'
         */
        parent.postMessage({ notification: notification }, '*');
    };
    /**
     * @param {?} programName
     * @param {?} params
     * @param {?=} parent
     * @return {?}
     */
    MenuDatasulService.prototype.openPath = /**
     * @param {?} programName
     * @param {?} params
     * @param {?=} parent
     * @return {?}
     */
    function (programName, params, parent) {
        if (parent === void 0) { parent = true; }
        /** @type {?} */
        var datasulPath = window.location.href.indexOf('totvs-menu') > 0 ? 'totvs-menu' : 'menu-html';
        /** @type {?} */
        var baseUrl = "/" + datasulPath + "/#/" + datasulPath + "/program-html/" + programName + "/#";
        if (parent) {
            window.open(baseUrl + "/" + params, '_parent');
        }
        else {
            window.open(baseUrl + "/" + params);
        }
    };
    MenuDatasulService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MenuDatasulService.ctorParameters = function () { return []; };
    return MenuDatasulService;
}());
export { MenuDatasulService };
