// dts-backoffice-util/esm5/lib/services/profile.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators/map';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
/**
 * @record
 */
export function IProfile() { }
if (false) {
    /** @type {?} */
    IProfile.prototype.pageId;
    /** @type {?} */
    IProfile.prototype.dataCode;
    /** @type {?} */
    IProfile.prototype.userCode;
    /** @type {?|undefined} */
    IProfile.prototype.dataValue;
}
/**
 * @record
 */
export function IProfileResponse() { }
if (false) {
    /** @type {?} */
    IProfileResponse.prototype.message;
    /** @type {?} */
    IProfileResponse.prototype.length;
    /** @type {?} */
    IProfileResponse.prototype.data;
}
/** @enum {number} */
var EDataType = {
    STRING: 1,
    JSON: 2,
};
EDataType[EDataType.STRING] = 'STRING';
EDataType[EDataType.JSON] = 'JSON';
var ProfileService = /** @class */ (function () {
    function ProfileService(http) {
        this.http = http;
        this.apiURL = '/dts/datasul-rest/resources/api/btb/btapi930za';
    }
    /**
     * @param {?} profile
     * @return {?}
     */
    ProfileService.prototype.setProfile = /**
     * @param {?} profile
     * @return {?}
     */
    function (profile) {
        /** @type {?} */
        var headers = { 'X-PO-Screen-Lock': 'true' };
        var pageId = profile.pageId, userCode = profile.userCode;
        return this.http
            .post(this.apiURL + "/setProfile?userCode=" + userCode + "&pageId=" + pageId, profile, { headers: headers });
    };
    /**
     * @param {?} profile
     * @param {?=} showLoading
     * @return {?}
     */
    ProfileService.prototype.getProfileAsString = /**
     * @param {?} profile
     * @param {?=} showLoading
     * @return {?}
     */
    function (profile, showLoading) {
        return this.getProfile(profile, EDataType.STRING, showLoading ? 'true' : 'false');
    };
    /**
     * @param {?} profile
     * @param {?=} showLoading
     * @return {?}
     */
    ProfileService.prototype.getProfileAsJSON = /**
     * @param {?} profile
     * @param {?=} showLoading
     * @return {?}
     */
    function (profile, showLoading) {
        return this.getProfile(profile, EDataType.JSON, showLoading ? 'true' : 'false');
    };
    /**
     * @private
     * @param {?} profile
     * @param {?} dataType
     * @param {?=} showLoading
     * @return {?}
     */
    ProfileService.prototype.getProfile = /**
     * @private
     * @param {?} profile
     * @param {?} dataType
     * @param {?=} showLoading
     * @return {?}
     */
    function (profile, dataType, showLoading) {
        /** @type {?} */
        var headers = { 'X-PO-Screen-Lock': showLoading };
        var pageId = profile.pageId, dataCode = profile.dataCode, userCode = profile.userCode;
        return this.http
            .get(this.apiURL + "/getProfile?userCode=" + userCode + "&pageId=" + pageId + "&dataCode=" + dataCode, { headers: headers })
            .pipe(map((/**
         * @param {?} preference
         * @return {?}
         */
        function (preference) {
            if (preference.data[0]) {
                if (dataType === EDataType.STRING) {
                    return preference.data[0].dataValue;
                }
                if (dataType === EDataType.JSON) {
                    return JSON.parse(preference.data[0].dataValue);
                }
            }
            else {
                return undefined;
            }
        })));
    };
    ProfileService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ProfileService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    /** @nocollapse */ ProfileService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProfileService_Factory() { return new ProfileService(i0.ɵɵinject(i1.HttpClient)); }, token: ProfileService, providedIn: "root" });
    return ProfileService;
}());
export { ProfileService };
if (false) {
    /** @type {?} */
    ProfileService.prototype.apiURL;
    /** @type {?} */
    ProfileService.prototype.http;
}
