// dts-backoffice-util/esm5/lib/services/report.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators/map';
/**
 * @record
 */
export function IProperty() { }
if (false) {
    /** @type {?} */
    IProperty.prototype.name;
    /** @type {?} */
    IProperty.prototype.value;
}
/**
 * @record
 */
export function IReportServiceParams() { }
if (false) {
    /** @type {?} */
    IReportServiceParams.prototype.reportName;
    /** @type {?} */
    IReportServiceParams.prototype.programName;
    /** @type {?} */
    IReportServiceParams.prototype.properties;
    /** @type {?} */
    IReportServiceParams.prototype.dialect;
    /** @type {?} */
    IReportServiceParams.prototype.downloadName;
    /** @type {?} */
    IReportServiceParams.prototype.download;
    /** @type {?} */
    IReportServiceParams.prototype.format;
}
/** @enum {string} */
var ReportFormats = {
    XLSX: 'xlsx',
    PDF: 'pdf',
    DOCX: 'docx',
    HTML: 'html',
};
export { ReportFormats };
var ReportService = /** @class */ (function () {
    function ReportService(httpClient) {
        this.httpClient = httpClient;
        this.URL = window.location.href.indexOf('totvs-menu') > 0 ? '/totvs-menu/rest/report/run/'
            : '/dts/datasul-report/resources/run/';
    }
    /**
     * @param {?} params
     * @param {?=} showLoading
     * @return {?}
     */
    ReportService.prototype.generate = /**
     * @param {?} params
     * @param {?=} showLoading
     * @return {?}
     */
    function (params, showLoading) {
        var _this = this;
        if (showLoading === void 0) { showLoading = true; }
        /** @type {?} */
        var headers = { 'X-PO-Screen-Lock': showLoading ? 'true' : 'false' };
        /** @type {?} */
        var reportURL = "" + this.URL + params.reportName;
        if (params.properties.length > 0) {
            reportURL += '?';
            params.properties.forEach((/**
             * @param {?} property
             * @param {?} index
             * @return {?}
             */
            function (property, index) {
                if (index > 0) {
                    reportURL += '&';
                }
                reportURL += "c_properties=" + property.name + "&c_values=" + property.value;
            }));
        }
        reportURL += "&dialect=" + params.dialect + "&format=" + params.format;
        reportURL += "&program=" + params.programName + "&resultFileName=" + params.downloadName;
        return this.httpClient.post(reportURL, {}, { headers: headers, responseType: 'blob' })
            .pipe(map((/**
         * @param {?} report
         * @return {?}
         */
        function (report) {
            if (params.download) {
                _this.download(report, params.downloadName + "." + params.format);
            }
            return report;
        })));
    };
    /**
     * @param {?} file
     * @param {?} fileName
     * @return {?}
     */
    ReportService.prototype.download = /**
     * @param {?} file
     * @param {?} fileName
     * @return {?}
     */
    function (file, fileName) {
        /** @type {?} */
        var binaryData = [file];
        /** @type {?} */
        var downloadLink = document.createElement('a');
        /** @type {?} */
        var urlDownload = window.URL.createObjectURL(new Blob(binaryData, { type: file.type }));
        downloadLink.href = urlDownload;
        downloadLink.setAttribute('download', fileName);
        document.body.appendChild(downloadLink);
        downloadLink.click();
        window.URL.revokeObjectURL(urlDownload);
        downloadLink.remove();
    };
    ReportService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ReportService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return ReportService;
}());
export { ReportService };
if (false) {
    /** @type {?} */
    ReportService.prototype.URL;
    /** @type {?} */
    ReportService.prototype.httpClient;
}
