// dts-backoffice-util/esm5/lib/services/session-info.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { of } from 'rxjs';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
var UserLoginService = /** @class */ (function () {
    function UserLoginService(http) {
        this.http = http;
        this.apiUrl = '/totvs-menu/rest/getSessionInfo';
        this.userLogin = undefined;
    }
    /**
     * @return {?}
     */
    UserLoginService.prototype.getUserLogin = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (localStorage.getItem('username')) {
            return of(localStorage.getItem('username'));
        }
        if (this.userLogin) {
            return of(this.userLogin);
        }
        else {
            return this.http.get("" + this.apiUrl)
                .pipe(map((/**
             * @param {?} response
             * @return {?}
             */
            function (response) {
                _this.userLogin = response.userName;
                return response.userName;
            })));
        }
    };
    UserLoginService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    UserLoginService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    /** @nocollapse */ UserLoginService.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserLoginService_Factory() { return new UserLoginService(i0.ɵɵinject(i1.HttpClient)); }, token: UserLoginService, providedIn: "root" });
    return UserLoginService;
}());
export { UserLoginService };
if (false) {
    /** @type {?} */
    UserLoginService.prototype.apiUrl;
    /** @type {?} */
    UserLoginService.prototype.userLogin;
    /** @type {?} */
    UserLoginService.prototype.http;
}
