// dts-backoffice-util/esm2015/lib/components/totvs-schedule-execution/totvs-schedule-execution.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { map } from 'rxjs/internal/operators';
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
export class RpwService {
    // private apiUrl = '/genericsZoom';
    // private urlJobScheduler = '/jobScheduler';
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.apiUrl = '/dts/datasul-rest/resources/prg/cdp/v1/genericsZoom';
        this.urlJobScheduler = '/dts/datasul-rest/resources/prg/framework/v1/jobScheduler';
        this.filterRpw = {
            fields: 'cod_servid_exec,des_servid_exec',
            filter: 'cod_servid_exec',
            filterZoom: 'cod_servid_exec',
            order: 'cod_servid_exec',
            table: 'servid_exec'
        };
    }
    /**
     * @param {?} param
     * @param {?} appId
     * @return {?}
     */
    getFilteredData(param, appId) {
        /** @type {?} */
        const localParams = Object.assign({}, this.filterRpw);
        /** @type {?} */
        let url = `${this.apiUrl}?pageSize=10&page=1&table=${localParams.table}`;
        url += `&fields=${localParams.fields}&order=${localParams.order}`;
        if (param.value) {
            url += `&${localParams.filterZoom}=*${param.value}*`;
        }
        return this.http.get(url, {})
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        (response) => this.convertToArrayComboOption(response.items, 'codServidExec', 'desServidExec'))));
    }
    /**
     * @param {?} value
     * @return {?}
     */
    getObjectByValue(value) {
        return this.http.get(`${this.apiUrl}/${value}`)
            .pipe(map((/**
         * @param {?} item
         * @return {?}
         */
        item => this.convertToThfComboOption(item, 'codServidExec', 'desServidExec'))));
    }
    /**
     * @param {?} parameters
     * @return {?}
     */
    createRpw(parameters) {
        /** @type {?} */
        const params = JSON.parse(JSON.stringify(parameters).replace(/\\\\/g, '*|'));
        return this.http.post(`${this.urlJobScheduler}`, params);
    }
    /* COMBO */
    /**
     * @param {?} items
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    convertToArrayComboOption(items, key, value) {
        if (items && items.length > 0) {
            return items.map((/**
             * @param {?} item
             * @return {?}
             */
            item => this.convertToThfComboOption(item, key, value)));
        }
        return [];
    }
    /**
     * @param {?} item
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    convertToThfComboOption(item, key, value) {
        item = item || {};
        return {
            value: item[key] || undefined,
            label: item[value] || undefined
        };
    }
}
RpwService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RpwService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    RpwService.prototype.apiUrl;
    /**
     * @type {?}
     * @private
     */
    RpwService.prototype.urlJobScheduler;
    /** @type {?} */
    RpwService.prototype.filterRpw;
    /** @type {?} */
    RpwService.prototype.http;
}
