// @po-ui/ng-components/esm5/lib/components/po-field/po-combo/po-combo-filter.service.js
import { __assign, __decorate, __metadata } from "tslib";
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { validateObjectType } from '../../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Serviço padrão utilizado para filtrar os dados do componente po-combo.
 */
var PoComboFilterService = /** @class */ (function () {
    function PoComboFilterService(http) {
        this.http = http;
        this.fieldLabel = 'label';
        this.fieldValue = 'value';
        this.messages = [];
    }
    Object.defineProperty(PoComboFilterService.prototype, "url", {
        get: function () {
            return this._url;
        },
        enumerable: true,
        configurable: true
    });
    PoComboFilterService.prototype.getFilteredData = function (param, filterParams) {
        var _this = this;
        var value = param.value;
        var filterParamsValidated = validateObjectType(filterParams);
        var params = __assign(__assign({}, filterParamsValidated), { filter: value });
        return this.http
            .get("" + this.url, { responseType: 'json', params: params })
            .pipe(map(function (response) { return _this.parseToArrayComboOption(response.items); }));
    };
    PoComboFilterService.prototype.getObjectByValue = function (value, filterParams) {
        var _this = this;
        var filterParamsValidated = validateObjectType(filterParams);
        return this.http
            .get(this.url + "/" + value, { params: filterParamsValidated })
            .pipe(map(function (item) { return _this.parseToComboOption(item); }));
    };
    PoComboFilterService.prototype.configProperties = function (url, fieldLabel, fieldValue) {
        this._url = url;
        this.fieldLabel = fieldLabel;
        this.fieldValue = fieldValue;
    };
    PoComboFilterService.prototype.parseToArrayComboOption = function (items) {
        var _this = this;
        if (items && items.length > 0) {
            var parsedOptions = items.map(function (item) { return _this.parseToComboOption(item); });
            this.displayMessages();
            return parsedOptions;
        }
        return [];
    };
    PoComboFilterService.prototype.parseToComboOption = function (item) {
        if (!(item === null || item === void 0 ? void 0 : item[this.fieldValue])) {
            this.addMessage(item, this.fieldValue);
            return { value: '' };
        }
        var label = item[this.fieldLabel];
        var value = item[this.fieldValue];
        return { label: label, value: value };
    };
    PoComboFilterService.prototype.addMessage = function (item, property) {
        this.messages.push("Cannot read property \"" + property + "\" of " + JSON.stringify(item) + ",\n      see [p-field-value] property at https://po-ui.io/documentation/po-combo");
    };
    PoComboFilterService.prototype.displayMessages = function () {
        if (this.messages.length) {
            this.messages.forEach(function (message) { return console.error(message); });
            this.messages = [];
        }
    };
    PoComboFilterService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    PoComboFilterService = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [HttpClient])
    ], PoComboFilterService);
    return PoComboFilterService;
}());
export { PoComboFilterService };
