// @po-ui/ng-templates/esm2015/lib/components/po-page-dynamic-search/po-advanced-filter/po-advanced-filter.component.js
import { __decorate, __metadata } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { PoDynamicFormComponent, PoLanguageService } from '@po-ui/ng-components';
import { PoAdvancedFilterBaseComponent } from './po-advanced-filter-base.component';
/**
 * @docsPrivate
 *
 * @docsExtends PoAdvancedFilterBaseComponent
 *
 * @examplePrivate
 *
 * <example-private name="po-advanced-filter" title="PO Busca Avançada">
 *   <file name="sample-po-advanced-filter.component.html"> </file>
 *   <file name="sample-po-advanced-filter.component.ts"> </file>
 * </example-private>
 */
let PoAdvancedFilterComponent = class PoAdvancedFilterComponent extends PoAdvancedFilterBaseComponent {
    constructor(languageService) {
        super(languageService);
    }
    open() {
        this.filter = this.keepFilters ? this.getInitialValuesFromFilter(this.filters) : {};
        this.poModal.open();
    }
    getInitialValuesFromFilter(filters) {
        return filters.reduce((result, item) => Object.assign(result, { [item.property]: item.initValue }), {});
    }
};
PoAdvancedFilterComponent.ctorParameters = () => [
    { type: PoLanguageService }
];
__decorate([
    ViewChild(PoDynamicFormComponent, { static: true }),
    __metadata("design:type", PoDynamicFormComponent)
], PoAdvancedFilterComponent.prototype, "poDynamicForm", void 0);
PoAdvancedFilterComponent = __decorate([
    Component({
        selector: 'po-advanced-filter',
        template: "<po-modal\n  p-hide-close\n  [p-primary-action]=\"primaryAction\"\n  [p-secondary-action]=\"secondaryAction\"\n  [p-title]=\"literals.title\"\n>\n  <po-dynamic-form [p-fields]=\"filters\" [p-value]=\"filter\"> </po-dynamic-form>\n</po-modal>\n"
    }),
    __metadata("design:paramtypes", [PoLanguageService])
], PoAdvancedFilterComponent);
export { PoAdvancedFilterComponent };
