// @po-ui/ng-templates/esm5/lib/components/po-page-background/po-page-background.component.js
import { __decorate, __metadata } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { browserLanguage, convertToBoolean, isTypeof } from './../../utils/util';
var PoPageBackgroundComponent = /** @class */ (function () {
    /**
     * @docsPrivate
     *
     * @description
     *
     * Componente para definição de cor de fundo e dos logotipos primário e secundário para os templates
     * de `po-page-login` e demais templates de login.
     */
    function PoPageBackgroundComponent() {
        this._showSelectLanguage = false;
        this.selectLanguageOptions = [
            { label: 'English', value: 'en' },
            { label: 'Español', value: 'es' },
            { label: 'Português', value: 'pt' },
            { label: 'Pусский', value: 'ru' }
        ];
        /**
         * Evento disparado ao selecionar alguma opção no seletor de idiomas.
         * Para este evento será passado como parâmetro o valor de idioma selecionado.
         */
        this.selectedLanguage = new EventEmitter();
    }
    Object.defineProperty(PoPageBackgroundComponent.prototype, "logo", {
        get: function () {
            return this._logo;
        },
        /** Caminho para a logomarca localizada na parte superior. */
        set: function (value) {
            this._logo = isTypeof(value, 'string') && value.trim() ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageBackgroundComponent.prototype, "secondaryLogo", {
        get: function () {
            return this._secondaryLogo;
        },
        /**
         * @optional
         *
         * @description
         *
         * Caminho para a logomarca localizada no rodapé.
         */
        set: function (value) {
            this._secondaryLogo = isTypeof(value, 'string') && value.trim() ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageBackgroundComponent.prototype, "showSelectLanguage", {
        get: function () {
            return this._showSelectLanguage;
        },
        /** Define se o seletor de idiomas deve ser exibido. */
        set: function (showSelectLanguage) {
            this._showSelectLanguage = convertToBoolean(showSelectLanguage);
        },
        enumerable: true,
        configurable: true
    });
    PoPageBackgroundComponent.prototype.ngOnInit = function () {
        this.selectedLanguageOption = browserLanguage();
    };
    PoPageBackgroundComponent.prototype.onChangeLanguage = function () {
        this.selectedLanguage.emit(this.selectedLanguageOption);
    };
    __decorate([
        Input('p-background'),
        __metadata("design:type", String)
    ], PoPageBackgroundComponent.prototype, "background", void 0);
    __decorate([
        Input('p-hide-logo'),
        __metadata("design:type", Boolean)
    ], PoPageBackgroundComponent.prototype, "hideLogo", void 0);
    __decorate([
        Input('p-highlight-info'),
        __metadata("design:type", String)
    ], PoPageBackgroundComponent.prototype, "highlightInfo", void 0);
    __decorate([
        Input('p-logo'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PoPageBackgroundComponent.prototype, "logo", null);
    __decorate([
        Input('p-secondary-logo'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PoPageBackgroundComponent.prototype, "secondaryLogo", null);
    __decorate([
        Input('p-show-select-language'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], PoPageBackgroundComponent.prototype, "showSelectLanguage", null);
    __decorate([
        Output('p-selected-language'),
        __metadata("design:type", EventEmitter)
    ], PoPageBackgroundComponent.prototype, "selectedLanguage", void 0);
    PoPageBackgroundComponent = __decorate([
        Component({
            selector: 'po-page-background',
            template: "<div class=\"po-page-login-container\">\n  <div class=\"po-page-login-panel\">\n    <img\n      *ngIf=\"logo\"\n      class=\"po-page-background-main-logo\"\n      [class.po-page-background-hide-logo]=\"hideLogo\"\n      alt=\"main-logo\"\n      [src]=\"logo\"\n    />\n\n    <div class=\"po-page-login-body\">\n      <div class=\"po-page-login-panel-content\">\n        <ng-content></ng-content>\n      </div>\n    </div>\n    <div class=\"po-page-background-footer po-sm-12\">\n      <po-divider class=\"po-page-background-footer-mobile-only\"></po-divider>\n\n      <div class=\"po-page-background-footer-content\">\n        <div *ngIf=\"showSelectLanguage\" class=\"po-page-background-footer-select\">\n          <po-select\n            name=\"selectedLanguageOption\"\n            [(ngModel)]=\"selectedLanguageOption\"\n            [p-options]=\"selectLanguageOptions\"\n            (p-change)=\"onChangeLanguage()\"\n          >\n          </po-select>\n        </div>\n\n        <div\n          class=\"po-page-background-secondary-logo\"\n          [ngClass]=\"\n            showSelectLanguage\n              ? 'po-page-background-secondary-logo-right'\n              : 'po-page-background-secondary-logo-centered'\n          \"\n        >\n          <img\n            *ngIf=\"secondaryLogo\"\n            class=\"po-page-background-secondary-logo-image\"\n            alt=\"secondary-logo\"\n            [src]=\"secondaryLogo\"\n          />\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <div\n    *ngIf=\"background\"\n    [ngClass]=\"background ? 'po-page-login-highlight-image' : 'po-page-login-highlight-image-off'\"\n    [style.background-image]=\"'url(' + background + ')'\"\n  >\n    <div class=\"po-page-login-highlight-text\">\n      <div class=\"po-font-display\">{{ highlightInfo }}</div>\n    </div>\n  </div>\n</div>\n"
        })
        /**
         * @docsPrivate
         *
         * @description
         *
         * Componente para definição de cor de fundo e dos logotipos primário e secundário para os templates
         * de `po-page-login` e demais templates de login.
         */
    ], PoPageBackgroundComponent);
    return PoPageBackgroundComponent;
}());
export { PoPageBackgroundComponent };
