// @po-ui/ng-templates/esm5/lib/components/po-page-dynamic-search/po-page-dynamic-search.component.js
import { __assign, __decorate, __extends, __metadata, __read, __spread } from "tslib";
import { Component, ViewChild, OnInit, OnDestroy, ChangeDetectorRef } from '@angular/core';
import { PoDisclaimerGroup, PoDynamicFieldType, PoDynamicFormField, PoLanguageService, PoPageFilter, PoDisclaimerGroupRemoveAction } from '@po-ui/ng-components';
import { capitalizeFirstLetter, getBrowserLanguage } from '../../utils/util';
import { PoPageCustomizationService } from '../../services/po-page-customization/po-page-customization.service';
import { PoAdvancedFilterComponent } from './po-advanced-filter/po-advanced-filter.component';
import { PoPageDynamicSearchBaseComponent } from './po-page-dynamic-search-base.component';
/**
 * @docsExtends PoPageDynamicSearchBaseComponent
 *
 * @example
 *
 * <example name="po-page-dynamic-search-basic" title="PO Page Dynamic Search Basic">
 *  <file name="sample-po-page-dynamic-search-basic/sample-po-page-dynamic-search-basic.component.html"> </file>
 *  <file name="sample-po-page-dynamic-search-basic/sample-po-page-dynamic-search-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-dynamic-search-hiring-processes" title="PO Page Dynamic Search - Hiring processes">
 *  <file name="sample-po-page-dynamic-search-hiring-processes/sample-po-page-dynamic-search-hiring-processes.component.html"> </file>
 *  <file name="sample-po-page-dynamic-search-hiring-processes/sample-po-page-dynamic-search-hiring-processes.component.ts"> </file>
 *  <file name="sample-po-page-dynamic-search-hiring-processes/sample-po-page-dynamic-search-hiring-processes.service.ts"> </file>
 * </example>
 */
var PoPageDynamicSearchComponent = /** @class */ (function (_super) {
    __extends(PoPageDynamicSearchComponent, _super);
    function PoPageDynamicSearchComponent(languageService, poPageCustomizationService, changeDetector) {
        var _this = _super.call(this, languageService) || this;
        _this.poPageCustomizationService = poPageCustomizationService;
        _this.changeDetector = changeDetector;
        _this._disclaimerGroup = {
            remove: _this.onRemoveDisclaimer.bind(_this),
            removeAll: _this.onRemoveAllDisclaimers.bind(_this),
            disclaimers: [],
            title: _this.literals.disclaimerGroupTitle
        };
        _this._filterSettings = {
            action: _this.onAction.bind(_this),
            advancedAction: _this.onAdvancedAction.bind(_this),
            placeholder: _this.literals.searchPlaceholder
        };
        return _this;
    }
    Object.defineProperty(PoPageDynamicSearchComponent.prototype, "disclaimerGroup", {
        get: function () {
            return Object.assign({}, this._disclaimerGroup, { title: this.literals.disclaimerGroupTitle });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDynamicSearchComponent.prototype, "filterSettings", {
        get: function () {
            this._filterSettings.advancedAction = this.filters.length === 0 ? undefined : this.onAdvancedAction.bind(this);
            return Object.assign({}, this._filterSettings, { placeholder: this.literals.searchPlaceholder });
        },
        enumerable: true,
        configurable: true
    });
    PoPageDynamicSearchComponent.prototype.ngOnInit = function () {
        this.setAdvancedFilterLiterals(this.literals);
        if (this.onLoad) {
            this.loadOptionsOnInitialize(this.onLoad);
        }
    };
    PoPageDynamicSearchComponent.prototype.ngOnDestroy = function () {
        if (this.loadSubscription) {
            this.loadSubscription.unsubscribe();
        }
    };
    PoPageDynamicSearchComponent.prototype.onChangeFilters = function (filters) {
        var filterObjectWithValue = filters
            .filter(function (filter) { return filter.initValue; })
            .reduce(function (prev, current) {
            var _a;
            return __assign(__assign({}, prev), (_a = {}, _a[current.property] = current.initValue, _a));
        }, {});
        if (Object.keys(filterObjectWithValue).length) {
            this.onAdvancedSearch(filterObjectWithValue);
        }
    };
    PoPageDynamicSearchComponent.prototype.onAction = function (quickFilter) {
        var _this = this;
        var disclaimerQuickSearchUpdated = {
            property: 'search',
            label: this.literals.quickSearchLabel + " " + quickFilter,
            value: quickFilter
        };
        var getDisclaimersWithConcatFilters = function () { return __spread(_this.getDisclaimersWithoutQuickSearch(), [
            disclaimerQuickSearchUpdated
        ]); };
        this._disclaimerGroup.disclaimers = this.concatFilters
            ? getDisclaimersWithConcatFilters()
            : [disclaimerQuickSearchUpdated];
        if (this.quickSearch.observers && this.quickSearch.observers.length > 0) {
            this.quickSearch.emit(quickFilter);
        }
        if (this.keepFilters && !this.concatFilters) {
            this.filters.forEach(function (element) { return delete element.initValue; });
        }
        this.changeDetector.detectChanges();
    };
    PoPageDynamicSearchComponent.prototype.onAdvancedAction = function () {
        this.poAdvancedFilter.open();
    };
    PoPageDynamicSearchComponent.prototype.onAdvancedSearch = function (filters) {
        this._disclaimerGroup.disclaimers = this.setDisclaimers(filters);
        this.setFilters(filters);
        this.advancedSearch.emit(filters);
    };
    PoPageDynamicSearchComponent.prototype.getDisclaimersWithoutQuickSearch = function () {
        var quickSearchProperty = 'search';
        return this._disclaimerGroup.disclaimers.filter(function (item) { return item.property !== quickSearchProperty; });
    };
    PoPageDynamicSearchComponent.prototype.setFilters = function (filters) {
        var formattedFilters = this.convertToFilters(filters);
        this.filters.forEach(function (element) {
            var compatibleObject = formattedFilters.find(function (item) { return item.property === element.property; });
            if (compatibleObject) {
                element.initValue = compatibleObject.value;
            }
            else {
                delete element.initValue;
            }
        });
    };
    PoPageDynamicSearchComponent.prototype.convertToFilters = function (filters) {
        return Object.entries(filters).map(function (_a) {
            var _b = __read(_a, 2), property = _b[0], value = _b[1];
            return ({ property: property, value: value });
        });
    };
    PoPageDynamicSearchComponent.prototype.applyDisclaimerLabelValue = function (field, filterValue) {
        var values = Array.isArray(filterValue) ? filterValue : [filterValue];
        var labels = values.map(function (value) {
            var filteredField = field.options.find(function (option) { return option.value === value || option === value; });
            if (filteredField) {
                return filteredField.label || filteredField.value || filteredField;
            }
        });
        return labels.join(', ');
    };
    PoPageDynamicSearchComponent.prototype.formatDate = function (date) {
        var year = parseInt(date.substr(0, 4), 10);
        var month = parseInt(date.substr(5, 2), 10);
        var day = parseInt(date.substr(8, 2), 10);
        return new Date(year, month - 1, day).toLocaleDateString(getBrowserLanguage());
    };
    PoPageDynamicSearchComponent.prototype.formatArrayToObjectKeyValue = function (filters) {
        var formattedObject = filters.reduce(function (result, item) {
            var _a;
            return Object.assign(result, (_a = {}, _a[item.property] = item.value || item.initValue, _a));
        }, {});
        Object.keys(formattedObject).forEach(function (key) {
            if (!formattedObject[key]) {
                delete formattedObject[key];
            }
        });
        return formattedObject;
    };
    PoPageDynamicSearchComponent.prototype.getFieldByProperty = function (fields, fieldName) {
        return fields.find(function (field) { return field.property === fieldName; });
    };
    PoPageDynamicSearchComponent.prototype.getFilterValueToDisclaimer = function (field, value) {
        if (field.type === PoDynamicFieldType.Date) {
            return this.formatDate(value);
        }
        if (field.options && value) {
            return this.applyDisclaimerLabelValue(field, value);
        }
        return value;
    };
    PoPageDynamicSearchComponent.prototype.onRemoveDisclaimer = function (removeData) {
        var currentDisclaimers = removeData.currentDisclaimers;
        this.emitChangesDisclaimers(currentDisclaimers);
    };
    PoPageDynamicSearchComponent.prototype.emitChangesDisclaimers = function (currentDisclaimers) {
        this.changeDisclaimers.emit(currentDisclaimers);
        this.setFilters(this.formatArrayToObjectKeyValue(currentDisclaimers));
    };
    PoPageDynamicSearchComponent.prototype.onRemoveAllDisclaimers = function () {
        this.emitChangesDisclaimers([]);
    };
    PoPageDynamicSearchComponent.prototype.setDisclaimers = function (filters) {
        var _this = this;
        var disclaimers = [];
        var properties = Object.keys(filters);
        properties.forEach(function (property) {
            var field = _this.getFieldByProperty(_this.filters, property);
            var label = field.label || capitalizeFirstLetter(field.property);
            var value = filters[property];
            var valueDisplayedOnTheDisclaimerLabel = _this.getFilterValueToDisclaimer(field, value);
            if (valueDisplayedOnTheDisclaimerLabel !== '') {
                disclaimers.push({
                    label: label + ": " + valueDisplayedOnTheDisclaimerLabel,
                    property: property,
                    value: value
                });
            }
        });
        return disclaimers;
    };
    PoPageDynamicSearchComponent.prototype.loadOptionsOnInitialize = function (onLoad) {
        var _this = this;
        this.loadSubscription = this.getPoDynamicPageOptions(onLoad).subscribe(function (responsePoOption) {
            return _this.poPageCustomizationService.changeOriginalOptionsToNewOptions(_this, responsePoOption);
        });
    };
    PoPageDynamicSearchComponent.prototype.getPoDynamicPageOptions = function (onLoad) {
        var originalOption = {
            title: this.title,
            actions: this.actions,
            breadcrumb: this.breadcrumb,
            filters: this.filters,
            keepFilters: this.keepFilters,
            concatFilters: this.concatFilters
        };
        var pageOptionSchema = {
            schema: [
                {
                    nameProp: 'filters',
                    merge: true,
                    keyForMerge: 'property'
                },
                {
                    nameProp: 'actions',
                    merge: true,
                    keyForMerge: 'label'
                },
                {
                    nameProp: 'breadcrumb'
                },
                {
                    nameProp: 'title'
                },
                {
                    nameProp: 'keepFilters'
                },
                {
                    nameProp: 'concatFilters'
                }
            ]
        };
        return this.poPageCustomizationService.getCustomOptions(onLoad, originalOption, pageOptionSchema);
    };
    PoPageDynamicSearchComponent.ctorParameters = function () { return [
        { type: PoLanguageService },
        { type: PoPageCustomizationService },
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        ViewChild(PoAdvancedFilterComponent, { static: true }),
        __metadata("design:type", PoAdvancedFilterComponent)
    ], PoPageDynamicSearchComponent.prototype, "poAdvancedFilter", void 0);
    PoPageDynamicSearchComponent = __decorate([
        Component({
            selector: 'po-page-dynamic-search',
            template: "<po-page-list\n  [p-actions]=\"actions\"\n  [p-breadcrumb]=\"breadcrumb\"\n  [p-disclaimer-group]=\"disclaimerGroup\"\n  [p-filter]=\"filterSettings\"\n  [p-title]=\"title\"\n>\n  <po-advanced-filter\n    [p-filters]=\"filters\"\n    [p-keep-filters]=\"keepFilters\"\n    [p-literals]=\"advancedFilterLiterals\"\n    (p-search-event)=\"onAdvancedSearch($event)\"\n  >\n  </po-advanced-filter>\n\n  <ng-content></ng-content>\n</po-page-list>\n"
        }),
        __metadata("design:paramtypes", [PoLanguageService,
            PoPageCustomizationService,
            ChangeDetectorRef])
    ], PoPageDynamicSearchComponent);
    return PoPageDynamicSearchComponent;
}(PoPageDynamicSearchBaseComponent));
export { PoPageDynamicSearchComponent };
