// @po-ui/ng-templates/esm5/lib/components/po-page-login/po-page-login-popover/po-page-login-popover.component.js
import { __decorate, __metadata } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { isExternalLink, isTypeof } from '../../../utils/util';
var PoPageLoginPopoverComponent = /** @class */ (function () {
    /**
     * @docsPrivate
     *
     * @description
     *
     * Componente para definição da mensagem de aviso de bloqueio do `po-page-login`.
     */
    function PoPageLoginPopoverComponent() {
        /** se 'p-recovery' for do tipo Function ou PoPageLoginRecovery, emite para o método 'openUrl' do componente 'po-page-login' */
        this.forgotPassword = new EventEmitter();
    }
    Object.defineProperty(PoPageLoginPopoverComponent.prototype, "recovery", {
        get: function () {
            return this._recovery;
        },
        /** exibe o link de 'esqueci minha senha' e verifica se o valor é um link interno ou externo */
        set: function (value) {
            this._recovery = value;
            if (isTypeof(value, 'string')) {
                this.recoveryType = isExternalLink(value) ? 'externalLink' : 'internalLink';
            }
        },
        enumerable: true,
        configurable: true
    });
    PoPageLoginPopoverComponent.prototype.onForgotPasswordClick = function (recovery) {
        this.forgotPassword.emit(recovery);
    };
    __decorate([
        Input('p-literals'),
        __metadata("design:type", Object)
    ], PoPageLoginPopoverComponent.prototype, "literals", void 0);
    __decorate([
        Input('p-recovery'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PoPageLoginPopoverComponent.prototype, "recovery", null);
    __decorate([
        Input('p-remaining-attempts'),
        __metadata("design:type", Number)
    ], PoPageLoginPopoverComponent.prototype, "remainingAttempts", void 0);
    __decorate([
        Output('p-forgot-password'),
        __metadata("design:type", Object)
    ], PoPageLoginPopoverComponent.prototype, "forgotPassword", void 0);
    PoPageLoginPopoverComponent = __decorate([
        Component({
            selector: 'po-page-login-popover',
            template: "<div class=\"po-page-login-popover-container\">\n  <div class=\"po-page-login-popover\">\n    <div class=\"po-page-login-popover-arrow\"></div>\n    <div class=\"po-page-login-popover-content\">\n      <p class=\"po-font-text-bold po-page-login-popover-title\">\n        {{ literals?.titlePopover }}\n      </p>\n      <p class=\"po-font-text\">\n        {{ literals?.forgotYourPassword }}\n      </p>\n      <p class=\"po-font-text\">\n        {{ literals?.ifYouTryHarder }}\n        <span class=\"po-font-text-bold po-page-login-popover-attempts\">\n          {{ literals?.attempts | poI18n: remainingAttempts }} </span\n        >{{ literals?.yourUserWillBeBlocked }}\n      </p>\n      <br />\n      <p class=\"po-font-text\">\n        {{ literals?.createANewPasswordNow }}\n      </p>\n      <div *ngIf=\"recovery\" class=\"po-font-text-bold po-mt-1 po-page-login-popover-link-container\">\n        <a *ngIf=\"recoveryType === 'internalLink'\" class=\"po-page-login-popover-link\" [routerLink]=\"recovery\">\n          {{ literals?.iForgotMyPassword }}\n        </a>\n        <a *ngIf=\"recoveryType === 'externalLink'\" class=\"po-page-login-popover-link\" [href]=\"recovery\" target=\"_blank\">\n          {{ literals?.iForgotMyPassword }}\n        </a>\n        <a *ngIf=\"!recoveryType\" class=\"po-page-login-popover-link\" (click)=\"onForgotPasswordClick(recovery)\">\n          {{ literals?.iForgotMyPassword }}\n        </a>\n      </div>\n    </div>\n  </div>\n</div>\n"
        })
        /**
         * @docsPrivate
         *
         * @description
         *
         * Componente para definição da mensagem de aviso de bloqueio do `po-page-login`.
         */
    ], PoPageLoginPopoverComponent);
    return PoPageLoginPopoverComponent;
}());
export { PoPageLoginPopoverComponent };
