// dts-backoffice-util/esm2015/lib/model/session-info.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function ISessionInfo() { }
if (false) {
    /** @type {?} */
    ISessionInfo.prototype.diStatus;
    /** @type {?} */
    ISessionInfo.prototype.productVersion;
    /** @type {?} */
    ISessionInfo.prototype.userPwd;
    /** @type {?} */
    ISessionInfo.prototype.sessionId;
    /** @type {?} */
    ISessionInfo.prototype.userDomain;
    /** @type {?} */
    ISessionInfo.prototype.userName;
    /** @type {?} */
    ISessionInfo.prototype.userSO;
    /** @type {?} */
    ISessionInfo.prototype.timeout;
}
export class SessionInfo {
    /**
     * @param {?=} values
     */
    constructor(values = {}) {
        Object.assign(this, values);
    }
    /**
     * @return {?}
     */
    get $diStatus() {
        return this.diStatus;
    }
    /**
     * @return {?}
     */
    get $productVersion() {
        return this.productVersion;
    }
    /**
     * @return {?}
     */
    get $userPwd() {
        return this.userPwd;
    }
    /**
     * @return {?}
     */
    get $sessionId() {
        return this.sessionId;
    }
    /**
     * @return {?}
     */
    get $userDomain() {
        return this.userDomain;
    }
    /**
     * @return {?}
     */
    get $userName() {
        return this.userName;
    }
    /**
     * @return {?}
     */
    get $userSO() {
        return this.userSO;
    }
    /**
     * @return {?}
     */
    get $timeout() {
        return this.timeout;
    }
    /**
     * @param {?} diStatus
     * @return {?}
     */
    set $diStatus(diStatus) {
        this.diStatus = diStatus;
    }
    /**
     * @param {?} productVersion
     * @return {?}
     */
    set $productVersion(productVersion) {
        this.productVersion = productVersion;
    }
    /**
     * @param {?} userPwd
     * @return {?}
     */
    set $userPwd(userPwd) {
        this.userPwd = userPwd;
    }
    /**
     * @param {?} sessionId
     * @return {?}
     */
    set $sessionId(sessionId) {
        this.sessionId = sessionId;
    }
    /**
     * @param {?} userDomain
     * @return {?}
     */
    set $userDomain(userDomain) {
        this.userDomain = userDomain;
    }
    /**
     * @param {?} userName
     * @return {?}
     */
    set $userName(userName) {
        this.userName = userName;
    }
    /**
     * @param {?} userSO
     * @return {?}
     */
    set $userSO(userSO) {
        this.userSO = userSO;
    }
    /**
     * @param {?} timeout
     * @return {?}
     */
    set $timeout(timeout) {
        this.timeout = timeout;
    }
    /**
     * @param {?=} json
     * @return {?}
     */
    static of(json = {}) {
        return new SessionInfo(json);
    }
    /**
     * @return {?}
     */
    static empty() {
        return new SessionInfo();
    }
    /**
     * @param {?=} json
     * @return {?}
     */
    static fromJson(json = []) {
        /** @type {?} */
        const items = [];
        for (const values of json) {
            items.push(new SessionInfo(values));
        }
        return items;
    }
}
if (false) {
    /** @type {?} */
    SessionInfo.prototype.diStatus;
    /** @type {?} */
    SessionInfo.prototype.productVersion;
    /** @type {?} */
    SessionInfo.prototype.userPwd;
    /** @type {?} */
    SessionInfo.prototype.sessionId;
    /** @type {?} */
    SessionInfo.prototype.userDomain;
    /** @type {?} */
    SessionInfo.prototype.userName;
    /** @type {?} */
    SessionInfo.prototype.userSO;
    /** @type {?} */
    SessionInfo.prototype.timeout;
}
