// dts-backoffice-util/esm5/lib/components/totvs-schedule-execution/totvs-schedule-execution.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, Output, EventEmitter } from '@angular/core';
import { PoI18nService, PoNotificationService } from '@po-ui/ng-components';
import { ParametersRpw } from './totvs-schedule-execution.model';
import { NgForm } from '@angular/forms';
import { RpwService } from './totvs-schedule-execution.service';
var TotvsScheduleExecutionComponent = /** @class */ (function () {
    function TotvsScheduleExecutionComponent(thfI18nService, rpwService, poNotification) {
        this.thfI18nService = thfI18nService;
        this.rpwService = rpwService;
        this.poNotification = poNotification;
        this.programEMS5 = false;
        this.programVersion = '';
        this.endExecution = new EventEmitter();
        this.jsonObject = {};
        this.literals = {};
    }
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /* VERIFICAR COMO FICARÁ A TRADUÇÃO
        forkJoin(
              this.thfI18nService.getLiterals(),
              this.thfI18nService.getLiterals({ context: 'generalRpw' })
          ).subscribe(literals => {
              literals.map(item => Object.assign(this.literals, item) );
              this.setupComponents();
          });*/
        this.setupComponents();
    };
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.setupComponents = /**
     * @return {?}
     */
    function () {
        this.executionTypeOptions = [
            { label: 'Executar Hoje', value: 1 },
            { label: 'Agendar Execução', value: 2 }
        ];
        this.frequencyOptions = [
            { label: 'Uma vez no dia', value: 'no' },
            { label: 'Várias vezes no dia', value: 'yes' }
        ];
        this.frequencyTypeOptions = [
            { label: 'Hora(s)', value: 'hour' },
            { label: 'Minuto(s)', value: 'minute' }
        ];
        this.weeklyOptions = [
            { label: 'Domingo', value: 'Sunday' },
            { label: 'Segunda', value: 'Monday' },
            { label: 'Terça', value: 'Tuesday' },
            { label: 'Quarta', value: 'Wednesday' },
            { label: 'Quinta', value: 'Thursday' },
            { label: 'Sexta', value: 'Friday' },
            { label: 'Sábado', value: 'Saturday' }
        ];
        this.columns = [
            { property: 'cod_servid_exec', label: 'Código' },
            { property: 'des_servid_exec', label: 'Descrição' }
        ];
        this.model = new ParametersRpw();
        this.model.executionType = 1;
        this.model.activeTab = 1;
        this.model.repeatExecution = false;
    };
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.isExecutionSchedule = /**
     * @return {?}
     */
    function () {
        return (this.model.executionType === 2);
    };
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.isRepeatExecution = /**
     * @return {?}
     */
    function () {
        return this.model.repeatExecution;
    };
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.isFrenquency = /**
     * @return {?}
     */
    function () {
        return (this.model.frequency === 'yes');
    };
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.changeRepeatExecution = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var date = new Date();
        this.model.execAppointHourInit = this.addZero(date.getHours()) + ":" + this.addZero(date.getMinutes());
        this.model.execAppointHourFinal = this.addZero(date.getHours()) + ":" + this.addZero(date.getMinutes());
        this.model.selectWeeklys = [];
        this.model.dayOfMonth = 0;
        this.model.frequency = 'no';
        this.model.frequencyType = 'hour';
        this.model.frequencyValue = 0;
    };
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.changeExecutionType = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var date = new Date();
        this.model.execAppointDate = date;
        this.model.execAppointHour = this.addZero(date.getHours()) + ":" + this.addZero(date.getMinutes());
    };
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.changeTypeFrequency = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} codTab
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.setActiveTab = /**
     * @param {?} codTab
     * @return {?}
     */
    function (codTab) {
        this.model.activeTab = codTab;
    };
    /**
     * @param {?} i
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.addZero = /**
     * @param {?} i
     * @return {?}
     */
    function (i) {
        if (i < 10) {
            i = '0' + i;
        }
        return i;
    };
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.executeSchedule = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.validate()) {
            return;
        }
        this.jsonObject = {};
        this.jsonObject.status = 'active';
        this.jsonObject.processID = this.programName;
        this.jsonObject.recurrent = this.model.repeatExecution;
        this.jsonObject.executionParameter = {};
        if (this.model.executionType === 1) {
            /** @type {?} */
            var date = new Date();
            this.jsonObject.firstExecution = date.getFullYear() + "-" + (date.getMonth() + 1) + "-" + date.getDate() + "T" + this.addZero(date.getHours()) + ":" + this.addZero(date.getMinutes()) + ":00.000Z";
        }
        if (this.model.executionType === 2) {
            this.jsonObject.firstExecution = this.model.execAppointDate + "T" + this.model.execAppointHour + ":00.000Z";
        }
        this.jsonObject.executionParameter.parametros = [];
        this.jsonObject.executionParameter.parametros[0] = { chave: 'rpwServer', valor: this.model.executionServer };
        this.jsonObject.executionParameter.parametros[1] = { chave: 'RPW_PROGRAM', valor: this.externalName };
        this.jsonObject.executionParameter.parametros[2] = { chave: 'RPW_PRG_EMS5', valor: this.programEMS5 ? 'yes' : 'no' };
        this.jsonObject.executionParameter.parametros[3] = { chave: 'RPW_PRG_VERS', valor: this.programVersion };
        this.jsonObject.executionParameter.parametros[4] = {};
        this.jsonObject.executionParameter.parametros[4].parametros_negocio = this.parameters;
        // Executa hoje ou agendada
        this.rpwService.createRpw(this.jsonObject).subscribe((/**
         * @return {?}
         */
        function () {
            _this.poNotification.success('Execução efetuada com sucesso !');
        }));
        if (this.model.repeatExecution) {
            if (this.model.activeTab === 1) {
                this.jsonObject.daily = {
                    hour: this.getHourOrMinute(this.model.execAppointHourInit, 'h'),
                    minute: this.getHourOrMinute(this.model.execAppointHourInit, 'm')
                };
            }
            if (this.model.activeTab === 2) {
                this.jsonObject.weekly = {
                    hour: this.getHourOrMinute(this.model.execAppointHourInit, 'h'),
                    minute: this.getHourOrMinute(this.model.execAppointHourInit, 'm'),
                    daysOfWeek: this.model.selectWeeklys
                };
            }
            if (this.model.activeTab === 3) {
                this.jsonObject.monthly = {
                    hour: this.getHourOrMinute(this.model.execAppointHourInit, 'h'),
                    minute: this.getHourOrMinute(this.model.execAppointHourInit, 'm'),
                    day: this.model.dayOfMonth
                };
            }
            if (this.isFrenquency()) {
                this.jsonObject.rangeExecutions = {
                    frequency: {
                        type: this.model.frequencyType,
                        value: this.model.frequencyValue
                    },
                    rangeLimit: {
                        hour: this.getHourOrMinute(this.model.execAppointHourFinal, 'h'),
                        minute: this.getHourOrMinute(this.model.execAppointHourFinal, 'm')
                    }
                };
            }
            // Executa a diária, semanal ou mensal
            this.rpwService.createRpw(this.jsonObject).subscribe((/**
             * @return {?}
             */
            function () {
            }));
        }
        this.endExecution.emit('endExecution');
    };
    /**
     * @param {?} value
     * @param {?} type
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.getHourOrMinute = /**
     * @param {?} value
     * @param {?} type
     * @return {?}
     */
    function (value, type) {
        if (type === 'h') {
            return +value.substring(0, 2);
        }
        if (type === 'm') {
            return +value.substring(3, 6);
        }
    };
    /**
     * @param {?} hourInit
     * @param {?} hourFinal
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.compareHour = /**
     * @param {?} hourInit
     * @param {?} hourFinal
     * @return {?}
     */
    function (hourInit, hourFinal) {
        if (hourInit === hourFinal) {
            return false;
        }
        if (this.getHourOrMinute(hourInit, 'h') < this.getHourOrMinute(hourFinal, 'h')) {
            return true;
        }
        if (this.getHourOrMinute(hourInit, 'm') > this.getHourOrMinute(hourFinal, 'm')) {
            return false;
        }
        return true;
    };
    /**
     * @return {?}
     */
    TotvsScheduleExecutionComponent.prototype.validate = /**
     * @return {?}
     */
    function () {
        if (!this.scheduleExecutionForm.valid) {
            this.poNotification.error('Verifique as inconsistências em tela.');
            return false;
        }
        if (!this.model.executionServer) {
            this.poNotification.error('Servidor de Execução não foi informado.');
            return false;
        }
        if (this.isExecutionSchedule() && (!this.model.execAppointDate || !this.model.execAppointHour)) {
            this.poNotification.error('Informar a data e a hora para agendamento da execução.');
            return false;
        }
        if (this.model.repeatExecution) {
            if (!this.model.execAppointHourInit) {
                this.poNotification.error('Informar a hora de início para a execução.');
                return false;
            }
            if (this.model.activeTab === 2 && (this.model.selectWeeklys.length === 0)) {
                this.poNotification.error('Informar os dias da semana para a execução.');
                return false;
            }
            if (this.model.activeTab === 3 && (!this.model.dayOfMonth)) {
                this.poNotification.error('Informar o dia para a execução.');
                return false;
            }
            if (this.isFrenquency()) {
                if (!this.compareHour(this.model.execAppointHourInit, this.model.execAppointHourFinal)) {
                    this.poNotification.error('Hora Fim deve ser maior que Hora Início.');
                    return false;
                }
                if (this.model.frequencyValue === 0) {
                    this.poNotification.error('Informar a frequência de execução (A cada hora/minuto).');
                    return false;
                }
            }
        }
        return true;
    };
    TotvsScheduleExecutionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-totvs-schedule-execution',
                    template: "<form #scheduleExecutionForm=\"ngForm\">\r\n    <div class=\"po-row po-sm-12 po-md-12 po-lg-12 po-xl-12\">\r\n        <po-radio-group \r\n            class=\"po-sm-12 po-md-12 po-lg-12 po-xl-12\"\r\n            name=\"executionType\"\r\n            p-label=\"Data de Execu\u00E7\u00E3o\"\r\n            [(ngModel)]=\"model.executionType\"\r\n            [p-options]=\"executionTypeOptions\"\r\n            (p-change)=\"changeExecutionType()\">\r\n        </po-radio-group>\r\n\r\n        <po-datepicker *ngIf=isExecutionSchedule()\r\n            class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\"\r\n            name=\"model.execAppointDate\"\r\n            p-label=\"Agendar para\"\r\n            [(ngModel)]=\"model.execAppointDate\">\r\n        </po-datepicker>\r\n\r\n        <po-input *ngIf=isExecutionSchedule()\r\n            class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\"\r\n            name=\"execAppointHour\"\r\n            p-label=\"Hora\"\r\n            p-mask=\"99:99\"\r\n            p-mask-format-model=\"true\"\r\n            p-pattern=\"^([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$\"\r\n            p-error-pattern=\"Hora inv\u00E1lida\"\r\n            [(ngModel)]=\"model.execAppointHour\">\r\n        </po-input>\r\n    </div>\r\n\r\n    <div class=\"po-row po-sm-12 po-md-12 po-lg-12 po-xl-12\">\r\n        <po-combo\r\n            class=\"po-sm-12 po-md-12 po-lg-12 po-xl-12\"\r\n            name=\"executionServer\"\r\n            p-field-label=\"des_servid_exec\"\r\n            p-field-value=\"cod_servid_exec\"\r\n            p-label=\"Servidor de Execu\u00E7\u00E3o\"\r\n            p-placeholder=\"Servidor de Execuc\u00E3o\"\r\n            [(ngModel)]=\"model.executionServer\"\r\n            [p-filter-service]=\"rpwService\">\r\n        </po-combo>\r\n    </div>\r\n\r\n    <div class=\"po-row po-sm-12 po-md-12 po-lg-12 po-xl-12\">\r\n        <po-switch\r\n            class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\"\r\n            name=\"repeatExecution\"\r\n            p-label=\"Repetir ocorr\u00EAncia ?\"\r\n            p-label-off=\"N\u00E3o\"\r\n            p-label-on=\"Sim\"\r\n            [(ngModel)]=\"model.repeatExecution\"\r\n            (click)=\"changeRepeatExecution()\">\r\n        </po-switch>\r\n    </div>\r\n\r\n    <div class=\"po-row\" *ngIf=\"isRepeatExecution()\">\r\n        <po-tabs>\r\n            <po-tab p-label=\"Di\u00E1ria\" p-active (p-click)=\"setActiveTab(1)\">\r\n                <div class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\">\r\n                    <po-radio-group \r\n                        class=\"po-sm-12 po-md-12 po-lg-12 po-xl-12\"\r\n                        name=\"frequencyDaily\"\r\n                        p-label=\"Frequ\u00EAncia\"\r\n                        p-columns=1\r\n                        [(ngModel)]=\"model.frequency\"\r\n                        [p-options]=\"frequencyOptions\"\r\n                        (p-change)=\"changeTypeFrequency()\">\r\n                    </po-radio-group>\r\n                </div>\r\n\r\n                <div class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\">    \r\n                    <po-input\r\n                        class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\"\r\n                        name=\"execAppointHourInitDaily\"\r\n                        p-label=\"Hora In\u00EDcio\"\r\n                        p-mask=\"99:99\"\r\n                        p-mask-format-model=\"true\"\r\n                        p-pattern=\"^([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$\"\r\n                        p-error-pattern=\"Hora inv\u00E1lida\"\r\n                        [(ngModel)]=\"model.execAppointHourInit\">\r\n                    </po-input>\r\n\r\n                    <po-input *ngIf=\"isFrenquency()\"\r\n                        class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\"\r\n                        name=\"execAppointHourFinalDaily\"\r\n                        p-label=\"Hora Fim\"\r\n                        p-mask=\"99:99\"\r\n                        p-mask-format-model=\"true\"\r\n                        p-pattern=\"^([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$\"\r\n                        p-error-pattern=\"Hora inv\u00E1lida\"\r\n                        [(ngModel)]=\"model.execAppointHourFinal\">\r\n                    </po-input>\r\n\r\n                    <po-number *ngIf=\"isFrenquency()\"\r\n                        class=\"po-sm-5 po-md-5 po-lg-5 po-xl-5\"\r\n                        name=\"frequencyValueDaily\"\r\n                        p-label=\"A cada\"\r\n                        p-minlength=\"1\"\r\n                        p-maxlength=\"3\"\r\n                        p-min=\"0\"\r\n                        p-max=\"999\"\r\n                        [(ngModel)]=\"model.frequencyValue\">\r\n                    </po-number>\r\n\r\n                    <po-select *ngIf=\"isFrenquency()\"\r\n                        class=\"po-mt-3 po-sm-7 po-md-7 po-lg-7 po-xl-7\"\r\n                        name=\"frequencyTypeDaily\"\r\n                        p-label=\"\"\r\n                        [p-options]=\"frequencyTypeOptions\"\r\n                        [(ngModel)]=\"model.frequencyType\">\r\n                    </po-select>\r\n                </div>\r\n            </po-tab>\r\n\r\n            <po-tab p-label=\"Semanal\" (p-click)=\"setActiveTab(2)\">\r\n                <div class=\"po-row po-sm-12 po-md-12 po-lg-12 po-xl-12\">\r\n                    <po-multiselect\r\n                        class=\"po-sm-12 po-md-12 po-lg-12 po-xl-12\"\r\n                        name=\"selectWeeklys\"\r\n                        p-label=\"Dias da Semana\"\r\n                        p-placeholder=\"Escolha os dias da semana\"\r\n                        [(ngModel)]=\"model.selectWeeklys\"\r\n                        [p-options]=\"weeklyOptions\">\r\n                    </po-multiselect>\r\n                </div>\r\n\r\n                <div class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\">\r\n                    <po-radio-group \r\n                        class=\"po-sm-12 po-md-12 po-lg-12 po-xl-12\"\r\n                        name=\"frequencyWeekly\"\r\n                        p-label=\"Frequ\u00EAncia\"\r\n                        p-columns=1\r\n                        [(ngModel)]=\"model.frequency\"\r\n                        [p-options]=\"frequencyOptions\"\r\n                        (p-change)=\"changeTypeFrequency()\">\r\n                    </po-radio-group>\r\n                </div>\r\n\r\n                <div class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\">\r\n                    <po-input\r\n                        class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\"\r\n                        name=\"execAppointHourInitWeekly\"\r\n                        p-label=\"Hora In\u00EDcio\"\r\n                        p-mask=\"99:99\"\r\n                        p-mask-format-model=\"true\"\r\n                        p-pattern=\"^([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$\"\r\n                        p-error-pattern=\"Hora inv\u00E1lida\"\r\n                        [(ngModel)]=\"model.execAppointHourInit\">\r\n                    </po-input>\r\n\r\n                    <po-input *ngIf=\"isFrenquency()\"\r\n                        class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\"\r\n                        name=\"execAppointHourFinalWeekly\"\r\n                        p-label=\"Hora Fim\"\r\n                        p-mask=\"99:99\"\r\n                        p-mask-format-model=\"true\"\r\n                        p-pattern=\"^([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$\"\r\n                        p-error-pattern=\"Hora inv\u00E1lida\"\r\n                        [(ngModel)]=\"model.execAppointHourFinal\">\r\n                    </po-input>\r\n\r\n                    <po-number *ngIf=\"isFrenquency()\"\r\n                        class=\"po-sm-5 po-md-5 po-lg-5 po-xl-5\"\r\n                        name=\"frequencyValueWeekly\"\r\n                        p-label=\"A cada\"\r\n                        p-minlength=\"1\"\r\n                        p-maxlength=\"3\"\r\n                        p-min=\"0\"\r\n                        p-max=\"999\"\r\n                        [(ngModel)]=\"model.frequencyValue\">\r\n                    </po-number>\r\n\r\n                    <po-select *ngIf=\"isFrenquency()\"\r\n                        class=\"po-mt-3 po-sm-7 po-md-7 po-lg-7 po-xl-7\"\r\n                        name=\"frequencyTypeWeekly\"\r\n                        p-label=\"\"\r\n                        [p-options]=\"frequencyTypeOptions\"\r\n                        [(ngModel)]=\"model.frequencyType\">\r\n                    </po-select>\r\n                </div>\r\n            </po-tab>\r\n\r\n            <po-tab p-label=\"Mensal\" (p-click)=\"setActiveTab(3)\">\r\n                <div class=\"po-row po-sm-12 po-md-12 po-lg-12 po-xl-12\">\r\n                    <po-number\r\n                        class=\"po-sm-3 po-md-3 po-lg-3 po-xl-3\"\r\n                        name=\"dayOfMonth\"\r\n                        p-label=\"Dia do m\u00EAs\"\r\n                        p-minlength=\"1\"\r\n                        p-maxlength=\"2\"\r\n                        p-min=\"0\"\r\n                        p-max=\"31\"\r\n                        [(ngModel)]=\"model.dayOfMonth\">\r\n                    </po-number>\r\n                </div>\r\n\r\n                <div class=\"po-row po-sm-6 po-md-6 po-lg-6 po-xl-6\">\r\n                    <po-radio-group \r\n                        class=\"po-sm-12 po-md-12 po-lg-12 po-xl-12\"\r\n                        name=\"frequencyMonthly\"\r\n                        p-label=\"Frequ\u00EAncia\"\r\n                        p-columns=1\r\n                        [(ngModel)]=\"model.frequency\"\r\n                        [p-options]=\"frequencyOptions\"\r\n                        (p-change)=\"changeTypeFrequency()\">\r\n                    </po-radio-group>\r\n                </div>\r\n\r\n                 <div class=\"po-row po-sm-6 po-md-6 po-lg-6 po-xl-6\">\r\n                    <po-input\r\n                        class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\"\r\n                        name=\"execAppointHourInitMonthly\"\r\n                        p-label=\"Hora In\u00EDcio\"\r\n                        p-mask=\"99:99\"\r\n                        p-mask-format-model=\"true\"\r\n                        p-pattern=\"^([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$\"\r\n                        p-error-pattern=\"Hora inv\u00E1lida\"\r\n                        [(ngModel)]=\"model.execAppointHourInit\">\r\n                    </po-input>\r\n\r\n                    <po-input *ngIf=\"isFrenquency()\"\r\n                        class=\"po-sm-6 po-md-6 po-lg-6 po-xl-6\"\r\n                        name=\"execAppointHourFinalWeekly\"\r\n                        p-label=\"Hora Fim\"\r\n                        p-mask=\"99:99\"\r\n                        p-mask-format-model=\"true\"\r\n                        p-pattern=\"^([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$\"\r\n                        p-error-pattern=\"Hora inv\u00E1lida\"\r\n                        [(ngModel)]=\"model.execAppointHourFinal\">\r\n                    </po-input>\r\n\r\n                    <po-number *ngIf=\"isFrenquency()\"\r\n                        class=\"po-sm-5 po-md-5 po-lg-5 po-xl-5\"\r\n                        name=\"frequencyValueWeekly\"\r\n                        p-label=\"A cada\"\r\n                        p-minlength=\"1\"\r\n                        p-maxlength=\"3\"\r\n                        p-min=\"0\"\r\n                        p-max=\"999\"\r\n                        [(ngModel)]=\"model.frequencyValue\">\r\n                    </po-number>\r\n\r\n                    <po-select *ngIf=\"isFrenquency()\"\r\n                        class=\"po-mt-3 po-sm-7 po-md-7 po-lg-7 po-xl-7\"\r\n                        name=\"frequencyTypeWeekly\"\r\n                        p-label=\"\"\r\n                        [p-options]=\"frequencyTypeOptions\"\r\n                        [(ngModel)]=\"model.frequencyType\">\r\n                    </po-select>\r\n                </div>\r\n            </po-tab>\r\n        </po-tabs>\r\n    </div>\r\n\r\n    <div class=\"po-sm-12 po-md-12 po-lg-12 po-xl-12\" style=\"padding-top: 15px;\">\r\n        <po-button\r\n            style=\"float: right;\"\r\n            p-label=\"Gerar Agendamento RPW\"\r\n            (click)=\"executeSchedule()\">\r\n        </po-button>\r\n    </div>\r\n</form>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TotvsScheduleExecutionComponent.ctorParameters = function () { return [
        { type: PoI18nService },
        { type: RpwService },
        { type: PoNotificationService }
    ]; };
    TotvsScheduleExecutionComponent.propDecorators = {
        scheduleExecutionForm: [{ type: ViewChild, args: ['scheduleExecutionForm', { static: true },] }],
        programName: [{ type: Input }],
        externalName: [{ type: Input }],
        programEMS5: [{ type: Input }],
        programVersion: [{ type: Input }],
        parameters: [{ type: Input }],
        endExecution: [{ type: Output }]
    };
    return TotvsScheduleExecutionComponent;
}());
export { TotvsScheduleExecutionComponent };
if (false) {
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.scheduleExecutionForm;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.programName;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.externalName;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.programEMS5;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.programVersion;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.parameters;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.endExecution;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.executionTypeOptions;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.frequencyOptions;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.frequencyTypeOptions;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.weeklyOptions;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.model;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.columns;
    /**
     * @type {?}
     * @private
     */
    TotvsScheduleExecutionComponent.prototype.jsonObject;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.literals;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.thfI18nService;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.rpwService;
    /** @type {?} */
    TotvsScheduleExecutionComponent.prototype.poNotification;
}
