// @po-ui/ng-templates/esm5/lib/components/po-page-dynamic-table/po-page-dynamic-table-actions.service.js
import { __decorate, __metadata } from "tslib";
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
var PoPageDynamicTableActionsService = /** @class */ (function () {
    function PoPageDynamicTableActionsService(http) {
        this.http = http;
        this.headers = new HttpHeaders({
            'X-PO-SCREEN-LOCK': 'true'
        });
    }
    PoPageDynamicTableActionsService.prototype.beforeDuplicate = function (action, id, body) {
        var resource = body !== null && body !== void 0 ? body : {};
        return this.executeAction({ action: action, resource: resource, id: id });
    };
    PoPageDynamicTableActionsService.prototype.beforeEdit = function (action, id, body) {
        var resource = body !== null && body !== void 0 ? body : {};
        return this.executeAction({ action: action, resource: resource, id: id });
    };
    PoPageDynamicTableActionsService.prototype.beforeNew = function (action) {
        return this.executeAction({ action: action });
    };
    PoPageDynamicTableActionsService.prototype.beforeRemove = function (action, id, resource) {
        return this.executeAction({ action: action, id: id, resource: resource });
    };
    PoPageDynamicTableActionsService.prototype.beforeRemoveAll = function (action, resources) {
        return this.executeAction({ action: action, resource: resources });
    };
    PoPageDynamicTableActionsService.prototype.beforeDetail = function (action, id, resource) {
        return this.executeAction({ action: action, id: id, resource: resource });
    };
    PoPageDynamicTableActionsService.prototype.executeAction = function (_a) {
        var action = _a.action, _b = _a.resource, resource = _b === void 0 ? {} : _b, id = _a.id;
        if (!action) {
            return of({});
        }
        if (typeof action === 'string') {
            var url = id ? action + "/" + id : action;
            return this.http.post(url, resource, { headers: this.headers });
        }
        if (id) {
            return of(action(id, resource));
        }
        return of(action(resource));
    };
    PoPageDynamicTableActionsService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    PoPageDynamicTableActionsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PoPageDynamicTableActionsService_Factory() { return new PoPageDynamicTableActionsService(i0.ɵɵinject(i1.HttpClient)); }, token: PoPageDynamicTableActionsService, providedIn: "root" });
    PoPageDynamicTableActionsService = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __metadata("design:paramtypes", [HttpClient])
    ], PoPageDynamicTableActionsService);
    return PoPageDynamicTableActionsService;
}());
export { PoPageDynamicTableActionsService };
