// @po-ui/ng-templates/esm5/lib/components/po-page-blocked-user/po-page-blocked-user-reason/po-page-blocked-user-reason.component.js
import { __assign, __decorate, __metadata } from "tslib";
import { ChangeDetectorRef, Component, Input, OnChanges, OnInit, SimpleChanges } from '@angular/core';
import { browserLanguage, poLocaleDefault } from '../../../utils/util';
import { poPageBlockedUserLiterals } from './../literals/i18n/po-page-blocked-user-literals';
import { PoPageBlockedUserReason } from '../enums/po-page-blocked-user-reason.enum';
var PoPageBlockedUserReasonComponent = /** @class */ (function () {
    function PoPageBlockedUserReasonComponent(changeDetector) {
        this.changeDetector = changeDetector;
    }
    PoPageBlockedUserReasonComponent.prototype.ngOnChanges = function (changes) {
        if (changes.reason || changes.params) {
            this.getLiterals();
        }
    };
    PoPageBlockedUserReasonComponent.prototype.ngOnInit = function () {
        this.getLiterals();
    };
    PoPageBlockedUserReasonComponent.prototype.getImageByReasonType = function () {
        var reasonImage;
        switch (this.reason) {
            case 'none': {
                reasonImage = 'big-lock';
                break;
            }
            case 'exceededAttempts': {
                reasonImage = 'blocked-user';
                break;
            }
            case 'expiredPassword': {
                reasonImage = 'expired';
                break;
            }
        }
        return "./assets/images/" + reasonImage + ".svg";
    };
    PoPageBlockedUserReasonComponent.prototype.getParams = function () {
        this.literalParams =
            this.reason === 'expiredPassword'
                ? [this.params.days, this.params.days]
                : [this.params.attempts, this.params.hours];
    };
    PoPageBlockedUserReasonComponent.prototype.getLiterals = function () {
        this.getParams();
        this.literals = __assign(__assign({}, poPageBlockedUserLiterals[this.reason][poLocaleDefault]), poPageBlockedUserLiterals[this.reason][browserLanguage()]);
        this.changeDetector.detectChanges();
    };
    PoPageBlockedUserReasonComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input('p-params'),
        __metadata("design:type", Object)
    ], PoPageBlockedUserReasonComponent.prototype, "params", void 0);
    __decorate([
        Input('p-reason'),
        __metadata("design:type", String)
    ], PoPageBlockedUserReasonComponent.prototype, "reason", void 0);
    PoPageBlockedUserReasonComponent = __decorate([
        Component({
            selector: 'po-page-blocked-user-reason',
            template: "<div class=\"po-page-blocked-user-reason-content\">\n  <img class=\"po-page-blocked-user-image po-mb-3\" [src]=\"getImageByReasonType()\" />\n  <div class=\"po-page-blocked-user-header po-mb-md-3 po-row\">\n    <div class=\"po-md-12\">{{ literals?.title }}</div>\n  </div>\n  <div class=\"po-page-blocked-user-text po-font-text po-row\">\n    <div class=\"po-mb-sm-2 po-mb-md-3 po-md-12 po-page-blocked-user-text-bold\">\n      {{ literals?.firstPhrase | poI18n: literalParams }}\n    </div>\n    <div class=\"po-mb-sm-2 po-mb-md-3 po-md-12\">{{ literals?.secondPhrase }}</div>\n    <div class=\"po-mb-sm-2 po-mb-md-3 po-md-12\" *ngIf=\"literals?.thirdPhrase\">{{ literals?.thirdPhrase }}</div>\n  </div>\n</div>\n"
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], PoPageBlockedUserReasonComponent);
    return PoPageBlockedUserReasonComponent;
}());
export { PoPageBlockedUserReasonComponent };
