// dts-backoffice-util/esm5/lib/utils/disclaimer.util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
Versão: 1.009
Data Criação: 06/08/2018
*/
var DisclaimerUtil = /** @class */ (function () {
    function DisclaimerUtil(poNotification, poI18nPipe, literals) {
        this.literals = {};
        this.poNotification = poNotification;
        this.poI18nPipe = poI18nPipe;
        this.literals = literals;
    }
    /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} tradValue
     * @param {?=} vldBlank
     * @return {?}
     */
    DisclaimerUtil.prototype.makeDisclaimer = /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} tradValue
     * @param {?=} vldBlank
     * @return {?}
     */
    function (property, value, hideClose, tradValue, vldBlank) {
        if (hideClose === void 0) { hideClose = false; }
        if (tradValue === void 0) { tradValue = false; }
        if (vldBlank === void 0) { vldBlank = false; }
        if (value === null || value === undefined) {
            return { label: '', property: '', value: '' };
        }
        if (vldBlank && value === '') {
            return { label: '', property: '', value: '' };
        }
        return {
            hideClose: hideClose,
            label: this.literals[property] + ": " + (tradValue ? this.literals[value] : value),
            property: "" + property,
            value: "" + value
        };
    };
    /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} vldBlank
     * @return {?}
     */
    DisclaimerUtil.prototype.makeDisclaimerFromNumber = /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} vldBlank
     * @return {?}
     */
    function (property, value, hideClose, vldBlank) {
        if (hideClose === void 0) { hideClose = false; }
        if (vldBlank === void 0) { vldBlank = false; }
        if (value === null || value === undefined) {
            return { label: '', property: '', value: '' };
        }
        if (vldBlank && value === 0) {
            return { label: '', property: '', value: '' };
        }
        return {
            hideClose: hideClose,
            label: this.literals[property] + ": " + value,
            property: "" + property,
            value: "" + value
        };
    };
    /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @return {?}
     */
    DisclaimerUtil.prototype.makeDisclaimerFromBoolean = /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @return {?}
     */
    function (property, value, hideClose) {
        if (hideClose === void 0) { hideClose = false; }
        if (value === null || value === undefined) {
            return { label: '', property: '', value: '' };
        }
        return {
            hideClose: hideClose,
            label: this.literals[property] + ": " + (value ? this.literals['yes'] : this.literals['no']),
            property: "" + property,
            value: "" + value
        };
    };
    /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @return {?}
     */
    DisclaimerUtil.prototype.makeDisclaimerFromDate = /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @return {?}
     */
    function (property, value, hideClose) {
        if (hideClose === void 0) { hideClose = false; }
        /** @type {?} */
        var dDate = this.ajustDate(value);
        if (!this.isValidDate(dDate)) {
            return { label: '', property: '', value: '' };
        }
        return {
            hideClose: hideClose,
            label: this.literals[property] + ": " + dDate.toLocaleDateString(),
            property: "" + property,
            value: "" + this.dateToQueryParam(dDate)
        };
    };
    /**
     * @param {?} property
     * @param {?} value
     * @param {?=} length
     * @param {?=} hideClose
     * @param {?=} tradValue
     * @return {?}
     */
    DisclaimerUtil.prototype.makeDisclaimerFromMultiSelect = /**
     * @param {?} property
     * @param {?} value
     * @param {?=} length
     * @param {?=} hideClose
     * @param {?=} tradValue
     * @return {?}
     */
    function (property, value, length, hideClose, tradValue) {
        var _this = this;
        if (length === void 0) { length = 0; }
        if (hideClose === void 0) { hideClose = false; }
        if (tradValue === void 0) { tradValue = false; }
        /** @type {?} */
        var lstLabels = '';
        /** @type {?} */
        var lstValues = '';
        if (!value) {
            return { label: '', property: '', value: '' };
        }
        if (value.length === 0 || (length > 0 && value.length === length)) {
            lstLabels = this.literals['all'];
            lstValues = 'all';
        }
        else {
            value.map((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (lstLabels !== '') {
                    lstLabels = lstLabels + ", ";
                }
                if (lstValues !== '') {
                    lstValues = lstValues + ",";
                }
                lstLabels = "" + lstLabels + (tradValue ? _this.literals[item] : item);
                lstValues = "" + lstValues + item;
            }));
        }
        return {
            hideClose: hideClose,
            label: this.literals[property] + ": " + lstLabels,
            property: "" + property,
            value: "" + lstValues
        };
    };
    /**
     * @param {?} property
     * @param {?} value
     * @param {?=} length
     * @param {?=} hideClose
     * @return {?}
     */
    DisclaimerUtil.prototype.makeDisclaimerFromCheckboxGroup = /**
     * @param {?} property
     * @param {?} value
     * @param {?=} length
     * @param {?=} hideClose
     * @return {?}
     */
    function (property, value, length, hideClose) {
        var _this = this;
        if (length === void 0) { length = 0; }
        if (hideClose === void 0) { hideClose = false; }
        /** @type {?} */
        var lstLabels = '';
        /** @type {?} */
        var lstValues = '';
        if (!value) {
            return { label: '', property: '', value: '' };
        }
        if (value.length === 0) {
            lstLabels = this.literals['none'];
            lstValues = 'none';
        }
        else if (length > 0 && value.length === length) {
            lstLabels = this.literals['all'];
            lstValues = 'all';
        }
        else {
            value.map((/**
             * @param {?} checkParam
             * @return {?}
             */
            function (checkParam) {
                if (lstLabels !== '') {
                    lstLabels = lstLabels + ", ";
                }
                if (lstValues !== '') {
                    lstValues = lstValues + ",";
                }
                lstLabels = "" + lstLabels + _this.literals[checkParam];
                lstValues = "" + lstValues + checkParam;
            }));
        }
        return {
            hideClose: hideClose,
            label: this.literals[property] + ": " + lstLabels,
            property: "" + property,
            value: "" + lstValues
        };
    };
    /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} validateEqual
     * @return {?}
     */
    DisclaimerUtil.prototype.makeDisclaimerFromRangeCharacter = /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} validateEqual
     * @return {?}
     */
    function (property, value, hideClose, validateEqual) {
        if (hideClose === void 0) { hideClose = false; }
        if (validateEqual === void 0) { validateEqual = true; }
        if (value.valInitial === null || value.valInitial === undefined) {
            value.valInitial = value.iniInitial;
        }
        if (value.valFinal === null || value.valFinal === undefined) {
            value.valFinal = value.iniFinal;
        }
        if (validateEqual &&
            value.iniInitial === value.valInitial &&
            value.iniFinal === value.valFinal) {
            return { label: '', property: '', value: '' };
        }
        return {
            hideClose: hideClose,
            label: this.poI18nPipe.transform(this.literals['filterRange'], [this.literals[property], "'" + value.valInitial + "'", "'" + value.valFinal + "'"]),
            property: "" + property,
            value: value.valInitial + ";" + value.valFinal
        };
    };
    /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} validateEqual
     * @return {?}
     */
    DisclaimerUtil.prototype.makeDisclaimerFromRangeNumber = /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} validateEqual
     * @return {?}
     */
    function (property, value, hideClose, validateEqual) {
        if (hideClose === void 0) { hideClose = false; }
        if (validateEqual === void 0) { validateEqual = true; }
        if (value.valInitial === null || value.valInitial === undefined) {
            value.valInitial = value.iniInitial;
        }
        if (value.valFinal === null || value.valFinal === undefined) {
            value.valFinal = value.iniFinal;
        }
        if (validateEqual &&
            value.iniInitial === value.valInitial &&
            value.iniFinal === value.valFinal) {
            return { label: '', property: '', value: '' };
        }
        return {
            hideClose: hideClose,
            label: this.poI18nPipe.transform(this.literals['filterRange'], [this.literals[property], value.valInitial, value.valFinal]),
            property: "" + property,
            value: value.valInitial + ";" + value.valFinal
        };
    };
    /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} validateEqual
     * @return {?}
     */
    DisclaimerUtil.prototype.makeDisclaimerFromRangeDate = /**
     * @param {?} property
     * @param {?} value
     * @param {?=} hideClose
     * @param {?=} validateEqual
     * @return {?}
     */
    function (property, value, hideClose, validateEqual) {
        if (hideClose === void 0) { hideClose = false; }
        if (validateEqual === void 0) { validateEqual = true; }
        /** @type {?} */
        var valInitDate = this.ajustDate(value.valInitial);
        /** @type {?} */
        var valFinalDate = this.ajustDate(value.valFinal);
        if (!this.isValidDate(valInitDate)) {
            valInitDate = value.iniInitial;
            value.valInitial = value.iniInitial;
        }
        if (!this.isValidDate(valFinalDate)) {
            valFinalDate = value.iniFinal;
            value.valFinal = value.iniFinal;
        }
        if (validateEqual &&
            this.dateToQueryParam(value.iniInitial) === this.dateToQueryParam(valInitDate) &&
            this.dateToQueryParam(value.iniFinal) === this.dateToQueryParam(valFinalDate)) {
            return { label: '', property: '', value: '' };
        }
        return {
            hideClose: hideClose,
            label: this.poI18nPipe.transform(this.literals['filterRange'], [this.literals[property], valInitDate.toLocaleDateString(), valFinalDate.toLocaleDateString()]),
            property: "" + property,
            value: this.dateToQueryParam(valInitDate) + ";" + this.dateToQueryParam(valFinalDate)
        };
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    DisclaimerUtil.prototype.atzCharFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    function (disclaimers, property, defaultValue) {
        /** @type {?} */
        var returnValue = defaultValue;
        if (disclaimers && disclaimers.length > 0) {
            disclaimers.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.property === property) {
                    if (item.value === null) {
                        return false;
                    }
                    returnValue = item.value;
                    return true;
                }
                return false;
            }));
        }
        return returnValue;
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    DisclaimerUtil.prototype.atzNumberFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    function (disclaimers, property, defaultValue) {
        /** @type {?} */
        var returnValue = defaultValue;
        if (disclaimers && disclaimers.length > 0) {
            disclaimers.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.property === property) {
                    if (item.value === null) {
                        return false;
                    }
                    returnValue = item.value;
                    return true;
                }
                return false;
            }));
        }
        return returnValue;
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    DisclaimerUtil.prototype.atzBooleanFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    function (disclaimers, property, defaultValue) {
        /** @type {?} */
        var returnValue = defaultValue;
        if (disclaimers && disclaimers.length > 0) {
            /** @type {?} */
            var disclaimerFind = disclaimers.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.property === property; }));
            if (disclaimerFind && disclaimerFind.value !== null && disclaimerFind.value !== undefined) {
                if (disclaimerFind.value === 'true') {
                    return true;
                }
                else {
                    return false;
                }
            }
        }
        return returnValue;
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    DisclaimerUtil.prototype.atzDateFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    function (disclaimers, property, defaultValue) {
        var _this = this;
        /** @type {?} */
        var returnValue = this.ajustDate(defaultValue);
        if (disclaimers && disclaimers.length > 0) {
            disclaimers.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.property === property) {
                    if (item.value === null) {
                        return false;
                    }
                    returnValue = _this.ajustDate(item.value);
                    return true;
                }
                return false;
            }));
        }
        return returnValue;
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    DisclaimerUtil.prototype.atzMultiSelectCharFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    function (disclaimers, property, defaultValue) {
        return this.atzMultiSelectFromDisclamer('char', disclaimers, property, defaultValue);
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    DisclaimerUtil.prototype.atzMultiSelectNumberFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    function (disclaimers, property, defaultValue) {
        return this.atzMultiSelectFromDisclamer('number', disclaimers, property, defaultValue);
    };
    /**
     * @param {?} type
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    DisclaimerUtil.prototype.atzMultiSelectFromDisclamer = /**
     * @param {?} type
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    function (type, disclaimers, property, defaultValue) {
        /** @type {?} */
        var returnValue = [];
        defaultValue.map((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            returnValue.push(item.value);
        }));
        if (disclaimers && disclaimers.length > 0) {
            disclaimers.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.property === property) {
                    if (item.value === null || item.value === 'all' || item.value === 'none') {
                        return false;
                    }
                    returnValue = [];
                    item.value.split(',').forEach((/**
                     * @param {?} itemOpt
                     * @return {?}
                     */
                    function (itemOpt) {
                        if (type === 'char') {
                            returnValue.push(itemOpt);
                        }
                        if (type === 'number') {
                            returnValue.push(+itemOpt);
                        }
                    }));
                    return true;
                }
                return false;
            }));
        }
        return returnValue;
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} options
     * @return {?}
     */
    DisclaimerUtil.prototype.atzCheckboxFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} options
     * @return {?}
     */
    function (disclaimers, property, options) {
        /** @type {?} */
        var returnValue = [];
        options.map((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            returnValue.push(item.value);
        }));
        if (disclaimers && disclaimers.length > 0) {
            disclaimers.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.property === property) {
                    if (item.value === null || item.value === 'all' || item.value === 'none') {
                        return false;
                    }
                    returnValue = [];
                    item.value.split(',').forEach((/**
                     * @param {?} itemOpt
                     * @return {?}
                     */
                    function (itemOpt) {
                        returnValue.push(itemOpt);
                    }));
                    return true;
                }
                return false;
            }));
        }
        return returnValue;
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} rangeChar
     * @return {?}
     */
    DisclaimerUtil.prototype.atzRangeCharFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} rangeChar
     * @return {?}
     */
    function (disclaimers, property, rangeChar) {
        /** @type {?} */
        var value = '';
        if (disclaimers && disclaimers.length > 0) {
            disclaimers.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.property === property) {
                    if (item.value === null) {
                        return false;
                    }
                    value = item.value;
                    return true;
                }
                return false;
            }));
        }
        return this.atzRangeCharFromQueryParam(value, rangeChar);
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} rangeNum
     * @return {?}
     */
    DisclaimerUtil.prototype.atzRangeNumFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} rangeNum
     * @return {?}
     */
    function (disclaimers, property, rangeNum) {
        /** @type {?} */
        var value = '';
        if (disclaimers && disclaimers.length > 0) {
            disclaimers.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.property === property) {
                    if (item.value === null) {
                        return false;
                    }
                    value = item.value;
                    return true;
                }
                return false;
            }));
        }
        return this.atzRangeNumFromQueryParam(value, rangeNum);
    };
    /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} rangeDate
     * @return {?}
     */
    DisclaimerUtil.prototype.atzRangeDateFromDisclamer = /**
     * @param {?} disclaimers
     * @param {?} property
     * @param {?} rangeDate
     * @return {?}
     */
    function (disclaimers, property, rangeDate) {
        /** @type {?} */
        var value = '';
        if (disclaimers && disclaimers.length > 0) {
            disclaimers.find((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.property === property) {
                    if (item.value === null) {
                        return false;
                    }
                    value = item.value;
                    return true;
                }
                return false;
            }));
        }
        return this.atzRangeDateFromQueryParam(value, rangeDate);
    };
    /**
     * @param {?} value
     * @param {?} rangeChar
     * @return {?}
     */
    DisclaimerUtil.prototype.atzRangeCharFromQueryParam = /**
     * @param {?} value
     * @param {?} rangeChar
     * @return {?}
     */
    function (value, rangeChar) {
        rangeChar.valInitial = rangeChar.iniInitial;
        rangeChar.valFinal = rangeChar.iniFinal;
        if (!value || value === '') {
            return rangeChar;
        }
        rangeChar.valInitial = value.split(';')[0];
        rangeChar.valFinal = value.split(';')[1];
        if (rangeChar.valFinal === null || rangeChar.valFinal === undefined) {
            rangeChar.valFinal = rangeChar.iniFinal;
        }
        return rangeChar;
    };
    /**
     * @param {?} value
     * @param {?} rangeNum
     * @return {?}
     */
    DisclaimerUtil.prototype.atzRangeNumFromQueryParam = /**
     * @param {?} value
     * @param {?} rangeNum
     * @return {?}
     */
    function (value, rangeNum) {
        rangeNum.valInitial = rangeNum.iniInitial;
        rangeNum.valFinal = rangeNum.iniFinal;
        if (!value || value === '') {
            return rangeNum;
        }
        rangeNum.valInitial = +value.split(';')[0];
        rangeNum.valFinal = +value.split(';')[1];
        if (rangeNum.valFinal === null || rangeNum.valFinal === undefined) {
            rangeNum.valFinal = rangeNum.iniFinal;
        }
        return rangeNum;
    };
    /**
     * @param {?} value
     * @param {?} rangeDate
     * @return {?}
     */
    DisclaimerUtil.prototype.atzRangeDateFromQueryParam = /**
     * @param {?} value
     * @param {?} rangeDate
     * @return {?}
     */
    function (value, rangeDate) {
        rangeDate.valInitial = rangeDate.iniInitial;
        rangeDate.valFinal = rangeDate.iniFinal;
        if (!value || value === '') {
            return rangeDate;
        }
        rangeDate.valInitial = this.queryParamToDate(value.split(';')[0]);
        rangeDate.valFinal = this.queryParamToDate(value.split(';')[1]);
        if (rangeDate.valFinal === null || rangeDate.valFinal === undefined) {
            rangeDate.valFinal = rangeDate.iniFinal;
        }
        return rangeDate;
    };
    /**
     * @param {?} date
     * @return {?}
     */
    DisclaimerUtil.prototype.dateToQueryParam = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        /** @type {?} */
        var iDay = date.getDate();
        /** @type {?} */
        var iMonth = date.getMonth() + 1;
        /** @type {?} */
        var iYear = date.getFullYear();
        return iYear + "-" + this.pad(iMonth) + "-" + this.pad(iDay);
    };
    /**
     * @param {?} param
     * @return {?}
     */
    DisclaimerUtil.prototype.queryParamToDate = /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        /** @type {?} */
        var iDay = +param.split('-')[2];
        /** @type {?} */
        var iMonth = +param.split('-')[1];
        /** @type {?} */
        var iYear = +param.split('-')[0];
        return new Date(iYear, iMonth - 1, iDay);
    };
    /**
     * @param {?} date
     * @return {?}
     */
    DisclaimerUtil.prototype.isValidDate = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return date instanceof Date && !isNaN(date.getTime());
    };
    /**
     * @param {?} param
     * @return {?}
     */
    DisclaimerUtil.prototype.ajustDate = /**
     * @param {?} param
     * @return {?}
     */
    function (param) {
        if (param instanceof Date) {
            return param;
        }
        if (!param) {
            return param;
        }
        return this.queryParamToDate(param.split('T')[0]);
    };
    /**
     * @param {?} number
     * @return {?}
     */
    DisclaimerUtil.prototype.pad = /**
     * @param {?} number
     * @return {?}
     */
    function (number) {
        if (number < 10) {
            return '0' + number;
        }
        return number;
    };
    return DisclaimerUtil;
}());
export { DisclaimerUtil };
if (false) {
    /** @type {?} */
    DisclaimerUtil.prototype.poNotification;
    /** @type {?} */
    DisclaimerUtil.prototype.poI18nPipe;
    /** @type {?} */
    DisclaimerUtil.prototype.literals;
}
