// dts-backoffice-util/esm5/lib/model/session-info.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __values } from "tslib";
/**
 * @record
 */
export function ISessionInfo() { }
if (false) {
    /** @type {?} */
    ISessionInfo.prototype.diStatus;
    /** @type {?} */
    ISessionInfo.prototype.productVersion;
    /** @type {?} */
    ISessionInfo.prototype.userPwd;
    /** @type {?} */
    ISessionInfo.prototype.sessionId;
    /** @type {?} */
    ISessionInfo.prototype.userDomain;
    /** @type {?} */
    ISessionInfo.prototype.userName;
    /** @type {?} */
    ISessionInfo.prototype.userSO;
    /** @type {?} */
    ISessionInfo.prototype.timeout;
}
var SessionInfo = /** @class */ (function () {
    function SessionInfo(values) {
        if (values === void 0) { values = {}; }
        Object.assign(this, values);
    }
    Object.defineProperty(SessionInfo.prototype, "$diStatus", {
        get: /**
         * @return {?}
         */
        function () {
            return this.diStatus;
        },
        set: /**
         * @param {?} diStatus
         * @return {?}
         */
        function (diStatus) {
            this.diStatus = diStatus;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SessionInfo.prototype, "$productVersion", {
        get: /**
         * @return {?}
         */
        function () {
            return this.productVersion;
        },
        set: /**
         * @param {?} productVersion
         * @return {?}
         */
        function (productVersion) {
            this.productVersion = productVersion;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SessionInfo.prototype, "$userPwd", {
        get: /**
         * @return {?}
         */
        function () {
            return this.userPwd;
        },
        set: /**
         * @param {?} userPwd
         * @return {?}
         */
        function (userPwd) {
            this.userPwd = userPwd;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SessionInfo.prototype, "$sessionId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.sessionId;
        },
        set: /**
         * @param {?} sessionId
         * @return {?}
         */
        function (sessionId) {
            this.sessionId = sessionId;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SessionInfo.prototype, "$userDomain", {
        get: /**
         * @return {?}
         */
        function () {
            return this.userDomain;
        },
        set: /**
         * @param {?} userDomain
         * @return {?}
         */
        function (userDomain) {
            this.userDomain = userDomain;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SessionInfo.prototype, "$userName", {
        get: /**
         * @return {?}
         */
        function () {
            return this.userName;
        },
        set: /**
         * @param {?} userName
         * @return {?}
         */
        function (userName) {
            this.userName = userName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SessionInfo.prototype, "$userSO", {
        get: /**
         * @return {?}
         */
        function () {
            return this.userSO;
        },
        set: /**
         * @param {?} userSO
         * @return {?}
         */
        function (userSO) {
            this.userSO = userSO;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SessionInfo.prototype, "$timeout", {
        get: /**
         * @return {?}
         */
        function () {
            return this.timeout;
        },
        set: /**
         * @param {?} timeout
         * @return {?}
         */
        function (timeout) {
            this.timeout = timeout;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?=} json
     * @return {?}
     */
    SessionInfo.of = /**
     * @param {?=} json
     * @return {?}
     */
    function (json) {
        if (json === void 0) { json = {}; }
        return new SessionInfo(json);
    };
    /**
     * @return {?}
     */
    SessionInfo.empty = /**
     * @return {?}
     */
    function () {
        return new SessionInfo();
    };
    /**
     * @param {?=} json
     * @return {?}
     */
    SessionInfo.fromJson = /**
     * @param {?=} json
     * @return {?}
     */
    function (json) {
        var e_1, _a;
        if (json === void 0) { json = []; }
        /** @type {?} */
        var items = [];
        try {
            for (var json_1 = __values(json), json_1_1 = json_1.next(); !json_1_1.done; json_1_1 = json_1.next()) {
                var values = json_1_1.value;
                items.push(new SessionInfo(values));
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (json_1_1 && !json_1_1.done && (_a = json_1.return)) _a.call(json_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return items;
    };
    return SessionInfo;
}());
export { SessionInfo };
if (false) {
    /** @type {?} */
    SessionInfo.prototype.diStatus;
    /** @type {?} */
    SessionInfo.prototype.productVersion;
    /** @type {?} */
    SessionInfo.prototype.userPwd;
    /** @type {?} */
    SessionInfo.prototype.sessionId;
    /** @type {?} */
    SessionInfo.prototype.userDomain;
    /** @type {?} */
    SessionInfo.prototype.userName;
    /** @type {?} */
    SessionInfo.prototype.userSO;
    /** @type {?} */
    SessionInfo.prototype.timeout;
}
