// @po-ui/ng-templates/esm5/lib/components/po-page-blocked-user/po-page-blocked-user-contacts/po-page-blocked-user-contacts.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, Input, ViewChild } from '@angular/core';
var poPageBlockedUserContactItemMargin = 16;
var PoPageBlockedUserContactsComponent = /** @class */ (function () {
    function PoPageBlockedUserContactsComponent(changeDetector) {
        this.changeDetector = changeDetector;
        this.overflowItem = true;
    }
    Object.defineProperty(PoPageBlockedUserContactsComponent.prototype, "email", {
        get: function () {
            return this._email;
        },
        set: function (value) {
            this._email = value;
            this.checkContactItemWidth();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageBlockedUserContactsComponent.prototype, "phone", {
        get: function () {
            return this._phone;
        },
        set: function (value) {
            this._phone = value;
            this.checkContactItemWidth();
        },
        enumerable: true,
        configurable: true
    });
    PoPageBlockedUserContactsComponent.prototype.checkContactItemWidth = function () {
        this.overflowItem = true;
        if (this.phone && this.email) {
            this.changeDetector.detectChanges();
            var phoneWidth = this.phoneItem.nativeElement.offsetWidth;
            var mailWidth = this.mailItem.nativeElement.offsetWidth;
            var contactGroupHalfWidth = this.contactGroup.nativeElement.offsetWidth / 2 - poPageBlockedUserContactItemMargin;
            this.overflowItem =
                phoneWidth > contactGroupHalfWidth || mailWidth > contactGroupHalfWidth - poPageBlockedUserContactItemMargin;
        }
    };
    PoPageBlockedUserContactsComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input('p-email'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageBlockedUserContactsComponent.prototype, "email", null);
    __decorate([
        Input('p-phone'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageBlockedUserContactsComponent.prototype, "phone", null);
    __decorate([
        ViewChild('contactGroup', { static: true }),
        __metadata("design:type", ElementRef)
    ], PoPageBlockedUserContactsComponent.prototype, "contactGroup", void 0);
    __decorate([
        ViewChild('mailItem', { static: true }),
        __metadata("design:type", ElementRef)
    ], PoPageBlockedUserContactsComponent.prototype, "mailItem", void 0);
    __decorate([
        ViewChild('phoneItem', { static: true }),
        __metadata("design:type", ElementRef)
    ], PoPageBlockedUserContactsComponent.prototype, "phoneItem", void 0);
    PoPageBlockedUserContactsComponent = __decorate([
        Component({
            selector: 'po-page-blocked-user-contacts',
            template: "<div #contactGroup class=\"po-page-blocked-user-contact-group po-row po-mb-3\">\n  <div\n    class=\"po-page-blocked-user-contact-item\"\n    [class.po-invisible]=\"!phone\"\n    [ngClass]=\"overflowItem ? 'po-md-12' : 'po-md-6'\"\n  >\n    <a\n      class=\"po-page-blocked-user-link po-clickable\"\n      [class.po-page-blocked-user-contact-content-inline]=\"overflowItem\"\n      href=\"tel:{{ phone }}\"\n      target=\"_self\"\n    >\n      <div class=\"po-page-blocked-user-contact-group-item\">\n        <span class=\"po-page-blocked-user-contact-icon po-icon po-icon-telephone po-pr-1\"></span>\n        <div #phoneItem class=\"po-page-blocked-user-contact-text po-font-text\">{{ phone }}</div>\n      </div>\n    </a>\n  </div>\n  <div\n    class=\"po-page-blocked-user-contact-item po-page-blocked-user-contact-mail\"\n    [class.po-invisible]=\"!email\"\n    [ngClass]=\"overflowItem ? 'po-md-12' : 'po-md-6'\"\n  >\n    <a\n      class=\"po-page-blocked-user-link po-clickable\"\n      [class.po-page-blocked-user-contact-content-inline]=\"overflowItem\"\n      href=\"mailto:{{ email }}\"\n      target=\"_self\"\n    >\n      <div class=\"po-page-blocked-user-contact-group-item\">\n        <span class=\"po-page-blocked-user-contact-icon po-icon po-icon-mail po-pr-1\"></span>\n        <div #mailItem class=\"po-page-blocked-user-contact-text po-font-text\">{{ email }}</div>\n      </div>\n    </a>\n  </div>\n</div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], PoPageBlockedUserContactsComponent);
    return PoPageBlockedUserContactsComponent;
}());
export { PoPageBlockedUserContactsComponent };
