// @po-ui/ng-templates/esm5/lib/components/po-page-login/po-page-login-base.component.js
import { __assign, __decorate, __metadata } from "tslib";
import { EventEmitter, Input, Output, Directive } from '@angular/core';
import { Router } from '@angular/router';
import { convertToBoolean, convertToInt, getShortBrowserLanguage, isExternalLink, isTypeof, poLocaleDefault } from './../../utils/util';
import { PoPageLoginAuthenticationType } from './enums/po-page-login-authentication-type.enum';
import { PoPageLoginService } from './po-page-login.service';
var poPageLoginContentMaxLength = 40;
export var poPageLoginLiteralsDefault = {
    en: {
        title: 'Welcome',
        loginErrorPattern: 'Invalid Login',
        loginHint: "Your login user was given to you at your first day.\n    If you don't have this information contact support",
        loginPlaceholder: 'Insert your e-mail',
        passwordErrorPattern: 'Invalid Password',
        passwordPlaceholder: 'Insert your password',
        customFieldErrorPattern: 'Invalid value',
        customFieldPlaceholder: 'Please enter a value',
        rememberUser: 'Automatic login',
        rememberUserHint: 'You can disable this option in system configuration',
        submitLabel: 'Enter',
        submittedLabel: 'Loading...',
        forgotPassword: 'Forgot your Password?',
        highlightInfo: '',
        registerUrl: 'New register',
        titlePopover: 'Oops!',
        forgotYourPassword: 'Forgot your password?',
        ifYouTryHarder: 'If you try ',
        attempts: '{0} more time(s) ',
        yourUserWillBeBlocked: 'without success your user will be blocked and you will be left 24 hours without being able to access :(',
        createANewPasswordNow: 'Better create a new password now! You will be able to log into the system right away.',
        iForgotMyPassword: 'I forgot my password'
    },
    es: {
        title: 'Bienvenido',
        loginErrorPattern: 'Login inválido',
        loginHint: "Su usuario ha sido entregado para usted en su primer d\u00EDa.\n    Si no tiene esta informaci\u00F3n, p\u00F3ngase en contacto con el soporte t\u00E9cnico",
        loginPlaceholder: 'Inserte su e-mail',
        passwordErrorPattern: 'Contraseña inválida',
        passwordPlaceholder: 'Inserte su contraseña',
        customFieldErrorPattern: 'Valor no válido.',
        customFieldPlaceholder: 'Por favor introduzca un valor',
        rememberUser: 'Inicio de sesión automáticamente',
        rememberUserHint: 'Puede deshabilitar esta opción en el menú del sistema.',
        submitLabel: 'Entrar',
        submittedLabel: 'Cargando...',
        forgotPassword: 'Olvidaste tu contraseña?',
        highlightInfo: '',
        registerUrl: 'Nuevo registro',
        titlePopover: 'Opa!',
        forgotYourPassword: 'Olvidaste tu contraseña?',
        ifYouTryHarder: 'Si intenta más ',
        attempts: '{0} vez/veces ',
        yourUserWillBeBlocked: 'sin éxito su usuario sera bloqueado y usted vás permanecer 24 horas sin poder acceder a :(',
        createANewPasswordNow: '¡Mejor crear una nueva contraseña ahora! Usted podrá entrar en el sistema inmediatamente después.',
        iForgotMyPassword: 'Olvide mi contraseña'
    },
    pt: {
        title: 'Bem-vindo',
        loginErrorPattern: 'Login inválido',
        loginHint: "Seu usu\u00E1rio foi entregue a voc\u00EA no seu primeiro dia.\n    Caso n\u00E3o tenha mais essa informa\u00E7\u00E3o contacte o suporte",
        loginPlaceholder: 'Insira seu e-mail',
        passwordErrorPattern: 'Senha inválida',
        passwordPlaceholder: 'Insira sua senha',
        customFieldErrorPattern: 'Valor inválido.',
        customFieldPlaceholder: 'Por favor insira um valor',
        rememberUser: 'Logar automaticamente',
        rememberUserHint: 'Você pode desabilitar essa opção no menu do sistema',
        submitLabel: 'Entrar',
        submittedLabel: 'Carregando...',
        forgotPassword: 'Esqueceu sua senha?',
        highlightInfo: '',
        registerUrl: 'Novo registro',
        titlePopover: 'Opa!',
        forgotYourPassword: 'Esqueceu sua senha?',
        ifYouTryHarder: 'Se tentar mais ',
        attempts: '{0} vez(es) ',
        yourUserWillBeBlocked: 'sem sucesso seu usuário será bloqueado e você fica 24 horas sem poder acessar :(',
        createANewPasswordNow: 'Melhor criar uma senha nova agora! Você vai poder entrar no sistema logo em seguida.',
        iForgotMyPassword: 'Esqueci minha senha'
    },
    ru: {
        title: 'Добро пожаловать!',
        loginErrorPattern: 'Неверный логин',
        loginHint: "\u0412\u0430\u0448 \u043B\u043E\u0433\u0438\u043D \u0431\u044B\u043B \u043F\u0440\u0435\u0434\u043E\u0441\u0442\u0430\u0432\u043B\u0435\u043D \u0432\u0430\u043C \u0432 \u043F\u0435\u0440\u0432\u044B\u0439 \u0434\u0435\u043D\u044C.\n    \u0415\u0441\u043B\u0438 \u0443 \u0432\u0430\u0441 \u043D\u0435\u0442 \u044D\u0442\u043E\u0439 \u0438\u043D\u0444\u043E\u0440\u043C\u0430\u0446\u0438\u0438, \u043E\u0431\u0440\u0430\u0442\u0438\u0442\u0435\u0441\u044C \u0432 \u0441\u043B\u0443\u0436\u0431\u0443 \u043F\u043E\u0434\u0434\u0435\u0440\u0436\u043A\u0438",
        loginPlaceholder: 'Вставьте свой адрес электронной почты',
        passwordErrorPattern: 'Неверный пароль',
        passwordPlaceholder: 'Введите свой пароль',
        customFieldErrorPattern: 'Неверное значение.',
        customFieldPlaceholder: 'Пожалуйста, введите значение',
        rememberUser: 'Автоматический вход',
        rememberUserHint: 'Вы можете отключить эту опцию в конфигурации системы',
        submitLabel: 'Войти',
        submittedLabel: '3агрузка...',
        forgotPassword: 'Забыли пароль?',
        highlightInfo: '',
        registerUrl: 'Новый регистр',
        titlePopover: 'Ой!',
        forgotYourPassword: 'Забыли пароль?',
        ifYouTryHarder: 'Если вы безуспешно попытаетесь войти еще ',
        attempts: '{0} раз(а) ',
        yourUserWillBeBlocked: 'Ваш пользователь будет заблокирован, и Вы останетесь на 24 часа без возможности доступа :(',
        createANewPasswordNow: 'Лучше создайте новый пароль сейчас! Вы сможете сразу войти в систему.',
        iForgotMyPassword: 'Я забыл свой пароль'
    }
};
export var poPageLoginLiteralIn = {
    en: 'in',
    es: 'en',
    pt: 'em',
    ru: 'в'
};
export var poPageLoginLiteralTo = {
    en: 'to',
    es: 'al',
    pt: 'ao',
    ru: 'к'
};
/**
 * @description
 *
 * O componente `po-page-login` é utilizado como template para tela de login.
 * Com ele é possível definirmos valores para usuário, senha e definir ações para recuperação de senha e gravação de dados do usuário.
 * Também é possível inserir uma imagem em conjunto com um texto de destaque.
 *
 *
 * A propriedade `p-authentication-url` automatiza a rotina do componente e simplifica o processo para autenticação do usuário, bastando
 * definir uma url para requisição da autenticação. A flexibilidade e praticidade podem chegar a um nível em que o desenvolvimento
 * da aplicação no *client side* é desprovida de qualquer código-fonte relacionado à rotina de login de usuário.
 * Seu detalhamento para uso pode ser visto logo abaixo em *propriedades*.
 * Caso julgue necessário, pode-se também definir manualmente a rotina do componente.
 *
 *
 * Para que as imagens sejam exibidas corretamente, é necessário incluir o caminho delas ao projeto. Para isso, edite
 * o *assets* no arquivo **angular.json** da aplicação na seguinte ordem:
 * ```
 *   "assets": [
 *     "src/assets",
 *     "src/favicon.ico",
 *     {
 *       "glob": "**\/*",
 *       "input": "node_modules/@po-ui/style/images",
 *       "output": "assets/images"
 *     }
 *   ]
 * ```
 */
var PoPageLoginBaseComponent = /** @class */ (function () {
    function PoPageLoginBaseComponent(loginService, router) {
        this.loginService = loginService;
        this.router = router;
        this.allLoginErrors = [];
        this.allPasswordErrors = [];
        this.rememberUser = false;
        this.showExceededAttemptsWarning = false;
        this._authenticationType = PoPageLoginAuthenticationType.Basic;
        this._hideRememberUser = false;
        this._loading = false;
        this._loginErrors = [];
        this._passwordErrors = [];
        /**
         * Evento disparado quando o usuário alterar o input do campo login.
         *
         * Esse evento receberá como parâmetro uma variável do tipo `string` com o texto informado no campo.
         *
         * > Esta propriedade será ignorada se for definido valor para a propriedade `p-authentication-url`.
         */
        this.loginChange = new EventEmitter();
        /**
         * Evento disparado ao submeter o formulário de login (apertando `Enter` dentro dos campos ou pressionando o botão de confirmação).
         *
         * Esse evento receberá como parâmetro um objeto do tipo `PoPageLogin` com os dados informados no formulário.
         *
         * > Esta propriedade será ignorada se for definido valor para a propriedade `p-url-recovery`.
         *
         * > Para mais detalhes consulte a documentação sobre a interface `PoPageLogin` mais abaixo.
         */
        this.loginSubmit = new EventEmitter();
        /**
         * Evento disparado quando o usuário alterar o input do campo password.
         *
         * Esse evento receberá como parâmetro uma variável do tipo `string` com o texto informado no campo.
         *
         * > Esta propriedade será ignorada se for definido valor para a propriedade `p-authentication-url`.
         */
        this.passwordChange = new EventEmitter();
    }
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "customField", {
        get: function () {
            return this._customField;
        },
        /**
         * @optional
         *
         * @description
         *
         * Ao informar um valor do tipo `string`, o mesmo será aplicado como a chave do campo customizado e utilizará
         * os valores padrões contidos na propriedade `literals` como `customFieldErrorPattern` e `customFieldPlaceholder`.
         *
         * Existe a possibilidade de informar um objeto que segue a definição da interface `PoPageLoginCustomField`, onde
         * através dos parâmetros enviados pode gerar um `po-input`, `po-combo` especificamente para serviços
         * ou `po-select` para valores fixos.
         *
         * Abaixo seguem os exemplos de cada tipo de campo.
         *
         * `po-input`:
         *
         * ```
         * {
         *   property: 'domain',
         *   value: 'jv01',
         *   placeholder: 'Enter your domain',
         *   pattern: '[a-z]',
         *   errorPattern: 'Invalid value'
         * }
         * ```
         *
         * `po-combo`:
         *
         * ```
         * {
         *   property: 'domain',
         *   value: 'jv01',
         *   placeholder: 'Enter your domain',
         *   url: 'https://po-ui.io/sample/api/comboOption/domains',
         *   fieldValue: 'nickname'
         * }
         * ```
         *
         * `po-select`:
         *
         * ```
         * {
         *   property: 'domain',
         *   value: 'jv01',
         *   placeholder: 'Enter your domain',
         *   options: [{label: 'Domain 1', value: '1'}, {label: 'Domain 2', value: '2'}]
         * }
         * ```
         *
         * Caso o customField possua options, url e fieldValue preenchidos, será priorizado o po-select
         * utilizando o options.
         *
         */
        set: function (value) {
            if (value) {
                if (isTypeof(value, 'string')) {
                    this.customFieldType = 'input';
                    this._customField = value;
                    this.customFieldObject = this.getDefaultCustomFieldObject(value);
                    return;
                }
                if (isTypeof(value, 'object') && !Array.isArray(value) && value['property']) {
                    this._customField = value;
                    this.customFieldObject = value;
                    if (!this.customFieldObject.options && !this.customFieldObject.url) {
                        this.customFieldType = 'input';
                    }
                    else {
                        this.customFieldType = this.customFieldObject.options ? 'select' : 'combo';
                    }
                    return;
                }
            }
            this._customField = undefined;
            this.customFieldObject = undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "contactEmail", {
        get: function () {
            return this._contactEmail;
        },
        /**
         * @optional
         *
         * @description
         *
         * Personaliza o e-mail que é exibido na mensagem de dica de login padrão para contato de suporte.
         */
        set: function (value) {
            this._contactEmail = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "productName", {
        get: function () {
            return this._productName;
        },
        /**
         * @optional
         *
         * @description
         *
         * Valor customizado que sucede o título de boas-vindas.
         *
         * > Esta propriedade é sobreposta se `p-literals` contiver uma definição customizada para a literal `title`.
         *
         * > Veja mais sobre as literais na propriedade `p-literals`.
         *
         */
        set: function (value) {
            this._productName = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "environment", {
        get: function () {
            return this._environment;
        },
        /**
         * @optional
         *
         * @description
         * Adiciona uma `tag` abaixo do título que especifica o ambiente que o usuário está fazendo o login.
         *
         * > Essa propriedade limita o texto em 40 caracteres.
         */
        set: function (environment) {
            if (environment && environment.length > poPageLoginContentMaxLength) {
                this._environment = environment.substring(0, poPageLoginContentMaxLength);
            }
            else {
                this._environment = environment;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "exceededAttemptsWarning", {
        get: function () {
            return this._exceededAttemptsWarning;
        },
        /**
         * @optional
         *
         * @description
         * Exibe um aviso de bloqueio de acordo com a quantidade restante de tentativas.
         * O aviso será exibido somente se a quantidade for maior que zero.
         *
         * > Caso tenha algum valor atribuído para o atributo `p-authentication-url` e o retorno da requisição estiver atribuindo valor
         * para o `p-exceeded-attempts-warning`, o valor considerado será o do retorno da requisição.
         *
         * @default `0`
         */
        set: function (value) {
            this._exceededAttemptsWarning = convertToInt(value);
            this.showExceededAttemptsWarning = this.exceededAttemptsWarning > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "hideRememberUser", {
        get: function () {
            return this._hideRememberUser;
        },
        /**
         * @optional
         *
         * @description
         *
         * Esconde a função "Lembrar usuário" do formulário de login.
         *
         * Quando essa propriedade é setada com `true` a propriedade `rememberUser` enviada no evento `p-login-submit` será sempre
         * `false`.
         *
         * > Veja a propriedade `p-literals` para customizar a literal `rememberUser`.
         *
         * @default `false`
         */
        set: function (value) {
            this._hideRememberUser = value === '' ? true : convertToBoolean(value);
            if (this._hideRememberUser) {
                this.rememberUser = false;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "literals", {
        get: function () {
            return this._literals;
        },
        /**
         * @optional
         *
         * @description
         *
         * Objeto com as literais usadas no `po-page-login`.
         *
         * Existem duas maneiras de customizar o componente, passando um objeto com todas as literais disponíveis:
         *
         * ```
         *  const customLiterals: PoPageLoginLiterals = {
         *    attempts: '{0} vez(es) ',
         *    createANewPasswordNow: 'Melhor criar uma senha nova agora! Você vai poder entrar no sistema logo em seguida.',
         *    forgotPassword: 'Esqueceu sua senha?',
         *    forgotYourPassword: 'Esqueceu sua senha?',
         *    highlightInfo: '',
         *    iForgotMyPassword: 'Esqueci minha senha',
         *    ifYouTryHarder: 'Se tentar mais ',
         *    title: 'Seja bem-vindo',
         *    loginErrorPattern: 'Login obrigatório',
         *    loginHint: 'Caso não possua usuário entre em contato com o suporte',
         *    loginLabel: 'Insira seu usuário',
         *    loginPlaceholder: 'Insira seu usuário de acesso',
         *    passwordErrorPattern: 'Senha obrigatória',
         *    passwordLabel: 'Insira sua senha',
         *    passwordPlaceholder: 'Insira sua senha de acesso',
         *    customFieldErrorPattern: 'Campo customizado inválido',
         *    customFieldPlaceholder: 'Por favor insira um valor',
         *    registerUrl: 'Novo registro',
         *    rememberUser: 'Lembrar usuário',
         *    rememberUserHint: 'Esta opção pode ser desabilitada nas configurações do sistema',
         *    submitLabel: 'Acessar sistema',
         *    submittedLabel: 'Carregando...',
         *    titlePopover: 'Opa!',
         *    yourUserWillBeBlocked: 'sem sucesso seu usuário será bloqueado e você fica 24 horas sem poder acessar :('
         *  };
         * ```
         *
         * Ou passando apenas as literais que deseja customizar:
         *
         * ```
         *  const customLiterals: PoPageLoginLiterals = {
         *    loginPlaceholder: 'Insira seu usuário de acesso',
         *    passwordPlaceholder: 'Insira sua senha de acesso',
         *    submitLabel: 'Acessar sistema'
         *  };
         * ```
         *
         * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
         *
         * ```
         * <po-page-login
         *   [p-literals]="customLiterals">
         * </po-page-login>
         * ```
         *
         *  > O objeto padrão de literais será traduzido de acordo com o idioma do browser (pt, en, es).
         *  > É também possível alternar o objeto padrão de literais através do seletor de idiomas localizado na parte inferior do template,
         * nesse caso, há também a opção do idioma russo.
         */
        set: function (value) {
            this._literals = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "loading", {
        get: function () {
            return this._loading;
        },
        /**
         * @optional
         *
         * @description
         *
         * Habilita um estado de carregamento ao botão de *login*.
         *
         * > É necessário atribuir `true` à esta propriedade na função definida em `p-login-submit`.
         *
         * @default `false`
         */
        set: function (value) {
            this._loading = convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "login", {
        get: function () {
            return this._login;
        },
        /**
         * @optional
         *
         * @description
         *
         * Valor do modelo do campo de login.
         */
        set: function (value) {
            this._login = value;
            if (!this.authenticationUrl) {
                this.loginChange.emit(this._login);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "loginErrors", {
        get: function () {
            return this._loginErrors;
        },
        /**
         * @optional
         *
         * @description
         *
         * Atributo que recebe uma lista de erros e exibe abaixo do campo de login.
         */
        set: function (value) {
            this._loginErrors = value || [];
            this.setLoginErrors(this._loginErrors);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "passwordErrors", {
        get: function () {
            return this._passwordErrors;
        },
        /**
         * @optional
         *
         * @description
         *
         * Atributo que recebe uma lista de erros e exibe abaixo do campo de password.
         */
        set: function (value) {
            this._passwordErrors = value || [];
            this.setPasswordErrors(this._passwordErrors);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "recovery", {
        get: function () {
            return this._recovery;
        },
        /**
         * @optional
         *
         * @description
         *
         * Exibe um link abaixo do formulário de login para que os usuários da aplicação façam a recuperação dos dados de autenticação.
         *
         * A propriedade aceita os seguintes tipos:
         *
         * - **String**: informe uma url externa ou uma rota válida;
         * - **Function**: pode-se customizar a ação. Para esta possilidade basta atribuir:
         * ```
         * <po-page-login>
         *   [recovery]="this.myRecovery.bind(this)">
         * </po-page-login>
         * ```
         *
         * - **PoPageLoginRecovery**: cria-se vínculo automático com o template **po-modal-password-recovery**.
         *   O objeto deve conter a **url** para requisição dos recursos e pode-se definir o **tipo** de modal para recuperação de senha,
         *   **email** para contato e **máscara** do campo de telefone.
         */
        set: function (value) {
            this._recovery = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "registerUrl", {
        get: function () {
            return this._registerUrl;
        },
        /**
         * @optional
         *
         * @description
         *
         * Caso a aplicação tenha um link para novos cadastros, informe uma url externa ou uma rota válida, dessa
         * forma será exibido um link abaixo do formulário de login para os usuários da aplicação.
         *
         * Exemplos de valores válidos:
         * - **local**: `/home`
         * - **url externa**: `https://po-ui.io`
         *
         * > Veja a propriedade `p-literals` para customizar a literal `registerUrl`.
         */
        set: function (value) {
            this._registerUrl = isTypeof(value, 'string') ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "authenticationType", {
        get: function () {
            return this._authenticationType;
        },
        /**
         * @optional
         *
         * @description
         *
         * Atributo que recebe o tipo de esquema da autenticação, sendo suportados apenas os valores `Basic` e `Bearer`.
         *
         * > Caso o tipo definido seja `Basic`, o componente fará uma requisição `POST` contendo:
         *
         * ```
         * headers {
         *  Authorization: Basic base64(login:password)
         * }
         *
         * body {
         *  rememberUser: rememberUser
         * }
         * ```
         *
         * > Caso o tipo definido seja `Bearer`, o componente fará uma requisição `POST` contendo:
         *
         * ```
         * body {
         *  login: login,
         *  password: base64(password),
         *  rememberUser: rememberUser
         * }
         * ```
         *
         * @default `PoPageLoginAuthenticationType.Basic`
         */
        set: function (value) {
            this._authenticationType = Object.values(PoPageLoginAuthenticationType).includes(value)
                ? value
                : PoPageLoginAuthenticationType.Basic;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "authenticationUrl", {
        get: function () {
            return this._authenticationUrl;
        },
        /**
         * @optional
         *
         * @description
         *
         * Endpoint usado pelo template para requisição do recurso. Quando preenchido, o método `p-login-submit` será ignorado e o
         * componente adquirirá automatização para o processo de autenticação.
         *
         * ### Processos
         * Ao digitar um valor válido no campo de login/password e pressionar **Enter**, o componente fará uma requisição `POST`
         * na url especificada nesta propriedade passando o objeto contendo o valor definido pelo usuário:
         *
         * ```
         * headers {
         *  Authorization: Basic base64(login:password)
         * }
         *
         * body {
         *  rememberUser: rememberUser
         * }
         * ```
         *
         * Em caso de **sucesso**, o objeto de retorno é armazenado no `sessionStorage` e o usuário é redirecionado para a página inicial da
         * aplicação `/`.
         *
         * ```
         * 200:
         *  {
         *    user: user
         *  }
         * ```
         *
         * Em caso de **erro** na autenticação, espera-se o seguinte retorno:
         *
         * ```
         * 400/401
         *  {
         *    code: 400/401,
         *    message: message,
         *    detailedMessage: detailedMessage,
         *    helpUrl?: helpUrl
         *  }
         * ```
         *
         * > Pode-se atribuir uma quantidade máxima de tentativas restantes (maxAttemptsRemaining) para o atributo `p-exceeded-attempts-warning`,
         * assim como os avisos relacionados aos campos login e password (loginWarnings, passwordWarnings) para os atributos `p-login-errors` e
         * `p-password-errors` conforme retorno abaixo:
         *
         * ```
         * 400
         *  {
         *    code: 400/401,
         *    message: message,
         *    detailedMessage: detailedMessage,
         *    helpUrl?: helpUrl,
         *    maxAttemptsRemaining?: maxAttemptsRemaining,
         *    loginWarnings?: [loginWarnings],
         *    passwordWarnings?: [passwordWarnings]
         *  }
         * ```
         *
         * > Caso o valor atribuído para `p-exceeded-attempts-warning` seja igual a 0(zero), poderá ser passado um valor para o
         * atributo `p-blocked-url` e o usuário será redirecionado para uma tela de bloqueio.
         *
         * *Processo finalizado.*
         *
         * _______________
         *
         * #### Praticidade
         * As informações do serviço de autenticação também podem ser transmitidas diretamente pelas configuraçãos de rota e, desta maneira,
         * dispensa-se qualquer menção e/ou importação do componente `po-page-login` no restante da aplicação. O exemplo abaixo exemplifica
         * a forma dinâmica com a qual o template de tela de login pode ser gerado ao navegar para rota `/login`, e também como ele se comunica
         * com o serviço para efetuação do processo de autenticação do usuário e solicitação de nova senha.
         * Basta definir nas configurações de rota:
         *
         *
         * ```
         *   import { PoPageLoginComponent, PoPageLoginAthenticationType } from '@po-ui/ng-templates';
         *
         *   ...
         *   const routes: Routes = [
         *     {
         *       path: 'login', component: PoPageLoginComponent, data: {
         *         serviceApi: 'https://po-ui.io/sample/api/users/authentication',
         *         environment: 'development',
         *         recovery: {
         *           url: 'https://po-ui.io/sample/api/users',
         *           type: PoModalPasswordRecoveryType.All,
         *           contactMail: 'dev.po@po-ui.com',
         *           phoneMask: '9-999-999-9999'
         *         },
         *         registerUrl: '/new-password',
         *         authenticationType: PoPageLoginAthenticationType.Basic
         *       }
         *     }
         *     ...
         *   ];
         *
         *   @NgModule({
         *     imports: [RouterModule.forRoot(routes)],
         *     exports: [RouterModule]
         *   })
         *   export class AppRoutingModule { }
         * ```
         *
         *
         * O metadado `serviceApi` deve ser a **url** para requisição dos recursos de autenticação, o `environment` alimenta a propriedade
         * `p-environment`, `recovery` é a interface `PoPageLoginRecovery` responsável pelas especificações contidas na modal de recuperação de
         * senha, `registerUrl` alimenta a propriedade `p-register-url` e `authenticationType` que define a propriedade `p-authentication-type`.
         *
         * > É essencial que siga a nomenclatura dos atributos exemplificados acima para sua efetiva funcionalidade.
         *
         */
        set: function (value) {
            this._authenticationUrl = isTypeof(value, 'string') ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "blockedUrl", {
        get: function () {
            return this._blockedUrl;
        },
        /**
         * @optional
         *
         * @description
         *
         * Caso o valor atribuído para `p-exceeded-attempts-warning` seja igual a 0(zero) e a aplicação tenha um link de bloqueio de usuário,
         * informe uma url externa ou uma rota válida, dessa forma em caso de bloqueio o usuário será redirecionado.
         */
        set: function (value) {
            this._blockedUrl = isTypeof(value, 'string') ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "language", {
        get: function () {
            return this.selectedLanguage || getShortBrowserLanguage();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageLoginBaseComponent.prototype, "pageLoginLiterals", {
        get: function () {
            var loginHintWithContactEmail = this.contactEmail
                ? this.concatenateLoginHintWithContactEmail(this.contactEmail)
                : undefined;
            var titleWithProductName = this.productName ? this.concatenateTitleWithProductName(this.productName) : undefined;
            return __assign(__assign(__assign(__assign(__assign({}, poPageLoginLiteralsDefault[poLocaleDefault]), poPageLoginLiteralsDefault[this.language]), loginHintWithContactEmail), titleWithProductName), this.literals);
        },
        enumerable: true,
        configurable: true
    });
    PoPageLoginBaseComponent.prototype.ngOnDestroy = function () {
        if (this.loginSubscription) {
            this.loginSubscription.unsubscribe();
        }
    };
    PoPageLoginBaseComponent.prototype.closePopover = function () {
        this.showExceededAttemptsWarning = false;
    };
    PoPageLoginBaseComponent.prototype.onLoginSubmit = function () {
        var _this = this;
        var loginForm = {
            login: this.login,
            password: this.password,
            rememberUser: this.rememberUser
        };
        if (this.customField) {
            loginForm[this.customFieldObject.property] = this.customFieldObject.value;
        }
        if (this.authenticationUrl) {
            this.loginSubscription = this.loginService
                .onLogin(this.authenticationUrl, this.authenticationType, loginForm)
                .subscribe(function (data) {
                _this.setValuesToProperties();
                sessionStorage.setItem('PO_USER_LOGIN', JSON.stringify(data));
                _this.openInternalLink('/');
            }, function (error) {
                if (error.error.code === '400' || error.error.code === '401') {
                    _this.setValuesToProperties(error);
                    _this.redirectBlockedUrl(_this.exceededAttemptsWarning, _this.blockedUrl);
                }
            });
        }
        else {
            this.loginSubmit.emit(loginForm);
            this.showExceededAttemptsWarning = this.exceededAttemptsWarning > 0;
        }
    };
    PoPageLoginBaseComponent.prototype.getDefaultCustomFieldObject = function (property) {
        return { property: property };
    };
    PoPageLoginBaseComponent.prototype.openExternalLink = function (url) {
        window.open(url, '_blank');
    };
    PoPageLoginBaseComponent.prototype.openInternalLink = function (url) {
        this.router.navigate([url]);
    };
    PoPageLoginBaseComponent.prototype.redirectBlockedUrl = function (attempts, blockedUrl) {
        if (attempts === 0 && blockedUrl) {
            this.showExceededAttemptsWarning = false;
            isExternalLink(blockedUrl) ? this.openExternalLink(blockedUrl) : this.openInternalLink(blockedUrl);
        }
    };
    PoPageLoginBaseComponent.prototype.setValuesToProperties = function (result) {
        if (result) {
            this.exceededAttemptsWarning = result.error.maxAttemptsRemaining;
            this.loginErrors = result.error.loginWarnings;
            this.passwordErrors = result.error.passwordWarnings;
            this.blockedUrl = result.error.blockedUrl;
        }
        else {
            this.exceededAttemptsWarning = 0;
            this.loginErrors = [];
            this.passwordErrors = [];
            this.blockedUrl = '';
        }
    };
    PoPageLoginBaseComponent.ctorParameters = function () { return [
        { type: PoPageLoginService },
        { type: Router }
    ]; };
    __decorate([
        Input('p-custom-field'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PoPageLoginBaseComponent.prototype, "customField", null);
    __decorate([
        Input('p-background'),
        __metadata("design:type", String)
    ], PoPageLoginBaseComponent.prototype, "background", void 0);
    __decorate([
        Input('p-contact-email'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageLoginBaseComponent.prototype, "contactEmail", null);
    __decorate([
        Input('p-product-name'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageLoginBaseComponent.prototype, "productName", null);
    __decorate([
        Input('p-environment'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageLoginBaseComponent.prototype, "environment", null);
    __decorate([
        Input('p-exceeded-attempts-warning'),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], PoPageLoginBaseComponent.prototype, "exceededAttemptsWarning", null);
    __decorate([
        Input('p-hide-remember-user'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], PoPageLoginBaseComponent.prototype, "hideRememberUser", null);
    __decorate([
        Input('p-literals'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PoPageLoginBaseComponent.prototype, "literals", null);
    __decorate([
        Input('p-loading'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], PoPageLoginBaseComponent.prototype, "loading", null);
    __decorate([
        Input('p-login'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageLoginBaseComponent.prototype, "login", null);
    __decorate([
        Input('p-logo'),
        __metadata("design:type", String)
    ], PoPageLoginBaseComponent.prototype, "logo", void 0);
    __decorate([
        Input('p-login-pattern'),
        __metadata("design:type", String)
    ], PoPageLoginBaseComponent.prototype, "loginPattern", void 0);
    __decorate([
        Input('p-login-errors'),
        __metadata("design:type", Array),
        __metadata("design:paramtypes", [Array])
    ], PoPageLoginBaseComponent.prototype, "loginErrors", null);
    __decorate([
        Input('p-password-errors'),
        __metadata("design:type", Array),
        __metadata("design:paramtypes", [Array])
    ], PoPageLoginBaseComponent.prototype, "passwordErrors", null);
    __decorate([
        Input('p-password-pattern'),
        __metadata("design:type", String)
    ], PoPageLoginBaseComponent.prototype, "passwordPattern", void 0);
    __decorate([
        Input('p-recovery'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PoPageLoginBaseComponent.prototype, "recovery", null);
    __decorate([
        Input('p-register-url'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageLoginBaseComponent.prototype, "registerUrl", null);
    __decorate([
        Input('p-secondary-logo'),
        __metadata("design:type", String)
    ], PoPageLoginBaseComponent.prototype, "secondaryLogo", void 0);
    __decorate([
        Input('p-authentication-type'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageLoginBaseComponent.prototype, "authenticationType", null);
    __decorate([
        Input('p-authentication-url'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageLoginBaseComponent.prototype, "authenticationUrl", null);
    __decorate([
        Input('p-blocked-url'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], PoPageLoginBaseComponent.prototype, "blockedUrl", null);
    __decorate([
        Output('p-login-change'),
        __metadata("design:type", EventEmitter)
    ], PoPageLoginBaseComponent.prototype, "loginChange", void 0);
    __decorate([
        Output('p-login-submit'),
        __metadata("design:type", Object)
    ], PoPageLoginBaseComponent.prototype, "loginSubmit", void 0);
    __decorate([
        Output('p-password-change'),
        __metadata("design:type", EventEmitter)
    ], PoPageLoginBaseComponent.prototype, "passwordChange", void 0);
    PoPageLoginBaseComponent = __decorate([
        Directive(),
        __metadata("design:paramtypes", [PoPageLoginService, Router])
    ], PoPageLoginBaseComponent);
    return PoPageLoginBaseComponent;
}());
export { PoPageLoginBaseComponent };
